<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Project Query Data onto Discriminant Space of Reference Data — calculateDiscriminantSpace • scDiagnostics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Project Query Data onto Discriminant Space of Reference Data — calculateDiscriminantSpace"><meta property="og:description" content="This function projects query single-cell RNA-seq data onto the discriminant space defined by reference data. The reference data
is used to identify important variables and compute discriminant vectors, which are then used to project both reference and query
data. Similarity between the query and reference projections is assessed using cosine similarity and Mahalanobis distance.
The S3 plot method plots the projected reference and query data on discriminant spaces."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/QCandAnnotationScores.html">Visualization of QC and Annotation Scores</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/StatisticalMeasures.html">Statistical Measures to Assess Dataset Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection of Annotation Anomalies</a>
    </li>
    <li>
      <a href="../articles/CellDistancesDiagnostics.html">Calculation of Distances Between Specific Cells and Cell Populations</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Project Query Data onto Discriminant Space of Reference Data</h1>
    <small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/R/calculateDiscriminantSpace.R" class="external-link"><code>R/calculateDiscriminantSpace.R</code></a>, <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/R/plot.calculateDiscriminantSpaceObject.R" class="external-link"><code>R/plot.calculateDiscriminantSpaceObject.R</code></a></small>
    <div class="hidden name"><code>calculateDiscriminantSpace.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function projects query single-cell RNA-seq data onto the discriminant space defined by reference data. The reference data
is used to identify important variables and compute discriminant vectors, which are then used to project both reference and query
data. Similarity between the query and reference projections is assessed using cosine similarity and Mahalanobis distance.</p>
<p>The S3 plot method plots the projected reference and query data on discriminant spaces.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calculateDiscriminantSpace</span><span class="op">(</span></span>
<span>  <span class="va">reference_data</span>,</span>
<span>  query_data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">ref_cell_type_col</span>,</span>
<span>  query_cell_type_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cell_types <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_tree <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  n_top <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  eigen_threshold <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  calculate_metrics <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"logcounts"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'calculateDiscriminantSpaceObject'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">cell_types</span>, plot_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"scatterplot"</span>, <span class="st">"boxplot"</span><span class="op">)</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-reference-data">reference_data<a class="anchor" aria-label="anchor" href="#arg-reference-data"></a></dt>
<dd><p>A <code>SingleCellExperiment</code> object containing numeric expression matrix for the reference cells.</p></dd>


<dt id="arg-query-data">query_data<a class="anchor" aria-label="anchor" href="#arg-query-data"></a></dt>
<dd><p>A <code>SingleCellExperiment</code> object containing numeric expression matrix for the query cells.
If NULL, only the projected reference data is returned. Default is NULL.</p></dd>


<dt id="arg-ref-cell-type-col">ref_cell_type_col<a class="anchor" aria-label="anchor" href="#arg-ref-cell-type-col"></a></dt>
<dd><p>The column name in <code>reference_data</code> indicating cell type labels.</p></dd>


<dt id="arg-query-cell-type-col">query_cell_type_col<a class="anchor" aria-label="anchor" href="#arg-query-cell-type-col"></a></dt>
<dd><p>The column name in <code>query_data</code> indicating cell type labels.</p></dd>


<dt id="arg-cell-types">cell_types<a class="anchor" aria-label="anchor" href="#arg-cell-types"></a></dt>
<dd><p>A character vector specifying the cell types to plot. If not provided, all cell types will be plotted.</p></dd>


<dt id="arg-n-tree">n_tree<a class="anchor" aria-label="anchor" href="#arg-n-tree"></a></dt>
<dd><p>An integer specifying the number of trees for the random forest used in variable importance calculation.</p></dd>


<dt id="arg-n-top">n_top<a class="anchor" aria-label="anchor" href="#arg-n-top"></a></dt>
<dd><p>An integer specifying the number of top variables to select based on importance scores.</p></dd>


<dt id="arg-eigen-threshold">eigen_threshold<a class="anchor" aria-label="anchor" href="#arg-eigen-threshold"></a></dt>
<dd><p>A numeric value specifying the threshold for retaining eigenvalues in discriminant analysis.</p></dd>


<dt id="arg-calculate-metrics">calculate_metrics<a class="anchor" aria-label="anchor" href="#arg-calculate-metrics"></a></dt>
<dd><p>Parameter to determine if cosine similarity and Mahalanobis distance metrics should be computed. Default is FALSE.</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>A numeric value specifying the significance level for Mahalanobis distance cutoff.</p></dd>


<dt id="arg-assay-name">assay_name<a class="anchor" aria-label="anchor" href="#arg-assay-name"></a></dt>
<dd><p>Name of the assay on which to perform computations. Default is "logcounts".</p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>An object of class <code>calculateDiscriminantSpace</code> containing the projected data on the discriminant space..
Each element of the list represents a combination of cell types and datasets. Each element should contain 'ref_proj' and 'query_proj' data frames.</p></dd>


<dt id="arg-plot-type">plot_type<a class="anchor" aria-label="anchor" href="#arg-plot-type"></a></dt>
<dd><p>Type of plot to generate. Options are "scatterplot" and "boxplot". Default is "scatterplot".</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments to be passed to the plotting functions.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list with the following components for each cell type combination:</p>
<dl><dt>discriminant_eigenvalues</dt>
<dd><p>Eigenvalues from the discriminant analysis.</p></dd>

<dt>discriminant_eigenvectors</dt>
<dd><p>Eigenvectors from the discriminant analysis.</p></dd>

<dt>ref_proj</dt>
<dd><p>Reference data projected onto the discriminant space.</p></dd>

<dt>query_proj</dt>
<dd><p>Query data projected onto the discriminant space.</p></dd>

<dt>query_mahalanobis_dist</dt>
<dd><p>Mahalanobis distances of query projections.</p></dd>

<dt>mahalanobis_crit</dt>
<dd><p>Cutoff value for Mahalanobis distance significance.</p></dd>

<dt>query_cosine_similarity</dt>
<dd><p>Cosine similarity scores of query projections.</p></dd>


</dl><p>The S3 plot method returns a <code>ggplot</code> object representing the scatterplot or boxplot of the projected data.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The function performs the following steps for each pairwise combination of cell types:</p><ul><li><p>Identifies the top important variables to distinguish the two cell types from the reference data.</p></li>
<li><p>Computes the Ledoit-Wolf shrinkage estimate of the covariance matrix for each cell type using the top important genes.</p></li>
<li><p>Constructs within-class and between-class scatter matrices.</p></li>
<li><p>Solves the generalized eigenvalue problem to obtain discriminant vectors.</p></li>
<li><p>Projects both reference and query data onto the discriminant space.</p></li>
<li><p>Assesses similarity of the query data projection to the reference data using cosine similarity and Mahalanobis distance.</p></li>
</ul><p>The S3 plot method generates either a scatterplot or a boxplot to visualize the projected data onto the discriminant spaces.
For scatterplot, each point represents a projected data point, and colors are used to differentiate between different cell types
and datasets. For boxplot, the distribution of the projected data values for each cell type is shown, separated by datasets.</p>
    </div>
    <div id="references">
    <h2>References</h2>

<ul><li><p>Fisher, R. A. (1936). "The Use of Multiple Measurements in Taxonomic Problems". *Annals of Eugenics*. 7 (2): 179–188. doi:10.1111/j.1469-1809.1936.tb02137.x.</p></li>
<li><p>Hastie, T., Tibshirani, R., &amp; Friedman, J. (2009). *The Elements of Statistical Learning: Data Mining, Inference, and Prediction*. Springer. Chapter 4: Linear Methods for Classification.</p></li>
<li><p>Ledoit, O., &amp; Wolf, M. (2004). "A well-conditioned estimator for large-dimensional covariance matrices". *Journal of Multivariate Analysis*. 88 (2): 365–411. doi:10.1016/S0047-259X(03)00096-4.</p></li>
<li><p>De Maesschalck, R., Jouan-Rimbaud, D., &amp; Massart, D. L. (2000). "The Mahalanobis distance". *Chemometrics and Intelligent Laboratory Systems*. 50 (1): 1–18. doi:10.1016/S0169-7439(99)00047-7.</p></li>
<li><p>Breiman, L. (2001). "Random Forests". *Machine Learning*. 45 (1): 5–32. doi:10.1023/A:1010933404324.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>plot.calculateDiscriminantSpaceObject</code></p>
<p><code>calculateDiscriminantSpace</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Anthony Christidis, <a href="mailto:anthony-alexander_christidis@hms.harvard.edu">anthony-alexander_christidis@hms.harvard.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Load data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Compute important variables for all pairwise cell comparisons</span></span></span>
<span class="r-in"><span><span class="va">disc_output</span> <span class="op">&lt;-</span> <span class="fu">calculateDiscriminantSpace</span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span></span>
<span class="r-in"><span>                                          query_data <span class="op">=</span> <span class="va">query_data</span>,</span></span>
<span class="r-in"><span>                                          query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span></span>
<span class="r-in"><span>                                          ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span></span>
<span class="r-in"><span>                                          n_tree <span class="op">=</span> <span class="fl">500</span>,</span></span>
<span class="r-in"><span>                                          n_top <span class="op">=</span> <span class="fl">50</span>,</span></span>
<span class="r-in"><span>                                          eigen_threshold  <span class="op">=</span> <span class="fl">1e-1</span>,</span></span>
<span class="r-in"><span>                                          calculate_metrics <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>                                          alpha <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Generate scatter and boxplot</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">disc_output</span>, plot_type <span class="op">=</span> <span class="st">"scatterplot"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calculateDiscriminantSpace-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">disc_output</span>, cell_types <span class="op">=</span> <span class="st">"CD4-CD8"</span>, plot_type <span class="op">=</span> <span class="st">"boxplot"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calculateDiscriminantSpace-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Check comparison</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span>Expert_Annotation <span class="op">=</span> <span class="va">query_data</span><span class="op">$</span><span class="va">expert_annotation</span>, SingleR <span class="op">=</span> <span class="va">query_data</span><span class="op">$</span><span class="va">SingleR_annotation</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  SingleR</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Expert_Annotation B_and_plasma CD4 CD8 Myeloid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      B_and_plasma           97   0   0       5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      CD4                     0 132   2       4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      CD8                     2 105 109      14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      Myeloid                 0   0   0      33</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

