<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate Graph Community Integration Diagnostics — calculateGraphIntegration • scDiagnostics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Calculate Graph Community Integration Diagnostics — calculateGraphIntegration"><meta property="og:description" content="This function performs graph-based community detection to identify annotation inconsistencies
by detecting query-only communities, true cross-cell-type mixing patterns, and local
annotation inconsistencies based on immediate neighborhood analysis.
The S3 plot method generates visualizations of annotation consistency diagnostics,
including query-only communities, cross-cell-type mixing, and local annotation inconsistencies."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection and Analysis of Annotation Anomalies</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate Graph Community Integration Diagnostics</h1>
    <small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/R/calculateGraphIntegration.R" class="external-link"><code>R/calculateGraphIntegration.R</code></a>, <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/R/plot.calculateGraphIntegrationObject.R" class="external-link"><code>R/plot.calculateGraphIntegrationObject.R</code></a></small>
    <div class="hidden name"><code>calculateGraphIntegration.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function performs graph-based community detection to identify annotation inconsistencies
by detecting query-only communities, true cross-cell-type mixing patterns, and local
annotation inconsistencies based on immediate neighborhood analysis.</p>
<p>The S3 plot method generates visualizations of annotation consistency diagnostics,
including query-only communities, cross-cell-type mixing, and local annotation inconsistencies.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calculateGraphIntegration</span><span class="op">(</span></span>
<span>  <span class="va">query_data</span>,</span>
<span>  <span class="va">reference_data</span>,</span>
<span>  <span class="va">query_cell_type_col</span>,</span>
<span>  <span class="va">ref_cell_type_col</span>,</span>
<span>  cell_types <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>  k_neighbors <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  resolution <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  min_cells_per_community <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  min_cells_per_celltype <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  high_query_prop_threshold <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  cross_type_threshold <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>  local_consistency_threshold <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  local_confidence_threshold <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  max_cells_query <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  max_cells_ref <span class="op">=</span> <span class="fl">5000</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'calculateGraphIntegrationObject'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  plot_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"community_network"</span>, <span class="st">"cell_network"</span>, <span class="st">"community_data"</span>, <span class="st">"summary"</span>,</span>
<span>    <span class="st">"local_issues"</span>, <span class="st">"annotation_issues"</span><span class="op">)</span>,</span>
<span>  color_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_type"</span>, <span class="st">"community_type"</span><span class="op">)</span>,</span>
<span>  max_nodes <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  point_size <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  exclude_reference_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-query-data">query_data<a class="anchor" aria-label="anchor" href="#arg-query-data"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code> object containing numeric expression matrix for the query cells.</p></dd>


<dt id="arg-reference-data">reference_data<a class="anchor" aria-label="anchor" href="#arg-reference-data"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code> object containing numeric expression matrix for the reference cells.</p></dd>


<dt id="arg-query-cell-type-col">query_cell_type_col<a class="anchor" aria-label="anchor" href="#arg-query-cell-type-col"></a></dt>
<dd><p>A character string specifying the column name in the query dataset containing cell type annotations.</p></dd>


<dt id="arg-ref-cell-type-col">ref_cell_type_col<a class="anchor" aria-label="anchor" href="#arg-ref-cell-type-col"></a></dt>
<dd><p>A character string specifying the column name in the reference dataset containing cell type annotations.</p></dd>


<dt id="arg-cell-types">cell_types<a class="anchor" aria-label="anchor" href="#arg-cell-types"></a></dt>
<dd><p>A character vector specifying the cell types to include in the analysis. If NULL, all cell types are included.</p></dd>


<dt id="arg-pc-subset">pc_subset<a class="anchor" aria-label="anchor" href="#arg-pc-subset"></a></dt>
<dd><p>A vector specifying the subset of principal components to use in the analysis. Default is 1:10.</p></dd>


<dt id="arg-k-neighbors">k_neighbors<a class="anchor" aria-label="anchor" href="#arg-k-neighbors"></a></dt>
<dd><p>An integer specifying the number of nearest neighbors for graph construction. Default is 30.</p></dd>


<dt id="arg-assay-name">assay_name<a class="anchor" aria-label="anchor" href="#arg-assay-name"></a></dt>
<dd><p>Name of the assay on which to perform computations. Default is "logcounts".</p></dd>


<dt id="arg-resolution">resolution<a class="anchor" aria-label="anchor" href="#arg-resolution"></a></dt>
<dd><p>Resolution parameter for Leiden clustering. Default is 0.15 for fewer, larger communities.</p></dd>


<dt id="arg-min-cells-per-community">min_cells_per_community<a class="anchor" aria-label="anchor" href="#arg-min-cells-per-community"></a></dt>
<dd><p>Minimum number of cells required for a community to be analyzed. Default is 10.</p></dd>


<dt id="arg-min-cells-per-celltype">min_cells_per_celltype<a class="anchor" aria-label="anchor" href="#arg-min-cells-per-celltype"></a></dt>
<dd><p>Minimum number of cells required per cell type for inclusion. Default is 20.</p></dd>


<dt id="arg-high-query-prop-threshold">high_query_prop_threshold<a class="anchor" aria-label="anchor" href="#arg-high-query-prop-threshold"></a></dt>
<dd><p>Minimum proportion of query cells to consider a community "query-only". Default is 0.9.</p></dd>


<dt id="arg-cross-type-threshold">cross_type_threshold<a class="anchor" aria-label="anchor" href="#arg-cross-type-threshold"></a></dt>
<dd><p>Minimum proportion needed to flag cross-cell-type mixing. Default is 0.1.</p></dd>


<dt id="arg-local-consistency-threshold">local_consistency_threshold<a class="anchor" aria-label="anchor" href="#arg-local-consistency-threshold"></a></dt>
<dd><p>Minimum proportion of reference neighbors that should support a query cell's annotation. Default is 0.6.</p></dd>


<dt id="arg-local-confidence-threshold">local_confidence_threshold<a class="anchor" aria-label="anchor" href="#arg-local-confidence-threshold"></a></dt>
<dd><p>Minimum confidence difference needed to suggest re-annotation. Default is 0.2.</p></dd>


<dt id="arg-max-cells-query">max_cells_query<a class="anchor" aria-label="anchor" href="#arg-max-cells-query"></a></dt>
<dd><p>Maximum number of query cells to retain after cell type filtering. If NULL,
no downsampling of query cells is performed. Default is 5000.</p></dd>


<dt id="arg-max-cells-ref">max_cells_ref<a class="anchor" aria-label="anchor" href="#arg-max-cells-ref"></a></dt>
<dd><p>Maximum number of reference cells to retain after cell type filtering. If NULL,
no downsampling of reference cells is performed. Default is 5000.</p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>An object of class <code>calculateGraphIntegrationObject</code> containing the diagnostic results.</p></dd>


<dt id="arg-plot-type">plot_type<a class="anchor" aria-label="anchor" href="#arg-plot-type"></a></dt>
<dd><p>Character string specifying visualization type. Options: "community_network" (default),
"cell_network", "community_data", "summary", "local_issues", or "annotation_issues".</p></dd>


<dt id="arg-color-by">color_by<a class="anchor" aria-label="anchor" href="#arg-color-by"></a></dt>
<dd><p>Character string specifying the variable to use for coloring points/elements if `plot_type` is
"community_network" or "cell_network". Default is "cell_type".</p></dd>


<dt id="arg-max-nodes">max_nodes<a class="anchor" aria-label="anchor" href="#arg-max-nodes"></a></dt>
<dd><p>Maximum number of nodes to display for performance. Default is 2000.</p></dd>


<dt id="arg-point-size">point_size<a class="anchor" aria-label="anchor" href="#arg-point-size"></a></dt>
<dd><p>Point size for graph nodes. Default is 0.8.</p></dd>


<dt id="arg-exclude-reference-only">exclude_reference_only<a class="anchor" aria-label="anchor" href="#arg-exclude-reference-only"></a></dt>
<dd><p>Logical indicating whether to exclude reference-only communities/cells from visualization.
Default is FALSE.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed to ggplot2 functions.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list containing:</p>
<dl><dt>high_query_prop_analysis</dt>
<dd><p>Analysis of communities with only query cells</p></dd>

<dt>cross_type_mixing</dt>
<dd><p>Analysis of communities with true query-reference cross-cell-type mixing</p></dd>

<dt>local_annotation_inconsistencies</dt>
<dd><p>Local neighborhood-based annotation inconsistencies</p></dd>

<dt>local_inconsistency_summary</dt>
<dd><p>Summary of local inconsistencies by cell type</p></dd>

<dt>community_composition</dt>
<dd><p>Detailed composition of each community</p></dd>

<dt>annotation_consistency</dt>
<dd><p>Summary of annotation consistency issues</p></dd>

<dt>overall_metrics</dt>
<dd><p>Overall diagnostic metrics</p></dd>

<dt>graph_info</dt>
<dd><p>Graph structure information for plotting</p></dd>

<dt>parameters</dt>
<dd><p>Analysis parameters used</p></dd>

</dl><p>The list is assigned the class <code>"calculateGraphIntegration"</code>.</p>
<p>A <code>ggplot</code> object showing integration diagnostics.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The function performs three types of analysis: (1) Communities containing only query cells,
(2) Communities where query cells are mixed with reference cells of different cell types
WITHOUT any reference cells of the same type, and (3) Local analysis of each query cell's
immediate neighbors to detect annotation inconsistencies even within mixed communities.</p>
<p>The S3 plot method creates optimized visualizations showing different types of annotation
issues including community-level and local neighborhood-level inconsistencies.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>calculateGraphIntegration</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Anthony Christidis, <a href="mailto:anthony-alexander_christidis@hms.harvard.edu">anthony-alexander_christidis@hms.harvard.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Load data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Remove a cell type (Myeloid)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SingleCellExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SummarizedExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: MatrixGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: matrixStats</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MatrixGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedSds, rowWeightedVars</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomicRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: stats4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: generics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘generics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     setequal, union</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘BiocGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.data.frame, basename, cbind, colnames, dirname, do.call,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     duplicated, eval, evalq, get, grep, grepl, is.unsorted, lapply,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     mapply, match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rank, rbind, rownames, sapply, saveRDS, table, tapply, unique,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     unsplit, which.max, which.min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: S4Vectors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘S4Vectors’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:utils’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     findMatches</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     I, expand.grid, unname</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biobase</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to Bioconductor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘Biobase’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:MatrixGenerics’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     anyMissing, rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: scuttle</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: ggplot2</span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SingleR-inc/SingleR" class="external-link">SingleR</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">reference_data</span> <span class="op">&lt;-</span> <span class="va">reference_data</span><span class="op">[</span>, <span class="va">reference_data</span><span class="op">$</span><span class="va">expert_annotation</span> <span class="op">!=</span> <span class="st">"Myeloid"</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">reference_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">reference_data</span>, ncomponents <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">SingleR_annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR</a></span><span class="op">(</span><span class="va">query_data</span>, <span class="va">reference_data</span>,</span></span>
<span class="r-in"><span>                              labels <span class="op">=</span> <span class="va">reference_data</span><span class="op">$</span><span class="va">expert_annotation</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">query_data</span><span class="op">$</span><span class="va">SingleR_annotation</span> <span class="op">&lt;-</span> <span class="va">SingleR_annotation</span><span class="op">$</span><span class="va">labels</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Check annotation data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span>Expert <span class="op">=</span> <span class="va">query_data</span><span class="op">$</span><span class="va">expert_annotation</span>, SingleR <span class="op">=</span> <span class="va">query_data</span><span class="op">$</span><span class="va">SingleR_annotation</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               SingleR</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Expert         B_and_plasma CD4 CD8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   B_and_plasma           98   3   1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CD4                     0 133   5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CD8                     0  44 186</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Myeloid                21  12   0</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run comprehensive annotation consistency diagnostics</span></span></span>
<span class="r-in"><span><span class="va">graph_diagnostics</span> <span class="op">&lt;-</span> <span class="fu">calculateGraphIntegration</span><span class="op">(</span></span></span>
<span class="r-in"><span>    query_data <span class="op">=</span> <span class="va">query_data</span>,</span></span>
<span class="r-in"><span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span></span>
<span class="r-in"><span>    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span></span>
<span class="r-in"><span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span></span>
<span class="r-in"><span>   pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span></span>
<span class="r-in"><span>   k_neighbors <span class="op">=</span> <span class="fl">30</span>,</span></span>
<span class="r-in"><span>    resolution <span class="op">=</span> <span class="fl">0.1</span>,</span></span>
<span class="r-in"><span>    high_query_prop_threshold <span class="op">=</span> <span class="fl">0.9</span>,</span></span>
<span class="r-in"><span>    cross_type_threshold <span class="op">=</span> <span class="fl">0.15</span>,</span></span>
<span class="r-in"><span>    local_consistency_threshold <span class="op">=</span> <span class="fl">0.6</span>,</span></span>
<span class="r-in"><span>    local_confidence_threshold <span class="op">=</span> <span class="fl">0.2</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Look at main output</span></span></span>
<span class="r-in"><span><span class="va">graph_diagnostics</span><span class="op">$</span><span class="va">overall_metrics</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $total_communities</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 25</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $high_query_prop_communities</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $true_cross_type_communities</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $total_high_query_prop_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 34</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $total_true_cross_mixing_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 41</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $total_locally_inconsistent_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 47</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $modularity</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.5956145</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $mean_query_isolation_rate</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.08245798</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $mean_true_cross_mixing_rate</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.09674078</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $mean_local_inconsistency_rate</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.08798728</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Network graph showing all issue types (color by cell type)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">graph_diagnostics</span>, plot_type <span class="op">=</span> <span class="st">"community_network"</span>, color_by <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calculateGraphIntegration-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Network graph showing all issue types</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">graph_diagnostics</span>, plot_type <span class="op">=</span> <span class="st">"cell_network"</span>,</span></span>
<span class="r-in"><span>     max_nodes <span class="op">=</span> <span class="fl">2000</span>, color_by <span class="op">=</span> <span class="st">"community_type"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calculateGraphIntegration-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Network graph showing all issue types</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">graph_diagnostics</span>, plot_type <span class="op">=</span> <span class="st">"community_data"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calculateGraphIntegration-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Summary bar chart of all issues by cell type</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">graph_diagnostics</span>, plot_type <span class="op">=</span> <span class="st">"summary"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calculateGraphIntegration-4.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Focus on local annotation inconsistencies</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">graph_diagnostics</span>, plot_type <span class="op">=</span> <span class="st">"local_issues"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calculateGraphIntegration-5.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Overall annotation issues overview</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">graph_diagnostics</span>, plot_type <span class="op">=</span> <span class="st">"annotation_issues"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calculateGraphIntegration-6.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

