<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate Top Loading Gene Expression Shifts — calculateTopLoadingGeneShifts • scDiagnostics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Calculate Top Loading Gene Expression Shifts — calculateTopLoadingGeneShifts"><meta property="og:description" content="This function identifies genes with the highest loadings for specified principal components
and performs statistical tests to detect distributional differences between query and reference data.
This function creates boxplots showing expression distributions for top loading genes
that exhibit distributional differences between query and reference datasets."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection and Analysis of Annotation Anomalies</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate Top Loading Gene Expression Shifts</h1>
    <small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/R/calculateTopLoadingGeneShifts.R" class="external-link"><code>R/calculateTopLoadingGeneShifts.R</code></a>, <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/R/plot.calculateTopLoadingGeneShifts.R" class="external-link"><code>R/plot.calculateTopLoadingGeneShifts.R</code></a></small>
    <div class="hidden name"><code>calculateTopLoadingGeneShifts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function identifies genes with the highest loadings for specified principal components
and performs statistical tests to detect distributional differences between query and reference data.</p>
<p>This function creates boxplots showing expression distributions for top loading genes
that exhibit distributional differences between query and reference datasets.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calculateTopLoadingGeneShifts</span><span class="op">(</span></span>
<span>  <span class="va">query_data</span>,</span>
<span>  <span class="va">reference_data</span>,</span>
<span>  <span class="va">query_cell_type_col</span>,</span>
<span>  <span class="va">ref_cell_type_col</span>,</span>
<span>  cell_types <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>  n_top_loadings <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  p_value_threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  adjust_method <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  max_cells <span class="op">=</span> <span class="fl">2500</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'calculateTopLoadingGeneShiftsObject'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">cell_type</span>,</span>
<span>  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,</span>
<span>  plot_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p_adjusted"</span>, <span class="st">"top_loading"</span><span class="op">)</span>,</span>
<span>  n_genes <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  significance_threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-query-data">query_data<a class="anchor" aria-label="anchor" href="#arg-query-data"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code> object containing numeric expression matrix for the query cells.</p></dd>


<dt id="arg-reference-data">reference_data<a class="anchor" aria-label="anchor" href="#arg-reference-data"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code> object containing numeric expression matrix for the reference cells.</p></dd>


<dt id="arg-query-cell-type-col">query_cell_type_col<a class="anchor" aria-label="anchor" href="#arg-query-cell-type-col"></a></dt>
<dd><p>The column name in the <code>colData</code> of <code>query_data</code> that identifies the cell types.</p></dd>


<dt id="arg-ref-cell-type-col">ref_cell_type_col<a class="anchor" aria-label="anchor" href="#arg-ref-cell-type-col"></a></dt>
<dd><p>The column name in the <code>colData</code> of <code>reference_data</code> that identifies the cell types.</p></dd>


<dt id="arg-cell-types">cell_types<a class="anchor" aria-label="anchor" href="#arg-cell-types"></a></dt>
<dd><p>A character vector specifying the cell types to analyze. If NULL, all common cell types are used.</p></dd>


<dt id="arg-pc-subset">pc_subset<a class="anchor" aria-label="anchor" href="#arg-pc-subset"></a></dt>
<dd><p>A numeric vector specifying which principal components to plot. Default is 1:3.</p></dd>


<dt id="arg-n-top-loadings">n_top_loadings<a class="anchor" aria-label="anchor" href="#arg-n-top-loadings"></a></dt>
<dd><p>Number of top loading genes to analyze per PC. Default is 50.</p></dd>


<dt id="arg-p-value-threshold">p_value_threshold<a class="anchor" aria-label="anchor" href="#arg-p-value-threshold"></a></dt>
<dd><p>P-value threshold for statistical significance. Default is 0.05.</p></dd>


<dt id="arg-adjust-method">adjust_method<a class="anchor" aria-label="anchor" href="#arg-adjust-method"></a></dt>
<dd><p>Method for multiple testing correction. Default is "fdr".</p></dd>


<dt id="arg-assay-name">assay_name<a class="anchor" aria-label="anchor" href="#arg-assay-name"></a></dt>
<dd><p>Name of the assay on which to perform computations. Default is "logcounts".</p></dd>


<dt id="arg-max-cells">max_cells<a class="anchor" aria-label="anchor" href="#arg-max-cells"></a></dt>
<dd><p>Maximum number of cells to retain. If the object has fewer cells, it is returned unchanged.
Default is 2500.</p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>An object of class <code>calculateTopLoadingGeneShiftsObject</code> containing the output of the <code>calculateTopLoadingGeneShifts</code> function.</p></dd>


<dt id="arg-cell-type">cell_type<a class="anchor" aria-label="anchor" href="#arg-cell-type"></a></dt>
<dd><p>A character string specifying the cell type to plot (must be exactly one).</p></dd>


<dt id="arg-plot-by">plot_by<a class="anchor" aria-label="anchor" href="#arg-plot-by"></a></dt>
<dd><p>A character string specifying gene selection method. Either "top_loading" or "p_adjusted". Default is "p_adjusted".</p></dd>


<dt id="arg-n-genes">n_genes<a class="anchor" aria-label="anchor" href="#arg-n-genes"></a></dt>
<dd><p>Number of genes to show per PC. Default is 10.</p></dd>


<dt id="arg-significance-threshold">significance_threshold<a class="anchor" aria-label="anchor" href="#arg-significance-threshold"></a></dt>
<dd><p>P-adjusted threshold for significance annotation. Default is 0.05.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments to be passed to the plotting functions.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list containing:</p><ul><li><p>PC results: Named elements for each PC (e.g., "PC1", "PC2") containing data frames with analysis results</p></li>
<li><p>expression_data: Matrix of expression values for all analyzed genes (genes × cells)</p></li>
<li><p>cell_metadata: Data frame with columns: cell_id, dataset, cell_type, original_index</p></li>
<li><p>gene_metadata: Data frame with columns: gene, pc, loading for all analyzed genes</p></li>
<li><p>percent_var: Named numeric vector of percent variance explained for each analyzed PC</p></li>
</ul><p>Each PC data frame contains columns:</p><ul><li><p>gene: Gene symbol</p></li>
<li><p>loading: PC loading value for the gene</p></li>
<li><p>cell_type: Cell type analyzed</p></li>
<li><p>p_value: Raw p-value from Wilcoxon rank-sum test</p></li>
<li><p>p_adjusted: Adjusted p-value</p></li>
<li><p>mean_query: Mean expression in query data</p></li>
<li><p>mean_reference: Mean expression in reference data</p></li>
<li><p>significant: Logical indicating statistical significance</p></li>
</ul><p>A ggplot object showing boxplots of gene expression by PC and dataset.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function extracts the top loading genes for each specified principal component from the reference
PCA space and performs distributional comparisons between query and reference data. For each gene,
it performs statistical tests to identify genes that may be causing PC-specific alignment issues
between datasets.</p>
<p>This function visualizes the results from <code>calculateTopLoadingGeneShifts</code> by creating
faceted boxplots. Each facet represents a principal component with its variance explained.
Within each facet, boxplots show expression distributions for selected genes, comparing
query and reference datasets for a specific cell type. Genes can be selected either by
highest absolute loadings or most significant p-adjusted values.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>plot.calculateTopLoadingGeneShiftsObject</code></p>
<p><code>calculateTopLoadingGeneShifts</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Anthony Christidis, <a href="mailto:anthony-alexander_christidis@hms.harvard.edu">anthony-alexander_christidis@hms.harvard.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Load data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Compute distributional shifts for genes with top loadings</span></span></span>
<span class="r-in"><span><span class="va">gene_shifts</span> <span class="op">&lt;-</span> <span class="fu">calculateTopLoadingGeneShifts</span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span></span>
<span class="r-in"><span>                                             query_data <span class="op">=</span> <span class="va">query_data</span>,</span></span>
<span class="r-in"><span>                                             query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span></span>
<span class="r-in"><span>                                             ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span></span>
<span class="r-in"><span>                                             cell_types <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                             pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,</span></span>
<span class="r-in"><span>                                             n_top_loadings <span class="op">=</span> <span class="fl">50</span>,</span></span>
<span class="r-in"><span>                                             assay_name <span class="op">=</span> <span class="st">"logcounts"</span>,</span></span>
<span class="r-in"><span>                                             p_value_threshold <span class="op">=</span> <span class="fl">0.05</span>,</span></span>
<span class="r-in"><span>                                             adjust_method <span class="op">=</span> <span class="st">"fdr"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Plot gene shifts</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gene_shifts</span>,</span></span>
<span class="r-in"><span>     cell_type <span class="op">=</span> <span class="st">"CD8"</span>,</span></span>
<span class="r-in"><span>     pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,</span></span>
<span class="r-in"><span>     plot_by <span class="op">=</span> <span class="st">"p_adjusted"</span>,</span></span>
<span class="r-in"><span>     n_genes <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>     significance_threshold <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calculateTopLoadingGeneShifts-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

