<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Compare Principal Components Analysis (PCA) Results — comparePCA • scDiagnostics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Compare Principal Components Analysis (PCA) Results — comparePCA"><meta property="og:description" content="This function compares the principal components (PCs) obtained from separate PCA on reference and query
datasets for a single cell type using either cosine similarity or correlation.
The S3 plot method generates a heatmap to visualize the similarities between
principal components from the output of the comparePCA function."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection and Analysis of Annotation Anomalies</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compare Principal Components Analysis (PCA) Results</h1>
    <small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/R/comparePCA.R" class="external-link"><code>R/comparePCA.R</code></a>, <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/R/plot.comparePCAObject.R" class="external-link"><code>R/plot.comparePCAObject.R</code></a></small>
    <div class="hidden name"><code>comparePCA.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function compares the principal components (PCs) obtained from separate PCA on reference and query
datasets for a single cell type using either cosine similarity or correlation.</p>
<p>The S3 plot method generates a heatmap to visualize the similarities between
principal components from the output of the <code>comparePCA</code> function.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">comparePCA</span><span class="op">(</span></span>
<span>  <span class="va">query_data</span>,</span>
<span>  <span class="va">reference_data</span>,</span>
<span>  <span class="va">query_cell_type_col</span>,</span>
<span>  <span class="va">ref_cell_type_col</span>,</span>
<span>  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>  n_top_vars <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cosine"</span>, <span class="st">"correlation"</span><span class="op">)</span>,</span>
<span>  correlation_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"spearman"</span>, <span class="st">"pearson"</span><span class="op">)</span>,</span>
<span>  n_permutations <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'comparePCAObject'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  show_values <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_significance <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  significance_threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  color_limits <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-query-data">query_data<a class="anchor" aria-label="anchor" href="#arg-query-data"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code> object containing numeric expression matrix for the query cells.</p></dd>


<dt id="arg-reference-data">reference_data<a class="anchor" aria-label="anchor" href="#arg-reference-data"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code> object containing numeric expression matrix for the reference cells.</p></dd>


<dt id="arg-query-cell-type-col">query_cell_type_col<a class="anchor" aria-label="anchor" href="#arg-query-cell-type-col"></a></dt>
<dd><p>The column name in the <code>colData</code> of <code>query_data</code> that identifies the cell types.</p></dd>


<dt id="arg-ref-cell-type-col">ref_cell_type_col<a class="anchor" aria-label="anchor" href="#arg-ref-cell-type-col"></a></dt>
<dd><p>The column name in the <code>colData</code> of <code>reference_data</code> that identifies the cell types.</p></dd>


<dt id="arg-pc-subset">pc_subset<a class="anchor" aria-label="anchor" href="#arg-pc-subset"></a></dt>
<dd><p>A numeric vector specifying the subset of principal components (PCs) to compare. Default is the first five PCs.</p></dd>


<dt id="arg-n-top-vars">n_top_vars<a class="anchor" aria-label="anchor" href="#arg-n-top-vars"></a></dt>
<dd><p>An integer indicating the number of top loading variables to consider for each PC. Default is 50.</p></dd>


<dt id="arg-metric">metric<a class="anchor" aria-label="anchor" href="#arg-metric"></a></dt>
<dd><p>The similarity metric to use. It can be either "cosine" or "correlation". Default is "cosine".</p></dd>


<dt id="arg-correlation-method">correlation_method<a class="anchor" aria-label="anchor" href="#arg-correlation-method"></a></dt>
<dd><p>The correlation method to use if metric is "correlation". It can be "spearman"
or "pearson". Default is "spearman".</p></dd>


<dt id="arg-n-permutations">n_permutations<a class="anchor" aria-label="anchor" href="#arg-n-permutations"></a></dt>
<dd><p>Number of permutations for statistical significance testing. If 0, no permutation test is performed. Default is 0.</p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A <code>comparePCAObject</code> output from the <code>comparePCA</code> function.</p></dd>


<dt id="arg-show-values">show_values<a class="anchor" aria-label="anchor" href="#arg-show-values"></a></dt>
<dd><p>Logical, whether to display similarity values on the heatmap. Default is TRUE.</p></dd>


<dt id="arg-show-significance">show_significance<a class="anchor" aria-label="anchor" href="#arg-show-significance"></a></dt>
<dd><p>Logical, whether to display significance indicators (requires permutation test). Default is TRUE.</p></dd>


<dt id="arg-significance-threshold">significance_threshold<a class="anchor" aria-label="anchor" href="#arg-significance-threshold"></a></dt>
<dd><p>Numeric, p-value threshold for significance. Default is 0.05.</p></dd>


<dt id="arg-color-limits">color_limits<a class="anchor" aria-label="anchor" href="#arg-color-limits"></a></dt>
<dd><p>Numeric vector of length 2 specifying color scale limits. If NULL, uses data range.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed to the plotting function.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list containing:</p>
<dl><dt>similarity_matrix</dt>
<dd><p>A matrix comparing the principal components of the reference and query datasets.</p></dd>

  <dt>top_variables</dt>
<dd><p>A list containing the top loading variables for each PC pair comparison.</p></dd>

  <dt>p_values</dt>
<dd><p>A matrix of permutation p-values (if n_permutations &gt; 0).</p></dd>

  <dt>metric</dt>
<dd><p>The similarity metric used.</p></dd>

  <dt>n_top_vars</dt>
<dd><p>Number of top variables used.</p></dd>


</dl><p>A <code>ggplot</code> object representing the heatmap of similarities.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function compares the PCA results between the reference and query datasets by computing cosine
similarities or correlations between the loadings of top variables for each pair of principal components. It first
extracts the PCA rotation matrices from both datasets and identifies the top variables with highest loadings for
each PC. Then, it computes the cosine similarities or correlations between the loadings of top variables for each
pair of PCs using vectorized operations for improved performance. The resulting matrix contains the similarity values,
where rows represent reference PCs and columns represent query PCs.</p>
<p>The S3 plot method creates an enhanced heatmap visualization with options to display
statistical significance and similarity values. The heatmap uses a blue-white-red
color gradient for similarity values, and optionally overlays significance indicators.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>plot.comparePCAObject</code></p>
<p><code>comparePCA</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Anthony Christidis, <a href="mailto:anthony-alexander_christidis@hms.harvard.edu">anthony-alexander_christidis@hms.harvard.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Load libraries</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Load data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Extract CD4 cells</span></span></span>
<span class="r-in"><span><span class="va">ref_data_subset</span> <span class="op">&lt;-</span> <span class="va">reference_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">reference_data</span><span class="op">$</span><span class="va">expert_annotation</span> <span class="op">==</span> <span class="st">"CD4"</span><span class="op">)</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">query_data_subset</span> <span class="op">&lt;-</span> <span class="va">query_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">$</span><span class="va">expert_annotation</span> <span class="op">==</span> <span class="st">"CD4"</span><span class="op">)</span><span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Selecting highly variable genes (can be customized by the user)</span></span></span>
<span class="r-in"><span><span class="va">ref_top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">ref_data_subset</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">query_top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">query_data_subset</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Intersect the gene symbols to obtain common genes</span></span></span>
<span class="r-in"><span><span class="va">common_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">ref_top_genes</span>, <span class="va">query_top_genes</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ref_data_subset</span> <span class="op">&lt;-</span> <span class="va">ref_data_subset</span><span class="op">[</span><span class="va">common_genes</span>,<span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">query_data_subset</span> <span class="op">&lt;-</span> <span class="va">query_data_subset</span><span class="op">[</span><span class="va">common_genes</span>,<span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run PCA on datasets separately</span></span></span>
<span class="r-in"><span><span class="va">ref_data_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">ref_data_subset</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">query_data_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">query_data_subset</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Call the PCA comparison function</span></span></span>
<span class="r-in"><span><span class="va">similarity_mat</span> <span class="op">&lt;-</span> <span class="fu">comparePCA</span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_subset</span>,</span></span>
<span class="r-in"><span>                             reference_data <span class="op">=</span> <span class="va">ref_data_subset</span>,</span></span>
<span class="r-in"><span>                             query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span></span>
<span class="r-in"><span>                             ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span></span>
<span class="r-in"><span>                             pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span></span>
<span class="r-in"><span>                             n_top_vars <span class="op">=</span> <span class="fl">50</span>,</span></span>
<span class="r-in"><span>                             metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cosine"</span>, <span class="st">"correlation"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                             correlation_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"spearman"</span>, <span class="st">"pearson"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                             n_permutation <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing 100 permutations for significance testing...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create the heatmap</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">similarity_mat</span>, show_significance <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="comparePCA-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

