<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Compare Marker Gene Expression between Query and Reference Data — compareMarkers • scDiagnostics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Compare Marker Gene Expression between Query and Reference Data — compareMarkers"><meta property="og:description" content="This function identifies marker genes for each cell type in both query and reference datasets
using the standard Bioconductor approach (Wilcoxon rank-sum test), and compares their expression
patterns to assess annotation quality. It can optionally filter query cells based on anomaly detection
results and restrict analysis to specific cell types.
The S3 plot method generates a comprehensive visualization of the output from the `compareMarkers` function.
The plot shows marker gene overlap and expression consistency between query and reference cell types,
with quality assessment and detailed annotations."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection and Analysis of Annotation Anomalies</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compare Marker Gene Expression between Query and Reference Data</h1>
    <small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/R/compareMarkers.R" class="external-link"><code>R/compareMarkers.R</code></a>, <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/R/plot.compareMarkersObject.R" class="external-link"><code>R/plot.compareMarkersObject.R</code></a></small>
    <div class="hidden name"><code>compareMarkers.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function identifies marker genes for each cell type in both query and reference datasets
using the standard Bioconductor approach (Wilcoxon rank-sum test), and compares their expression
patterns to assess annotation quality. It can optionally filter query cells based on anomaly detection
results and restrict analysis to specific cell types.</p>
<p>The S3 plot method generates a comprehensive visualization of the output from the `compareMarkers` function.
The plot shows marker gene overlap and expression consistency between query and reference cell types,
with quality assessment and detailed annotations.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">compareMarkers</span><span class="op">(</span></span>
<span>  <span class="va">query_data</span>,</span>
<span>  <span class="va">reference_data</span>,</span>
<span>  <span class="va">query_cell_type_col</span>,</span>
<span>  <span class="va">ref_cell_type_col</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  n_markers <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  min_cells <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  anomaly_filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"none"</span>, <span class="st">"anomalous_only"</span>, <span class="st">"non_anomalous_only"</span><span class="op">)</span>,</span>
<span>  cell_types <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'compareMarkersObject'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">x</span>, cell_types <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-query-data">query_data<a class="anchor" aria-label="anchor" href="#arg-query-data"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code> object containing query cells.</p></dd>


<dt id="arg-reference-data">reference_data<a class="anchor" aria-label="anchor" href="#arg-reference-data"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code> object containing reference cells.</p></dd>


<dt id="arg-query-cell-type-col">query_cell_type_col<a class="anchor" aria-label="anchor" href="#arg-query-cell-type-col"></a></dt>
<dd><p>The column name in the <code>colData</code> of <code>query_data</code> that identifies the cell types.</p></dd>


<dt id="arg-ref-cell-type-col">ref_cell_type_col<a class="anchor" aria-label="anchor" href="#arg-ref-cell-type-col"></a></dt>
<dd><p>The column name in the <code>colData</code> of <code>reference_data</code> that identifies the cell types.</p></dd>


<dt id="arg-assay-name">assay_name<a class="anchor" aria-label="anchor" href="#arg-assay-name"></a></dt>
<dd><p>Name of the assay to use for computations. Default is "logcounts".</p></dd>


<dt id="arg-n-markers">n_markers<a class="anchor" aria-label="anchor" href="#arg-n-markers"></a></dt>
<dd><p>Number of top marker genes to consider for each cell type. Default is 50.</p></dd>


<dt id="arg-min-cells">min_cells<a class="anchor" aria-label="anchor" href="#arg-min-cells"></a></dt>
<dd><p>Minimum number of cells required per cell type for marker identification. Default is 10.</p></dd>


<dt id="arg-anomaly-filter">anomaly_filter<a class="anchor" aria-label="anchor" href="#arg-anomaly-filter"></a></dt>
<dd><p>Character string specifying how to filter query cells based on anomaly detection.
Options: "none" (default), "anomalous_only", "non_anomalous_only".</p></dd>


<dt id="arg-cell-types">cell_types<a class="anchor" aria-label="anchor" href="#arg-cell-types"></a></dt>
<dd><p>Character vector specifying which cell types to plot. If NULL, all cell types are plotted.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed to the plotting function.</p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A list containing the output from the <code>compareMarkers</code> function.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list containing the following elements:</p><dl><dt>marker_overlap</dt>
<dd><p>Matrix showing overlap of top markers between query and reference for each cell type.</p></dd>

  <dt>expression_consistency</dt>
<dd><p>Matrix showing expression consistency of reference markers in query data.</p></dd>

  <dt>quality_scores</dt>
<dd><p>Named vector of quality assessments for each cell type.</p></dd>

  <dt>markers_query</dt>
<dd><p>List of marker gene results for each cell type in query data.</p></dd>

  <dt>markers_ref</dt>
<dd><p>List of marker gene results for each cell type in reference data.</p></dd>

  <dt>common_cell_types</dt>
<dd><p>Vector of cell types present in both datasets.</p></dd>

  <dt>n_cells_query</dt>
<dd><p>Named vector of cell counts per type in query data.</p></dd>

  <dt>n_cells_ref</dt>
<dd><p>Named vector of cell counts per type in reference data.</p></dd>

  <dt>anomaly_filter_used</dt>
<dd><p>Character string indicating the anomaly filter applied.</p></dd>

  <dt>selected_cell_types</dt>
<dd><p>Character vector of cell types analyzed.</p></dd>

  <dt>anomaly_output</dt>
<dd><p>Output from anomaly detection if performed.</p></dd>


</dl><p>The S3 plot method returns a <code>ggplot</code> object representing the marker gene comparison results.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The function performs the following steps:
1. Optionally performs anomaly detection and filters query cells based on results.
2. Identifies marker genes for each cell type in both datasets using <code>findMarkers</code> approach.
3. Reference markers are always computed using all reference cells for each cell type.
4. Query markers are computed using filtered cells (anomalous/non-anomalous) if specified.
5. Compares the overlap of top marker genes between corresponding cell types.
6. Evaluates the expression consistency of reference markers in query data.
7. Provides quality scores based on marker gene concordance.</p>
<p>Marker genes are identified using Wilcoxon rank-sum tests comparing each cell type against all others.
High overlap and consistent expression of markers indicate good annotation quality.</p>
<p>The S3 plot method creates a scatter plot showing the relationship between marker overlap (x-axis)
and expression consistency (y-axis) for each cell type. Points are colored by quality score and
sized by the minimum number of cells. Quality zones provide visual guidance for interpretation.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>plot.compareMarkersObject</code>, <code><a href="detectAnomaly.html">detectAnomaly</a></code></p>
<p><code>compareMarkers</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Anthony Christidis, <a href="mailto:anthony-alexander_christidis@hms.harvard.edu">anthony-alexander_christidis@hms.harvard.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Load data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Compare marker genes</span></span></span>
<span class="r-in"><span><span class="va">marker_comparison</span> <span class="op">&lt;-</span> <span class="fu">compareMarkers</span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>,</span></span>
<span class="r-in"><span>                                    reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span></span>
<span class="r-in"><span>                                    query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span></span>
<span class="r-in"><span>                                    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># With anomaly filtering</span></span></span>
<span class="r-in"><span><span class="va">marker_comparison_filtered</span> <span class="op">&lt;-</span> <span class="fu">compareMarkers</span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>,</span></span>
<span class="r-in"><span>                                            reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span></span>
<span class="r-in"><span>                                            query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span></span>
<span class="r-in"><span>                                            ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span></span>
<span class="r-in"><span>                                            anomaly_filter <span class="op">=</span> <span class="st">"non_anomalous_only"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Visualize results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">marker_comparison</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="compareMarkers-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

