<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Package index â€¢ scDiagnostics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Package index"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection and Analysis of Annotation Anomalies</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="visualization-of-cell-type-annotations">Visualization of Cell Type Annotations <a href="#visualization-of-cell-type-annotations" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="boxplotPCA.html">boxplotPCA()</a></code> </p>
        </td>
        <td><p>Plot Principal Components for Different Cell Types</p></td>
      </tr><tr><td>
          <p><code><a href="calculateDiscriminantSpace.html">calculateDiscriminantSpace()</a></code> <code><a href="calculateDiscriminantSpace.html">plot(<i>&lt;calculateDiscriminantSpaceObject&gt;</i>)</a></code> </p>
        </td>
        <td><p>Project Query Data onto a Unified Discriminant Space of Reference Data</p></td>
      </tr><tr><td>
          <p><code><a href="calculateSIRSpace.html">calculateSIRSpace()</a></code> <code><a href="calculateSIRSpace.html">plot(<i>&lt;calculateSIRSpaceObject&gt;</i>)</a></code> </p>
        </td>
        <td><p>Calculate Sliced Inverse Regression (SIR) Space for Different Cell Types</p></td>
      </tr><tr><td>
          <p><code><a href="plotCellTypeMDS.html">plotCellTypeMDS()</a></code> </p>
        </td>
        <td><p>Plot Reference and Query Cell Types using MDS</p></td>
      </tr><tr><td>
          <p><code><a href="plotCellTypePCA.html">plotCellTypePCA()</a></code> </p>
        </td>
        <td><p>Plot Principal Components for Different Cell Types</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="evaluation-of-dataset-alignment">Evaluation of Dataset Alignment <a href="#evaluation-of-dataset-alignment" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="calculateGraphIntegration.html">calculateGraphIntegration()</a></code> <code><a href="calculateGraphIntegration.html">plot(<i>&lt;calculateGraphIntegrationObject&gt;</i>)</a></code> </p>
        </td>
        <td><p>Calculate Graph Community Integration Diagnostics</p></td>
      </tr><tr><td>
          <p><code><a href="calculateWassersteinDistance.html">calculateWassersteinDistance()</a></code> <code><a href="calculateWassersteinDistance.html">plot(<i>&lt;calculateWassersteinDistanceObject&gt;</i>)</a></code> </p>
        </td>
        <td><p>Compute Wasserstein Distance Distributions Between Query and Reference Datasets</p></td>
      </tr><tr><td>
          <p><code><a href="comparePCA.html">comparePCA()</a></code> <code><a href="comparePCA.html">plot(<i>&lt;comparePCAObject&gt;</i>)</a></code> </p>
        </td>
        <td><p>Compare Principal Components Analysis (PCA) Results</p></td>
      </tr><tr><td>
          <p><code><a href="comparePCASubspace.html">comparePCASubspace()</a></code> <code><a href="comparePCASubspace.html">plot(<i>&lt;comparePCASubspaceObject&gt;</i>)</a></code> </p>
        </td>
        <td><p>Compare Subspaces Spanned by Top Principal Components</p></td>
      </tr><tr><td>
          <p><code><a href="plotPairwiseDistancesDensity.html">plotPairwiseDistancesDensity()</a></code> </p>
        </td>
        <td><p>Ridgeline Plot of Pairwise Distance Analysis</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="calculation-of-statistical-measures-to-compare-two-datasets">Calculation of Statistical Measures to Compare Two Datasets <a href="#calculation-of-statistical-measures-to-compare-two-datasets" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation()</a></code> <code><a href="calculateAveragePairwiseCorrelation.html">plot(<i>&lt;calculateAveragePairwiseCorrelationObject&gt;</i>)</a></code> </p>
        </td>
        <td><p>Compute Average Pairwise Correlation between Cell Types</p></td>
      </tr><tr><td>
          <p><code><a href="calculateCramerPValue.html">calculateCramerPValue()</a></code> </p>
        </td>
        <td><p>Calculate Cramer Test P-Values for Two-Sample Comparison of Multivariate ECDFs</p></td>
      </tr><tr><td>
          <p><code><a href="calculateHotellingPValue.html">calculateHotellingPValue()</a></code> </p>
        </td>
        <td><p>Perform Hotelling's T-squared Test on PCA Scores for Single-cell RNA-seq Data</p></td>
      </tr><tr><td>
          <p><code><a href="calculateMMDPValue.html">calculateMMDPValue()</a></code> </p>
        </td>
        <td><p>Calculate Maximum Mean Discrepancy P-Values for Two-Sample Comparison</p></td>
      </tr><tr><td>
          <p><code><a href="regressPC.html">plot(<i>&lt;regressPCObject&gt;</i>)</a></code> <code><a href="regressPC.html">regressPC()</a></code> </p>
        </td>
        <td><p>Plot Regression Results on Principal Components</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="anomaly-detection-global-and-cell-type-specific-">Anomaly Detection (Global and Cell Type-Specific) <a href="#anomaly-detection-global-and-cell-type-specific-" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="detectAnomaly.html">detectAnomaly()</a></code> <code><a href="detectAnomaly.html">plot(<i>&lt;detectAnomalyObject&gt;</i>)</a></code> </p>
        </td>
        <td><p>PCA Anomaly Scores via Isolation Forests with Visualization</p></td>
      </tr><tr><td>
          <p><code><a href="calculateCellSimilarityPCA.html">calculateCellSimilarityPCA()</a></code> <code><a href="calculateCellSimilarityPCA.html">plot(<i>&lt;calculateCellSimilarityPCAObject&gt;</i>)</a></code> </p>
        </td>
        <td><p>Calculate Cell Similarity Using PCA Loadings</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="calculation-of-distances-between-specific-cells-and-cell-populations">Calculation of Distances Between Specific Cells and Cell Populations <a href="#calculation-of-distances-between-specific-cells-and-cell-populations" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="calculateCellDistances.html">calculateCellDistances()</a></code> <code><a href="calculateCellDistances.html">plot(<i>&lt;calculateCellDistancesObject&gt;</i>)</a></code> </p>
        </td>
        <td><p>Compute Cell Distances Between Reference and Query Data</p></td>
      </tr><tr><td>
          <p><code><a href="calculateCellDistancesSimilarity.html">calculateCellDistancesSimilarity()</a></code> </p>
        </td>
        <td><p>Function to Calculate Bhattacharyya Coefficients and Hellinger Distances</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="evaluation-of-marker-gene-alignment">Evaluation of Marker Gene Alignment <a href="#evaluation-of-marker-gene-alignment" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="calculateHVGOverlap.html">calculateHVGOverlap()</a></code> </p>
        </td>
        <td><p>Calculate the Overlap Coefficient for Highly Variable Genes</p></td>
      </tr><tr><td>
          <p><code><a href="calculateVarImpOverlap.html">calculateVarImpOverlap()</a></code> </p>
        </td>
        <td><p>Compare Gene Importance Across Datasets Using Random Forest</p></td>
      </tr><tr><td>
          <p><code><a href="compareMarkers.html">compareMarkers()</a></code> <code><a href="compareMarkers.html">plot(<i>&lt;compareMarkersObject&gt;</i>)</a></code> </p>
        </td>
        <td><p>Compare Marker Gene Expression between Query and Reference Data</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualization-of-marker-expressions">Visualization of Marker Expressions <a href="#visualization-of-marker-expressions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotGeneExpressionDimred.html">plotGeneExpressionDimred()</a></code> </p>
        </td>
        <td><p>Visualize gene expression on a dimensional reduction plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotMarkerExpression.html">plotMarkerExpression()</a></code> </p>
        </td>
        <td><p>Plot gene expression distribution from overall and cell type-specific perspective</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualization-of-qc-and-annotation-scores">Visualization of QC and Annotation Scores <a href="#visualization-of-qc-and-annotation-scores" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="histQCvsAnnotation.html">histQCvsAnnotation()</a></code> </p>
        </td>
        <td><p>Histograms: QC Stats and Annotation Scores Visualization</p></td>
      </tr><tr><td>
          <p><code><a href="plotQCvsAnnotation.html">plotQCvsAnnotation()</a></code> </p>
        </td>
        <td><p>Scatter plot: QC stats vs Cell Type Annotation Scores</p></td>
      </tr><tr><td>
          <p><code><a href="plotGeneSetScores.html">plotGeneSetScores()</a></code> </p>
        </td>
        <td><p>Visualization of gene sets or pathway scores on dimensional reduction plot</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="misc">Misc <a href="#misc" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="processPCA.html">processPCA()</a></code> </p>
        </td>
        <td><p>Process PCA for SingleCellExperiment Objects</p></td>
      </tr><tr><td>
          <p><code><a href="projectPCA.html">projectPCA()</a></code> </p>
        </td>
        <td><p>Project Query Data Onto PCA Space of Reference Data</p></td>
      </tr><tr><td>
          <p><code><a href="projectSIR.html">projectSIR()</a></code> </p>
        </td>
        <td><p>Project Query Data Onto SIR Space of Reference Data</p></td>
      </tr><tr><td>
          <p><code><a href="calculateCategorizationEntropy.html">calculateCategorizationEntropy()</a></code> </p>
        </td>
        <td><p>Calculate Categorization Entropy</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="datasets">Datasets <a href="#datasets" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="reference_data.html">reference_data</a></code> </p>
        </td>
        <td><p>Reference Single-Cell RNA-Seq Dataset</p></td>
      </tr><tr><td>
          <p><code><a href="query_data.html">query_data</a></code> </p>
        </td>
        <td><p>Query Single-Cell RNA-Seq Dataset</p></td>
      </tr><tr><td>
          <p><code><a href="qc_data.html">qc_data</a></code> </p>
        </td>
        <td><p>Quality Control Single-Cell RNA-Seq Dataset</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

