<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot Regression Results on Principal Components — plot.regressPCObject • scDiagnostics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Plot Regression Results on Principal Components — plot.regressPCObject"><meta property="og:description" content="The S3 plot method generates plots to visualize the results of regression analyses
performed on principal components (PCs) against cell types, datasets, or their interactions.
This function performs linear regression of a covariate of interest onto one
or more principal components, based on the data in a SingleCellExperiment
object."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection and Analysis of Annotation Anomalies</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot Regression Results on Principal Components</h1>
    <small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/R/plot.regressPCObject.R" class="external-link"><code>R/plot.regressPCObject.R</code></a>, <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/R/regressPC.R" class="external-link"><code>R/regressPC.R</code></a></small>
    <div class="hidden name"><code>regressPC.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The S3 plot method generates plots to visualize the results of regression analyses
performed on principal components (PCs) against cell types, datasets, or their interactions.</p>
<p>This function performs linear regression of a covariate of interest onto one
or more principal components, based on the data in a <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code>
object.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S3 method for class 'regressPCObject'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  plot_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"r_squared"</span>, <span class="st">"variance_contribution"</span>, <span class="st">"coefficient_heatmap"</span><span class="op">)</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  coefficients_include <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">regressPC</span><span class="op">(</span></span>
<span>  <span class="va">query_data</span>,</span>
<span>  reference_data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">query_cell_type_col</span>,</span>
<span>  ref_cell_type_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  query_batch_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cell_types <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>  adjust_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BH"</span>, <span class="st">"holm"</span>, <span class="st">"hochberg"</span>, <span class="st">"hommel"</span>, <span class="st">"bonferroni"</span>, <span class="st">"BY"</span>, <span class="st">"fdr"</span>,</span>
<span>    <span class="st">"none"</span><span class="op">)</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  max_cells <span class="op">=</span> <span class="fl">2500</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>An object of class <code>regressPCObject</code> containing the output of the <code>regressPC</code> function</p></dd>


<dt id="arg-plot-type">plot_type<a class="anchor" aria-label="anchor" href="#arg-plot-type"></a></dt>
<dd><p>Type of plot to generate. Available options:
"r_squared", "variance_contribution", "coefficient_heatmap"</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>Significance threshold for p-values. Default is 0.05.</p></dd>


<dt id="arg-coefficients-include">coefficients_include<a class="anchor" aria-label="anchor" href="#arg-coefficients-include"></a></dt>
<dd><p>Character vector specifying which coefficient types to include
in the coefficient heatmap. Options are <code>c("cell_type", "batch", "interaction")</code>.
Default is <code>NULL</code>, which includes all available coefficient types. Only applies
to <code>plot_type = "coefficient_heatmap"</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments to be passed to the plotting functions.</p></dd>


<dt id="arg-query-data">query_data<a class="anchor" aria-label="anchor" href="#arg-query-data"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code> object containing numeric expression matrix for the query cells.</p></dd>


<dt id="arg-reference-data">reference_data<a class="anchor" aria-label="anchor" href="#arg-reference-data"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code> object containing numeric expression matrix for the reference cells.
If NULL, the PC scores are regressed against the cell types of the query data.</p></dd>


<dt id="arg-query-cell-type-col">query_cell_type_col<a class="anchor" aria-label="anchor" href="#arg-query-cell-type-col"></a></dt>
<dd><p>The column name in the <code>colData</code> of <code>query_data</code> that identifies the cell types.</p></dd>


<dt id="arg-ref-cell-type-col">ref_cell_type_col<a class="anchor" aria-label="anchor" href="#arg-ref-cell-type-col"></a></dt>
<dd><p>The column name in the <code>colData</code> of <code>reference_data</code> that identifies the cell types.</p></dd>


<dt id="arg-query-batch-col">query_batch_col<a class="anchor" aria-label="anchor" href="#arg-query-batch-col"></a></dt>
<dd><p>The column name in the <code>colData</code> of <code>query_data</code> that identifies the batch or sample.
If provided, performs interaction analysis with cell types. Default is NULL.</p></dd>


<dt id="arg-cell-types">cell_types<a class="anchor" aria-label="anchor" href="#arg-cell-types"></a></dt>
<dd><p>A character vector specifying the cell types to include in the analysis. If NULL, all cell types are included.</p></dd>


<dt id="arg-pc-subset">pc_subset<a class="anchor" aria-label="anchor" href="#arg-pc-subset"></a></dt>
<dd><p>A numeric vector specifying which principal components to include in the analysis. Default is PC1 to PC10.</p></dd>


<dt id="arg-adjust-method">adjust_method<a class="anchor" aria-label="anchor" href="#arg-adjust-method"></a></dt>
<dd><p>A character string specifying the method to adjust the p-values.
Options include "BH", "holm", "hochberg", "hommel", "bonferroni", "BY", "fdr", or "none".
Default is "BH" (Benjamini-Hochberg).</p></dd>


<dt id="arg-assay-name">assay_name<a class="anchor" aria-label="anchor" href="#arg-assay-name"></a></dt>
<dd><p>Name of the assay on which to perform computations. Default is "logcounts".</p></dd>


<dt id="arg-max-cells">max_cells<a class="anchor" aria-label="anchor" href="#arg-max-cells"></a></dt>
<dd><p>Maximum number of cells to retain. If the object has fewer cells, it is returned unchanged.
Default is 2500.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>The S3 plot method returns a <code>ggplot</code> object representing the specified plot type.</p>
<p>A <code>list</code> containing</p><ul><li><p>summaries of the linear
regression models for each specified principal component,</p></li>
<li><p>the
corresponding R-squared (R2) values,</p></li>
<li><p>the variance contributions for
each principal component, and</p></li>
<li><p>the total variance explained.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>Principal component regression, derived from PCA, can be used to quantify the
variance explained by a covariate of interest. Applications for single-cell
analysis include quantification of batch effects, assessing clustering
homogeneity, and evaluating alignment of query and reference datasets in cell
type annotation settings.</p>
<p>The function supports multiple regression scenarios:</p><ul><li><p>Query only, no batch: PC  cell_type</p></li>
<li><p>Query only, with batch: PC  cell_type * batch</p></li>
<li><p>Query + Reference, no batch: PC  cell_type * dataset</p></li>
<li><p>Query + Reference, with batch: PC  cell_type * batch (where batch includes Reference)</p></li>
</ul><p>When batch information is provided with reference data, batches are labeled as
"Reference" for reference data and "Query_BatchName" for query batches, with
Reference set as the first factor level for interpretation.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Luecken et al. Benchmarking atlas-level data integration in
single-cell genomics. Nature Methods, 19:41-50, 2022.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>regressPC</code></p>
<p><code>plot.regressPCObject</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Anthony Christidis, <a href="mailto:anthony-alexander_christidis@hms.harvard.edu">anthony-alexander_christidis@hms.harvard.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Load data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Query only analysis</span></span></span>
<span class="r-in"><span><span class="va">regress_res</span> <span class="op">&lt;-</span> <span class="fu">regressPC</span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>,</span></span>
<span class="r-in"><span>                         query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span></span>
<span class="r-in"><span>                         cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                         pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Visualize results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"r_squared"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="regressPC-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"variance_contribution"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="regressPC-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"coefficient_heatmap"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="regressPC-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Query + Reference analysis</span></span></span>
<span class="r-in"><span><span class="va">regress_res</span> <span class="op">&lt;-</span> <span class="fu">regressPC</span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>,</span></span>
<span class="r-in"><span>                         reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span></span>
<span class="r-in"><span>                         query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span></span>
<span class="r-in"><span>                         ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span></span>
<span class="r-in"><span>                         cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                         pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Visualize results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"r_squared"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="regressPC-4.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"variance_contribution"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="regressPC-5.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"coefficient_heatmap"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="regressPC-6.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

