<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>scDiagnostics: Single-Cell Diagnostics Package — scDiagnostics-package • scDiagnostics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="scDiagnostics: Single-Cell Diagnostics Package — scDiagnostics-package"><meta property="og:description" content="`scDiagnostics` is a comprehensive toolkit designed for the analysis and diagnostics of single-cell RNA sequencing (scRNA-seq) data. This package provides functionalities for comparing principal components, visualizing canonical correlation analysis (CCA) outputs, and plotting cell type-specific MDS and PCA projections."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection and Analysis of Annotation Anomalies</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>scDiagnostics: Single-Cell Diagnostics Package</h1>
    <small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/R/scDiagnostics-package.R" class="external-link"><code>R/scDiagnostics-package.R</code></a></small>
    <div class="hidden name"><code>scDiagnostics-package.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>`scDiagnostics` is a comprehensive toolkit designed for the analysis and diagnostics of single-cell RNA sequencing (scRNA-seq) data. This package provides functionalities for comparing principal components, visualizing canonical correlation analysis (CCA) outputs, and plotting cell type-specific MDS and PCA projections.</p>
    </div>


    <div id="details">
    <h2>Details</h2>
    <p>The package includes the following key functionalities, organized by their specific purposes:</p>
    </div>
    <div id="visualization-of-cell-type-annotations">
    <h2>Visualization of Cell Type Annotations</h2>


<p>Functions for visualizing differences between query and reference datasets across multiple cell types.</p><ul><li><p><code><a href="boxplotPCA.html">boxplotPCA</a></code>: Boxplots of PCA scores for cell types.</p></li>
<li><p><code><a href="calculateDiscriminantSpace.html">calculateDiscriminantSpace</a></code>: Calculates discriminant space, with a plot method for visualization.</p></li>
<li><p><code><a href="plotCellTypeMDS.html">plotCellTypeMDS</a></code>: Creates MDS plots for cell types using query and reference datasets.</p></li>
<li><p><code><a href="plotCellTypePCA.html">plotCellTypePCA</a></code>: Plots principal components for different cell types.</p></li>
</ul></div>
    <div id="evaluation-of-dataset-alignment">
    <h2>Evaluation of Dataset Alignment</h2>


<p>Functions for visualizing differences between query and reference datasets for a specific cell type.</p><ul><li><p><code><a href="calculateGraphIntegration.html">calculateGraphIntegration</a></code>: Calculate a graph network and form communities of cells.</p></li>
<li><p><code><a href="calculateWassersteinDistance.html">calculateWassersteinDistance</a></code>: Wasserstein distance for different cell types.</p></li>
<li><p><code><a href="comparePCA.html">comparePCA</a></code>: Compares PCA results, with a plot method for visualization.</p></li>
<li><p><code><a href="comparePCASubspace.html">comparePCASubspace</a></code>: Compares PCA subspace, with a plot method for visualization.</p></li>
<li><p><code><a href="plotPairwiseDistancesDensity.html">plotPairwiseDistancesDensity</a></code>: Plots the density of pairwise distances.</p></li>
</ul></div>
    <div id="calculation-of-statistical-measures-to-compare-two-datasets">
    <h2>Calculation of Statistical Measures to Compare Two Datasets</h2>


<p>Functions to compute statistical measures to compare two datasets.</p><ul><li><p><code><a href="calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation</a></code>: Calculates average pairwise correlation, with a plot method for visualization.</p></li>
<li><p><code><a href="calculateCramerPValue.html">calculateCramerPValue</a></code>: Calculates the p-value using Cramer's test.</p></li>
<li><p><code><a href="calculateHotellingPValue.html">calculateHotellingPValue</a></code>: Calculates the p-value using Hotelling's T-squared test.</p></li>
<li><p><code><a href="calculateMMDPValue.html">calculateMMDPValue</a></code>: Calculates the p-value using maximum mean discrepancy.</p></li>
<li><p><code><a href="regressPC.html">regressPC</a></code>: Performs regression on principal components, with a plot method for visualization.</p></li>
</ul></div>
    <div id="evaluation-of-marker-gene-alignment">
    <h2>Evaluation of Marker Gene Alignment</h2>


<p>Functions for calculating overlap measures of genes between two datasets.</p><ul><li><p><code><a href="calculateHVGOverlap.html">calculateHVGOverlap</a></code>: Calculates overlap of highly variable genes (HVG) between datasets.</p></li>
<li><p><code><a href="calculateVarImpOverlap.html">calculateVarImpOverlap</a></code>: Calculates overlap of variable importance measures between datasets.</p></li>
<li><p><code><a href="compareMarkers.html">compareMarkers</a></code>: Compares marker genes between datasets.</p></li>
</ul></div>
    <div id="visualization-of-marker-expressions">
    <h2>Visualization of Marker Expressions</h2>


<p>Functions for to visualize and compare the expression of markers between a reference and a query dataset.</p><ul><li><p><code><a href="plotGeneExpressionDimred.html">plotGeneExpressionDimred</a></code>: Plots gene expression in a dimensionality reduction space.</p></li>
<li><p><code><a href="plotMarkerExpression.html">plotMarkerExpression</a></code>: Plots marker expression levels.</p></li>
</ul></div>
    <div id="anomaly-detection-global-and-cell-type-specific-">
    <h2>Anomaly Detection (Global and Cell Type-Specific)</h2>


<p>Functions for detecting anomalies at both the global and cell type-specific levels.</p><ul><li><p><code><a href="detectAnomaly.html">detectAnomaly</a></code>: Detects anomalies in the data, with a plot method for visualization.</p></li>
<li><p><code><a href="calculateCellSimilarityPCA.html">calculateCellSimilarityPCA</a></code>: Calculates cell similarity in PCA space, with a plot method for visualization.</p></li>
</ul></div>
    <div id="calculation-of-distances-between-specific-cells-and-cell-populations">
    <h2>Calculation of Distances Between Specific Cells and Cell Populations</h2>


<p>Functions for calculating distances between specific cells and cell populations.</p><ul><li><p><code><a href="calculateCellDistances.html">calculateCellDistances</a></code>: Calculates distances between cells, with a plot method for visualization.</p></li>
<li><p><code><a href="calculateCellDistancesSimilarity.html">calculateCellDistancesSimilarity</a></code>: Calculates similarity based on cell distances, with a plot method for visualization.</p></li>
</ul></div>
    <div id="visualization-of-qc-and-annotation-scores">
    <h2>Visualization of QC and Annotation Scores</h2>


<p>Functions for visualizing quality control (QC) metrics or other characteristics of the data.</p><ul><li><p><code><a href="histQCvsAnnotation.html">histQCvsAnnotation</a></code>: Plots histograms of QC metrics versus annotations.</p></li>
<li><p><code><a href="plotGeneSetScores.html">plotGeneSetScores</a></code>: Plots scores of gene sets.</p></li>
<li><p><code><a href="plotQCvsAnnotation.html">plotQCvsAnnotation</a></code>: Plots QC metrics versus annotations.</p></li>
</ul></div>
    <div id="misc">
    <h2>Misc</h2>


<p>Miscellaneous functions for various tasks.</p><ul><li><p><code><a href="projectPCA.html">projectPCA</a></code>: Projects new data into PCA space.</p></li>
<li><p><code><a href="projectSIR.html">projectSIR</a></code>: Projects new data into SIR space.</p></li>
<li><p><code><a href="calculateCategorizationEntropy.html">calculateCategorizationEntropy</a></code>: Calculates categorization entropy for clusters.</p></li>
</ul><p>The package is built to facilitate in-depth analysis and visualization of single-cell data, enhancing the understanding of cell type similarities and differences across datasets.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Useful links:</p><ul><li><p><a href="https://github.com/ccb-hms/scDiagnostics" class="external-link">https://github.com/ccb-hms/scDiagnostics</a></p></li>
<li><p>Report bugs at <a href="https://github.com/ccb-hms/scDiagnostics/issues" class="external-link">https://github.com/ccb-hms/scDiagnostics/issues</a></p></li>
</ul></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p><strong>Maintainer</strong>: Anthony Christidis <a href="mailto:anthony-alexander_christidis@hms.harvard.edu">anthony-alexander_christidis@hms.harvard.edu</a> (<a href="https://orcid.org/0000-0002-4565-6279" class="external-link">ORCID</a>)</p>
<p>Authors:</p><ul><li><p>Andrew Ghazi</p></li>
<li><p>Smriti Chawla</p></li>
<li><p>Ludwig Geistlinger</p></li>
<li><p>Robert Gentleman</p></li>
</ul><p>Other contributors:</p><ul><li><p>Nitesh Turaga [contributor]</p></li>
</ul></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

