<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Detection and Analysis of Annotation Anomalies • scDiagnostics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Detection and Analysis of Annotation Anomalies">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection and Analysis of Annotation Anomalies</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Detection and Analysis of Annotation
Anomalies</h1>
                        <h4 data-toc-skip class="author">Anthony
Christidis</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Andrew
Ghazi</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Smriti
Chawla</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Nitesh
Turaga</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Ludwig
Geistlinger</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Robert
Gentleman</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/vignettes/AnnotationAnomalies.Rmd" class="external-link"><code>vignettes/AnnotationAnomalies.Rmd</code></a></small>
      <div class="hidden name"><code>AnnotationAnomalies.Rmd</code></div>

    </address>
</address>
</address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <code>scDiagnostics</code> package provides powerful tools for
anomaly detection in single-cell data, enabling researchers to identify
and analyze outliers in complex datasets. Central to this process is the
<code><a href="../reference/detectAnomaly.html">detectAnomaly()</a></code> function, which integrates dimensionality
reduction through Principal Component Analysis (PCA) with the robust
capabilities of the isolation forest algorithm.</p>
<p>In single-cell analysis, detecting anomalies is crucial for
identifying potential data issues, such as mislabeled cells, technical
artifacts, or biologically distinct subpopulations. The
<code><a href="../reference/detectAnomaly.html">detectAnomaly()</a></code> function offers a versatile approach to
anomaly detection by allowing users to project data onto a PCA space and
apply isolation forests to uncover outliers. Whether working solely with
a reference dataset or comparing a query dataset against a
well-characterized reference, this function provides detailed insights
into potential anomalies.</p>
<p>This vignette illustrates how to effectively use the
<code><a href="../reference/detectAnomaly.html">detectAnomaly()</a></code> function in various scenarios. We explore
both cell-type-specific and global anomaly detection, demonstrate the
utility of integrating query data with reference data, and offer
guidance on interpreting the results. Additionally, we show how to
extend this analysis by combining anomaly detection with PCA loadings
using the <code><a href="../reference/calculateCellSimilarityPCA.html">calculateCellSimilarityPCA()</a></code> function, providing
a comprehensive toolkit for investigating the structure and quality of
single-cell data.</p>
<p>This vignette also demonstrates the use of two functions,
<code><a href="../reference/calculateCellDistances.html">calculateCellDistances()</a></code> and
<code><a href="../reference/calculateCellDistancesSimilarity.html">calculateCellDistancesSimilarity()</a></code>, to analyze the
distances between cells in query and reference datasets and to measure
the similarity of density distributions. These functions are useful for
investigating how cells in different datasets relate to each other,
particularly in the context of identifying anomalies and understanding
distribution overlaps.</p>
</div>
<div class="section level2">
<h2 id="preliminaries">Preliminaries<a class="anchor" aria-label="anchor" href="#preliminaries"></a>
</h2>
<p>In the context of the <code>scDiagnostics</code> package, the
following datasets illustrate the application of these tools:</p>
<ul>
<li><p><code>reference_data</code>: A curated and processed dataset
containing expert-assigned cell type annotations. This dataset serves as
a reference for comparison and can be used alone to detect anomalies
within the reference annotations.</p></li>
<li><p><code>query_data</code>: A dataset that also includes
expert-assigned cell type annotations, but additionally features
annotations generated by the <em><a href="https://bioconductor.org/packages/3.21/SingleR" class="external-link">SingleR</a></em>
package. This allows for the comparison of expert annotations with those
produced by an automated method to detect inconsistencies or
anomalies.</p></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/scDiagnostics" class="external-link">scDiagnostics</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>By using these datasets, you can leverage the package’s tools to
compare annotations between <code>reference_data</code> and
<code>query_data</code>, or analyze <code>reference_data</code> alone to
identify potential issues. The package’s flexibility supports various
analysis scenarios, whether you need to assess overall annotation
quality or focus on specific cell types.</p>
<p>Through these capabilities, <code>scDiagnostics</code> empowers you
to perform thorough and nuanced assessments of cell type annotations,
enhancing the accuracy and reliability of your analyses.</p>
</div>
<div class="section level2">
<h2 id="the-detectanomaly-function">The <code>detectAnomaly()</code> Function<a class="anchor" aria-label="anchor" href="#the-detectanomaly-function"></a>
</h2>
<div class="section level3">
<h3 id="function-overview">Function Overview<a class="anchor" aria-label="anchor" href="#function-overview"></a>
</h3>
<div class="section level4">
<h4 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h4>
<p>The <code><a href="../reference/detectAnomaly.html">detectAnomaly()</a></code> function integrates dimensionality
reduction via PCA with the isolation forest algorithm to detect
anomalies in single-cell data. By projecting both reference and query
datasets (if available) onto the PCA space of the reference data, the
function trains an isolation forest model on reference data in PCA space
to pinpoint anomalies in the reference or query data. This approach is
highly versatile:</p>
<ul>
<li>
<strong>Reference Only</strong>: Compute anomaly scores solely for
the reference dataset to identify potential issues within the reference
itself.</li>
<li>
<strong>Reference and Query</strong>: Compare the query dataset
against the reference to find anomalies in the query data that may not
align with the established reference.</li>
<li>
<strong>Global and Specific Analysis</strong>: Assess anomalies at a
global level or focus on specific cell types to gain targeted insights
into your data. The function also provides detailed visualizations and
statistical outputs to help you interpret the anomalies detected.</li>
</ul>
</div>
<div class="section level4">
<h4 id="parameters">Parameters<a class="anchor" aria-label="anchor" href="#parameters"></a>
</h4>
<p>The function takes a <em><a href="https://bioconductor.org/packages/3.21/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
object as <code>reference_data</code> and trains an isolation forest
model on the reference PCA-projected data, with an optional
<code>query_data</code> for projecting onto this PCA space for anomaly
detection. You can specify cell type annotations through
<code>ref_cell_type_col</code> and <code>query_cell_type_col</code>, and
limit the analysis to certain cell types using the
<code>cell_types</code> parameter. The function allows you to select
specific principal components to use to train the isolation forest via
<code>pc_subset</code>, adjust the number of trees with
<code>n_tree</code>, and set an <code>anomaly_threshold</code> for
classifying anomalies.</p>
</div>
<div class="section level4">
<h4 id="return-value">Return Value<a class="anchor" aria-label="anchor" href="#return-value"></a>
</h4>
<p>The function returns several outputs: <code>anomaly_scores</code>
indicating the likelihood of each cell being an anomaly, a logical
vector (<code>anomaly</code>) identifying these anomalies, PCA
projections for the reference data (<code>reference_mat_subset</code>)
and optionally for the query data (<code>query_mat_subset</code>), and
the proportion of variance explained by the selected principal
components (<code>var_explained</code>).</p>
</div>
</div>
<div class="section level3">
<h3 id="detectanomaly-examples">
<code>detectAnomaly()</code> Examples<a class="anchor" aria-label="anchor" href="#detectanomaly-examples"></a>
</h3>
<div class="section level4">
<h4 id="anomaly-detection-with-reference-and-query-data">Anomaly Detection with Reference and Query Data<a class="anchor" aria-label="anchor" href="#anomaly-detection-with-reference-and-query-data"></a>
</h4>
<p>This section demonstrates how to use the <code><a href="../reference/detectAnomaly.html">detectAnomaly()</a></code>
function when both reference and query datasets are provided. It
includes examples of analyzing anomalies for specific cell types and
globally across all data.</p>
<div class="section level5">
<h5 id="example-1-cell-type-specific-anomaly-detection">Example 1: Cell-Type Specific Anomaly Detection<a class="anchor" aria-label="anchor" href="#example-1-cell-type-specific-anomaly-detection"></a>
</h5>
<p>In this example, we analyze anomalies specifically for the “CD4” cell
type. The anomaly scores are trained on the PCA projections of the “CD4”
cells from the reference dataset. If query data is provided, anomaly
scores for the query data are predicted based on the PCA projections of
the query data onto the reference PCA space for the “CD4” cell type.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform anomaly detection</span></span>
<span><span class="va">anomaly_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectAnomaly.html">detectAnomaly</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>                                pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>                                n_tree <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                anomaly_threshold <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the output for the "CD4" cell type</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">anomaly_output</span>, </span>
<span>     cell_type <span class="op">=</span> <span class="st">"CD4"</span>, </span>
<span>     pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, </span>
<span>     data_type <span class="op">=</span> <span class="st">"query"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/AnnotationAnomalies/detectAnomaly1.png"></p>
<p>In this example, we analyze anomalies specifically for the “CD4” cell
type. The anomaly scores are trained on the PCA projections of the “CD4”
cells from the reference dataset. If query data is provided, anomaly
scores for the query data are predicted based on the PCA projections of
the query data onto the reference PCA space for the “CD4” cell type.</p>
</div>
<div class="section level5">
<h5 id="example-2-global-anomaly-detection">Example 2: Global Anomaly Detection<a class="anchor" aria-label="anchor" href="#example-2-global-anomaly-detection"></a>
</h5>
<p>Here, we perform global anomaly detection by setting
<code>cell_type = NULL</code>. In this case, the isolation forest is
trained on PCA projections of all cells in the reference data combined.
The global anomaly scores are then computed for both reference and query
datasets.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform anomaly detection</span></span>
<span><span class="va">anomaly_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectAnomaly.html">detectAnomaly</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>                                pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>                                n_tree <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                anomaly_threshold <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the global anomaly scores</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">anomaly_output</span>, </span>
<span>     cell_type <span class="op">=</span> <span class="cn">NULL</span>,  <span class="co"># Plot all cell types</span></span>
<span>     pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, </span>
<span>     data_type <span class="op">=</span> <span class="st">"query"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Warning in max(ids, na.rm = TRUE): no non-missing arguments to max; returning</span></span>
<span><span class="co">#&gt; -Inf</span></span>
<span><span class="co">#&gt; Warning: Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Warning in max(ids, na.rm = TRUE): no non-missing arguments to max; returning</span></span>
<span><span class="co">#&gt; -Inf</span></span>
<span><span class="co">#&gt; Warning: Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Warning in max(ids, na.rm = TRUE): no non-missing arguments to max; returning</span></span>
<span><span class="co">#&gt; -Inf</span></span>
<span><span class="co">#&gt; Warning: Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Warning in max(ids, na.rm = TRUE): no non-missing arguments to max; returning</span></span>
<span><span class="co">#&gt; -Inf</span></span>
<span><span class="co">#&gt; Warning: Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Warning in max(ids, na.rm = TRUE): no non-missing arguments to max; returning</span></span>
<span><span class="co">#&gt; -Inf</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/AnnotationAnomalies/detectAnomaly2.png"></p>
<p>Setting <code>cell_type = NULL</code> means that the anomaly
detection is done globally. The isolation forest is trained on PCA
projections of all cells from the reference dataset. Anomaly scores are
then predicted for the query data based on these global PCA projections.
The plot provides a comprehensive view of anomalies across all cell
types in the query dataset.</p>
</div>
<div class="section level5">
<h5 id="anomaly-detection-on-reference-data">Anomaly Detection on Reference Data<a class="anchor" aria-label="anchor" href="#anomaly-detection-on-reference-data"></a>
</h5>
<p>Note that if you do not provide a query dataset to the
<code><a href="../reference/detectAnomaly.html">detectAnomaly()</a></code> function, the function will only return the
anomaly output data on the reference data.</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="integrating-anomaly-detection-with-cell-similarity-analysis-using-pca-loadings">Integrating Anomaly Detection with Cell Similarity Analysis Using
PCA Loadings<a class="anchor" aria-label="anchor" href="#integrating-anomaly-detection-with-cell-similarity-analysis-using-pca-loadings"></a>
</h2>
<p>The <code><a href="../reference/detectAnomaly.html">detectAnomaly()</a></code> function identifies anomalous cells
in a dataset by detecting outliers based on PCA results. Once anomalies
are detected, <code><a href="../reference/calculateCellSimilarityPCA.html">calculateCellSimilarityPCA()</a></code> evaluates how
these outliers influence principal component directions. This analysis
helps determine if anomalous cells significantly impact later PCs, which
capture finer variations in the data. By combining these functions, you
can both pinpoint anomalies and understand their effect on PCA
directions, providing deeper insights into the data structure.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Detect anomalies and select top anomalies for further analysis</span></span>
<span><span class="va">anomaly_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectAnomaly.html">detectAnomaly</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>                                ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>                                pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                                n_tree <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                anomaly_threshold <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select the top 6 anomalies based on the anomaly scores</span></span>
<span><span class="va">top6_anomalies</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">anomaly_output</span><span class="op">$</span><span class="va">Combined</span><span class="op">$</span><span class="va">reference_anomaly_scores</span>, </span>
<span>                             decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate cosine similarity between the top anomalies and top 25 PCs</span></span>
<span><span class="va">cosine_similarities</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateCellSimilarityPCA.html">calculateCellSimilarityPCA</a></span><span class="op">(</span><span class="va">reference_data</span>, </span>
<span>                                                  cell_names <span class="op">=</span> <span class="va">top6_anomalies</span>, </span>
<span>                                                  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">25</span>, </span>
<span>                                                  n_top_vars <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the cosine similarities across PCs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">cosine_similarities</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC"</span>, <span class="fl">15</span><span class="op">:</span><span class="fl">25</span><span class="op">)</span><span class="op">]</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt;                    PC15  PC16  PC17  PC18  PC19  PC20  PC21  PC22  PC23  PC24</span></span>
<span><span class="co">#&gt; ACACCAATCTTAACCT-1 0.10  0.05 -0.24 -0.01 -0.05  0.08 -0.16 -0.15  0.01  0.08</span></span>
<span><span class="co">#&gt; GGCTGGTAGCCAACAG-1 0.06 -0.13 -0.33  0.06 -0.13 -0.03 -0.07 -0.01  0.11  0.04</span></span>
<span><span class="co">#&gt; CATGGCGCATGCTGGC-1 0.13  0.12 -0.08 -0.14 -0.24  0.03  0.01 -0.17 -0.17 -0.05</span></span>
<span><span class="co">#&gt; TACACGAAGCGATAGC-1 0.07 -0.02  0.01 -0.12  0.16 -0.02  0.08  0.14 -0.06 -0.26</span></span>
<span><span class="co">#&gt; GTACTTTTCCGTTGTC-1 0.16 -0.01 -0.24  0.03 -0.02  0.04  0.11 -0.08 -0.15 -0.16</span></span>
<span><span class="co">#&gt; ATTTCTGAGCGATATA-1 0.05 -0.29 -0.24 -0.18 -0.20 -0.05  0.15  0.08  0.15  0.11</span></span>
<span><span class="co">#&gt;                     PC25</span></span>
<span><span class="co">#&gt; ACACCAATCTTAACCT-1 -0.01</span></span>
<span><span class="co">#&gt; GGCTGGTAGCCAACAG-1 -0.26</span></span>
<span><span class="co">#&gt; CATGGCGCATGCTGGC-1 -0.01</span></span>
<span><span class="co">#&gt; TACACGAAGCGATAGC-1 -0.36</span></span>
<span><span class="co">#&gt; GTACTTTTCCGTTGTC-1 -0.17</span></span>
<span><span class="co">#&gt; ATTTCTGAGCGATATA-1 -0.06</span></span></code></pre></div>
<p>Note that there is also a plot method for the object return for
<code><a href="../reference/calculateCellSimilarityPCA.html">calculateCellSimilarityPCA()</a></code>. See reference manual for
details.</p>
</div>
<div class="section level2">
<h2 id="analyzing-cell-distances">Analyzing Cell Distances<a class="anchor" aria-label="anchor" href="#analyzing-cell-distances"></a>
</h2>
<div class="section level3">
<h3 id="calculatecelldistances">
<code>calculateCellDistances()</code><a class="anchor" aria-label="anchor" href="#calculatecelldistances"></a>
</h3>
<p>The <code><a href="../reference/calculateCellDistances.html">calculateCellDistances()</a></code> function calculates
Euclidean distances between cells within the reference dataset and
between query cells and reference cells. This helps in understanding how
closely related cells are within and across datasets based on their
principal component (PC) scores.</p>
</div>
<div class="section level3">
<h3 id="function-usage">Function Usage<a class="anchor" aria-label="anchor" href="#function-usage"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute cell distances</span></span>
<span><span class="va">distance_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateCellDistances.html">calculateCellDistances</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>In the code above:</p>
<ul>
<li>
<code>query_data</code> and reference_data: These are <em><a href="https://bioconductor.org/packages/3.21/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects containing the respective datasets for analysis.</li>
<li>
<code>query_cell_type_col</code> and ref_cell_type_col: These
arguments specify the columns in the <code>colData</code> of each
dataset that contain cell type annotations.</li>
<li>
<code>pc_subset</code>: Specifies which principal components (1 to
10) are used to compute distances. PCA is applied for dimensionality
reduction before calculating distances.</li>
</ul>
</div>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<p>The output <code>distance_data</code> is a list where each entry
contains:</p>
<ul>
<li>
<code>ref_distances</code>: Pairwise distances within the reference
dataset for each cell type.</li>
<li>
<code>query_to_ref_distances</code>: Distances from each query cell
to all reference cells for each cell type.</li>
</ul>
</div>
<div class="section level3">
<h3 id="example-workflow">Example Workflow<a class="anchor" aria-label="anchor" href="#example-workflow"></a>
</h3>
<p>In the example workflow, the first step involves plotting the
distributions of cell distances for selected query cells against various
reference cell populations. This involves visualizing both the distances
between the query cells and the reference cells as well as the distances
among the reference cells themselves. For each cell type, such as CD4
and CD8, these plots illustrate how the selected query cells compare in
terms of their proximity to different reference populations.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify top 6 anomalies for CD4 cells</span></span>
<span><span class="va">cd4_anomalies</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectAnomaly.html">detectAnomaly</a></span><span class="op">(</span></span>
<span>  reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>  query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>  query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>  ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>  n_tree <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  anomaly_threshold <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Top 6 CD4 anomalies</span></span>
<span><span class="va">cd4_top6_anomalies</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">cd4_anomalies</span><span class="op">$</span><span class="va">CD4</span><span class="op">$</span><span class="va">query_anomaly_scores</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot distances for CD4 and CD8 cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">distance_data</span>, ref_cell_type <span class="op">=</span> <span class="st">"CD4"</span>, cell_names <span class="op">=</span> <span class="va">cd4_top6_anomalies</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/AnnotationAnomalies/calculateCellDistances1.png"></p>
<p>For CD4 cells, you might observe that the query cells are located
much further away from the reference cells, indicating a significant
divergence in their density distributions. This suggests that the CD4
query cells are less similar to the CD4 reference cells, showing greater
variation or potential anomalies.</p>
<p>In contrast, for CD8 cells, the plots might reveal more overlap
between the density distributions of query cells and reference cells.
This overlap indicates that the CD8 query cells are more similar to the
CD8 reference cells, suggesting better alignment or less divergence
between their densities. Such insights are crucial for understanding how
closely the query cells match with reference populations and can help in
identifying cell types with notable differences or similarities.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot distances for CD8 cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">distance_data</span>, ref_cell_type <span class="op">=</span> <span class="st">"CD8"</span>, cell_names <span class="op">=</span> <span class="va">cd4_top6_anomalies</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/AnnotationAnomalies/calculateCellDistances2.png"></p>
</div>
<div class="section level3">
<h3 id="calculatecelldistancessimilarity">
<code>calculateCellDistancesSimilarity()</code><a class="anchor" aria-label="anchor" href="#calculatecelldistancessimilarity"></a>
</h3>
<p>The <code><a href="../reference/calculateCellDistancesSimilarity.html">calculateCellDistancesSimilarity()</a></code> function measures
how similar the density distributions of query cells are to those in the
reference dataset. It computes <strong>Bhattacharyya
coefficients</strong> and <strong>Hellinger distances</strong>, which
quantify the overlap between the distributions.</p>
<p>The Bhattacharyya coefficient and Hellinger distance are two measures
used to quantify the similarity between probability distributions. The
Bhattacharyya coefficient ranges from 0 to 1, where a value closer to 1
indicates higher similarity between distributions, while a value closer
to 0 suggests greater dissimilarity. The Hellinger distance, also
ranging from 0 to 1, reflects the divergence between distributions, with
lower values indicating greater similarity and higher values showing
more distinct distributions. In the context of comparing query cells to
reference cells, a high Bhattacharyya coefficient and a low Hellinger
distance both suggest that the query cells have density profiles similar
to those in the reference dataset, which can be desirable depending on
the experimental objectives.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute similarity measures for top 6 CD4 anomalies</span></span>
<span><span class="va">overlap_measures</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateCellDistancesSimilarity.html">calculateCellDistancesSimilarity</a></span><span class="op">(</span></span>
<span>  query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>  reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>  cell_names <span class="op">=</span> <span class="va">cd4_top6_anomalies</span>,</span>
<span>  query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>  ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">overlap_measures</span></span>
<span><span class="co">#&gt; $bhattacharyya_coef</span></span>
<span><span class="co">#&gt;                 Cell        CD4        CD8 B_and_plasma   Myeloid</span></span>
<span><span class="co">#&gt; 1 AACCGCGCAAACCCAT-1 0.14963173 0.15973512    0.6063560 0.2146414</span></span>
<span><span class="co">#&gt; 2 GTTACAGCAGCTGTGC-1 0.05978566 0.07361651    0.5716079 0.1379982</span></span>
<span><span class="co">#&gt; 3 TGAGCATTCCAAACTG-1 0.21854137 0.24409183    0.5797450 0.2559222</span></span>
<span><span class="co">#&gt; 4 TGCTACCTCGGTTCGG-1 0.76503890 0.96527336    0.3480598 0.2815120</span></span>
<span><span class="co">#&gt; 5 TCAGCTCCATACTCTT-1 0.72368485 0.98681446    0.2903559 0.2965596</span></span>
<span><span class="co">#&gt; 6 CTCGAAATCAGTCCCT-1 0.75713220 0.93154956    0.2975623 0.2300326</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $hellinger_dist</span></span>
<span><span class="co">#&gt;                 Cell       CD4       CD8 B_and_plasma   Myeloid</span></span>
<span><span class="co">#&gt; 1 AACCGCGCAAACCCAT-1 0.9221541 0.9166596    0.6274106 0.8862046</span></span>
<span><span class="co">#&gt; 2 GTTACAGCAGCTGTGC-1 0.9696465 0.9624882    0.6545167 0.9284405</span></span>
<span><span class="co">#&gt; 3 TGAGCATTCCAAACTG-1 0.8840015 0.8694298    0.6482708 0.8625994</span></span>
<span><span class="co">#&gt; 4 TGCTACCTCGGTTCGG-1 0.4847279 0.1863509    0.8074282 0.8476367</span></span>
<span><span class="co">#&gt; 5 TCAGCTCCATACTCTT-1 0.5256569 0.1148283    0.8424037 0.8387135</span></span>
<span><span class="co">#&gt; 6 CTCGAAATCAGTCCCT-1 0.4928162 0.2616304    0.8381156 0.8774779</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="r-session-info">R Session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h2>
<pre><code>R version 4.5.1 (2025-06-13)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] scDiagnostics_1.1.0 BiocStyle_2.36.0   

loaded via a namespace (and not attached):
 [1] SummarizedExperiment_1.38.1 gtable_0.3.6               
 [3] xfun_0.52                   bslib_0.9.0                
 [5] ggplot2_3.5.2               htmlwidgets_1.6.4          
 [7] GGally_2.3.0                Biobase_2.68.0             
 [9] lattice_0.22-7              vctrs_0.6.5                
[11] tools_4.5.1                 generics_0.1.4             
[13] parallel_4.5.1              stats4_4.5.1               
[15] tibble_3.3.0                pkgconfig_2.0.3            
[17] Matrix_1.7-3                RColorBrewer_1.1-3         
[19] S7_0.2.0                    desc_1.4.3                 
[21] S4Vectors_0.46.0            ggridges_0.5.6             
[23] lifecycle_1.0.4             GenomeInfoDbData_1.2.14    
[25] compiler_4.5.1              farver_2.1.2               
[27] textshaping_1.0.1           RhpcBLASctl_0.23-42        
[29] GenomeInfoDb_1.44.1         htmltools_0.5.8.1          
[31] sass_0.4.10                 yaml_2.3.10                
[33] tidyr_1.3.1                 pkgdown_2.1.3              
[35] pillar_1.11.0               crayon_1.5.3               
[37] jquerylib_0.1.4             SingleCellExperiment_1.30.1
[39] DelayedArray_0.34.1         cachem_1.1.0               
[41] abind_1.4-8                 ggstats_0.10.0             
[43] tidyselect_1.2.1            digest_0.6.37              
[45] purrr_1.1.0                 dplyr_1.1.4                
[47] bookdown_0.43               labeling_0.4.3             
[49] fastmap_1.2.0               grid_4.5.1                 
[51] cli_3.6.5                   SparseArray_1.8.1          
[53] magrittr_2.0.3              S4Arrays_1.8.1             
[55] withr_3.0.2                 UCSC.utils_1.4.0           
[57] scales_1.4.0                rmarkdown_2.29             
[59] XVector_0.48.0              httr_1.4.7                 
[61] matrixStats_1.5.0           ragg_1.4.0                 
[63] isotree_0.6.1-4             evaluate_1.0.4             
[65] knitr_1.50                  GenomicRanges_1.60.0       
[67] IRanges_2.42.0              rlang_1.1.6                
[69] Rcpp_1.1.0                  glue_1.8.0                 
[71] BiocManager_1.30.26         BiocGenerics_0.54.0        
[73] jsonlite_2.0.0              R6_2.6.1                   
[75] MatrixGenerics_1.20.0       systemfonts_1.2.3          
[77] fs_1.6.6                   </code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
