<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Calculation of Distances Between Specific Cells and Cell Populations â€¢ scDiagnostics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Calculation of Distances Between Specific Cells and Cell Populations">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/QCandAnnotationScores.html">Visualization of QC and Annotation Scores</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/StatisticalMeasures.html">Statistical Measures to Assess Dataset Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection of Annotation Anomalies</a>
    </li>
    <li>
      <a href="../articles/CellDistancesDiagnostics.html">Calculation of Distances Between Specific Cells and Cell Populations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Calculation of Distances Between Specific Cells
and Cell Populations</h1>
                        <h4 data-toc-skip class="author">Anthony
Christidis</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Andrew
Ghazi</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Smriti
Chawla</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Nitesh
Turaga</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Ludwig
Geistlinger</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Robert
Gentleman</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/vignettes/CellDistancesDiagnostics.Rmd" class="external-link"><code>vignettes/CellDistancesDiagnostics.Rmd</code></a></small>
      <div class="hidden name"><code>CellDistancesDiagnostics.Rmd</code></div>

    </address>
</address>
</address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette demonstrates the use of two functions,
<code><a href="../reference/calculateCellDistances.html">calculateCellDistances()</a></code> and
<code><a href="../reference/calculateCellDistancesSimilarity.html">calculateCellDistancesSimilarity()</a></code>, to analyze the
distances between cells in query and reference datasets and to measure
the similarity of density distributions. These functions are useful for
investigating how cells in different datasets relate to each other,
particularly in the context of identifying anomalies and understanding
distribution overlaps.</p>
</div>
<div class="section level2">
<h2 id="datasets">Datasets<a class="anchor" aria-label="anchor" href="#datasets"></a>
</h2>
<p>In the context of the <code>scDiagnostics</code> package, the
following datasets illustrate the application of these tools:</p>
<ul>
<li><p><code>reference_data</code>: A curated and processed dataset
containing expert-assigned cell type annotations. This dataset serves as
a reference for comparison and can be used alone to detect anomalies
within the reference annotations.</p></li>
<li><p><code>query_data</code>: A dataset that also includes
expert-assigned cell type annotations, but additionally features
annotations generated by the <em><a href="https://bioconductor.org/packages/3.19/SingleR" class="external-link">SingleR</a></em>
package. This allows for the comparison of expert annotations with those
produced by an automated method to detect inconsistencies or
anomalies.</p></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/scDiagnostics" class="external-link">scDiagnostics</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="computing-cell-distances">Computing Cell Distances<a class="anchor" aria-label="anchor" href="#computing-cell-distances"></a>
</h2>
<p>The <code><a href="../reference/calculateCellDistances.html">calculateCellDistances()</a></code> function calculates
Euclidean distances between cells within the reference dataset and
between query cells and reference cells. This helps in understanding how
closely related cells are within and across datasets based on their
principal component (PC) scores.</p>
<div class="section level3">
<h3 id="function-usage">Function Usage<a class="anchor" aria-label="anchor" href="#function-usage"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute cell distances</span></span>
<span><span class="va">distance_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateCellDistances.html">calculateCellDistances</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>In the code above:</p>
<ul>
<li>
<code>query_data</code> and reference_data: These are <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects containing the respective datasets for analysis.</li>
<li>
<code>query_cell_type_col</code> and ref_cell_type_col: These
arguments specify the columns in the <code>colData</code> of each
dataset that contain cell type annotations.</li>
<li>
<code>pc_subset</code>: Specifies which principal components (1 to
10) are used to compute distances. PCA is applied for dimensionality
reduction before calculating distances.</li>
</ul>
</div>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<p>The output <code>distance_data</code> is a list where each entry
contains:</p>
<ul>
<li>
<code>ref_distances</code>: Pairwise distances within the reference
dataset for each cell type.</li>
<li>
<code>query_to_ref_distances</code>: Distances from each query cell
to all reference cells for each cell type.</li>
</ul>
</div>
<div class="section level3">
<h3 id="example-workflow">Example Workflow<a class="anchor" aria-label="anchor" href="#example-workflow"></a>
</h3>
<p>In the example workflow, the first step involves plotting the
distributions of cell distances for selected query cells against various
reference cell populations. This involves visualizing both the distances
between the query cells and the reference cells as well as the distances
among the reference cells themselves. For each cell type, such as CD4
and CD8, these plots illustrate how the selected query cells compare in
terms of their proximity to different reference populations.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify top 6 anomalies for CD4 cells</span></span>
<span><span class="va">cd4_anomalies</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectAnomaly.html">detectAnomaly</a></span><span class="op">(</span></span>
<span>  reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>  query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>  query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>  ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>  n_tree <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  anomaly_treshold <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Top 6 CD4 anomalies</span></span>
<span><span class="va">cd4_top6_anomalies</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">cd4_anomalies</span><span class="op">$</span><span class="va">CD4</span><span class="op">$</span><span class="va">query_anomaly_scores</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot distances for CD4 and CD8 cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">distance_data</span>, ref_cell_type <span class="op">=</span> <span class="st">"CD4"</span>, cell_names <span class="op">=</span> <span class="va">cd4_top6_anomalies</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/CellDistancesDiagnostics/calculateCellDistances1.png"></p>
<p>For CD4 cells, you might observe that the query cells are located
much further away from the reference cells, indicating a significant
divergence in their density distributions. This suggests that the CD4
query cells are less similar to the CD4 reference cells, showing greater
variation or potential anomalies.</p>
<p>In contrast, for CD8 cells, the plots might reveal more overlap
between the density distributions of query cells and reference cells.
This overlap indicates that the CD8 query cells are more similar to the
CD8 reference cells, suggesting better alignment or less divergence
between their densities. Such insights are crucial for understanding how
closely the query cells match with reference populations and can help in
identifying cell types with notable differences or similarities.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot distances for CD8 cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">distance_data</span>, ref_cell_type <span class="op">=</span> <span class="st">"CD8"</span>, cell_names <span class="op">=</span> <span class="va">cd4_top6_anomalies</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/CellDistancesDiagnostics/calculateCellDistances2.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="calculating-similarity-measures">Calculating Similarity Measures<a class="anchor" aria-label="anchor" href="#calculating-similarity-measures"></a>
</h2>
<p>The <code><a href="../reference/calculateCellDistancesSimilarity.html">calculateCellDistancesSimilarity()</a></code> function measures
how similar the density distributions of query cells are to those in the
reference dataset. It computes <strong>Bhattacharyya
coefficients</strong> and <strong>Hellinger distances</strong>, which
quantify the overlap between the distributions.</p>
<p>The Bhattacharyya coefficient and Hellinger distance are two measures
used to quantify the similarity between probability distributions. The
Bhattacharyya coefficient ranges from 0 to 1, where a value closer to 1
indicates higher similarity between distributions, while a value closer
to 0 suggests greater dissimilarity. The Hellinger distance, also
ranging from 0 to 1, reflects the divergence between distributions, with
lower values indicating greater similarity and higher values showing
more distinct distributions. In the context of comparing query cells to
reference cells, a high Bhattacharyya coefficient and a low Hellinger
distance both suggest that the query cells have density profiles similar
to those in the reference dataset, which can be desirable depending on
the experimental objectives.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute similarity measures for top 6 CD4 anomalies</span></span>
<span><span class="va">overlap_measures</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateCellDistancesSimilarity.html">calculateCellDistancesSimilarity</a></span><span class="op">(</span></span>
<span>  query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>  reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>  cell_names <span class="op">=</span> <span class="va">cd4_top6_anomalies</span>,</span>
<span>  query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>  ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">overlap_measures</span></span>
<span><span class="co">#&gt; $bhattacharyya_coef</span></span>
<span><span class="co">#&gt;                 Cell       CD4       CD8 B_and_plasma   Myeloid</span></span>
<span><span class="co">#&gt; 1 TTGGAACTCGCTTAGA-1 0.4664148 0.9678914    0.3587941 0.2574115</span></span>
<span><span class="co">#&gt; 2 GATCGCGTCCCATTAT-1 0.6695226 0.9709480    0.3327168 0.2051765</span></span>
<span><span class="co">#&gt; 3 AGCATACGTTGAGTTC-1 0.6526016 0.9609405    0.3085990 0.2102890</span></span>
<span><span class="co">#&gt; 4 TAAGCGTGTAATCGTC-1 0.6049953 0.9677820    0.3331004 0.1622547</span></span>
<span><span class="co">#&gt; 5 CCATGTCTCGTCTGCT-1 0.6158507 0.9812964    0.3421234 0.2000697</span></span>
<span><span class="co">#&gt; 6 TGCCCTAAGCGGCTTC-1 0.6184935 0.9558004    0.3118239 0.1380546</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $hellinger_dist</span></span>
<span><span class="co">#&gt;                 Cell       CD4       CD8 B_and_plasma   Myeloid</span></span>
<span><span class="co">#&gt; 1 TTGGAACTCGCTTAGA-1 0.7304692 0.1791887    0.8007534 0.8617357</span></span>
<span><span class="co">#&gt; 2 GATCGCGTCCCATTAT-1 0.5748716 0.1704463    0.8168741 0.8915288</span></span>
<span><span class="co">#&gt; 3 AGCATACGTTGAGTTC-1 0.5894051 0.1976348    0.8315053 0.8886569</span></span>
<span><span class="co">#&gt; 4 TAAGCGTGTAATCGTC-1 0.6284940 0.1794938    0.8166392 0.9152843</span></span>
<span><span class="co">#&gt; 5 CCATGTCTCGTCTGCT-1 0.6197978 0.1367610    0.8110959 0.8943882</span></span>
<span><span class="co">#&gt; 6 TGCCCTAAGCGGCTTC-1 0.6176621 0.2102369    0.8295638 0.9284102</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>This vignette illustrates how to leverage the
<code><a href="../reference/calculateCellDistances.html">calculateCellDistances()</a></code> and
<code><a href="../reference/calculateCellDistancesSimilarity.html">calculateCellDistancesSimilarity()</a></code> functions to analyze
cell distances and distribution similarities in single-cell datasets. By
calculating Euclidean distances between cells within and across
datasets, you gain insights into how closely related cells are in the
PCA-reduced space. The subsequent step of computing similarity measures,
such as Bhattacharyya coefficients and Hellinger distances, allows for a
deeper understanding of how similar the density distributions of query
cells are to those in the reference dataset.</p>
<hr>
</div>
<div class="section level2">
<h2 id="r-session-info">R Session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h2>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] scDiagnostics_1.0.0 BiocStyle_2.32.1   

loaded via a namespace (and not attached):
 [1] SummarizedExperiment_1.34.0 gtable_0.3.5               
 [3] xfun_0.47                   bslib_0.8.0                
 [5] ggplot2_3.5.1               htmlwidgets_1.6.4          
 [7] Biobase_2.64.0              lattice_0.22-6             
 [9] generics_0.1.3              vctrs_0.6.5                
[11] tools_4.4.1                 stats4_4.4.1               
[13] parallel_4.4.1              tibble_3.2.1               
[15] fansi_1.0.6                 pkgconfig_2.0.3            
[17] Matrix_1.7-0                desc_1.4.3                 
[19] S4Vectors_0.42.1            lifecycle_1.0.4            
[21] GenomeInfoDbData_1.2.12     farver_2.1.2               
[23] compiler_4.4.1              textshaping_0.4.0          
[25] munsell_0.5.1               GenomeInfoDb_1.40.1        
[27] htmltools_0.5.8.1           sass_0.4.9                 
[29] yaml_2.3.10                 pkgdown_2.1.1              
[31] pillar_1.9.0                crayon_1.5.3               
[33] jquerylib_0.1.4             SingleCellExperiment_1.26.0
[35] DelayedArray_0.30.1         cachem_1.1.0               
[37] abind_1.4-8                 tidyselect_1.2.1           
[39] digest_0.6.37               dplyr_1.1.4                
[41] bookdown_0.40               labeling_0.4.3             
[43] fastmap_1.2.0               grid_4.4.1                 
[45] colorspace_2.1-1            cli_3.6.3                  
[47] SparseArray_1.4.8           magrittr_2.0.3             
[49] S4Arrays_1.4.1              utf8_1.2.4                 
[51] withr_3.0.1                 UCSC.utils_1.0.0           
[53] scales_1.3.0                rmarkdown_2.28             
[55] XVector_0.44.0              httr_1.4.7                 
[57] matrixStats_1.4.1           ragg_1.3.3                 
[59] isotree_0.6.1-1             evaluate_1.0.0             
[61] knitr_1.48                  GenomicRanges_1.56.1       
[63] IRanges_2.38.1              rlang_1.1.4                
[65] Rcpp_1.0.13                 glue_1.7.0                 
[67] BiocManager_1.30.25         BiocGenerics_0.50.0        
[69] jsonlite_1.8.9              R6_2.5.1                   
[71] MatrixGenerics_1.16.0       systemfonts_1.1.0          
[73] fs_1.6.4                    zlibbioc_1.50.0            </code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
