<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Evaluation of Dataset and Marker Gene Alignment • scDiagnostics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Evaluation of Dataset and Marker Gene Alignment">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/QCandAnnotationScores.html">Visualization of QC and Annotation Scores</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/StatisticalMeasures.html">Statistical Measures to Assess Dataset Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection of Annotation Anomalies</a>
    </li>
    <li>
      <a href="../articles/CellDistancesDiagnostics.html">Calculation of Distances Between Specific Cells and Cell Populations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Evaluation of Dataset and Marker Gene
Alignment</h1>
                        <h4 data-toc-skip class="author">Anthony
Christidis</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Andrew
Ghazi</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Smriti
Chawla</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Nitesh
Turaga</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Ludwig
Geistlinger</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Robert
Gentleman</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/vignettes/DatasetMarkerGeneAlignment.Rmd" class="external-link"><code>vignettes/DatasetMarkerGeneAlignment.Rmd</code></a></small>
      <div class="hidden name"><code>DatasetMarkerGeneAlignment.Rmd</code></div>

    </address>
</address>
</address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In the realm of single-cell genomics, the ability to compare and
integrate data across different conditions, datasets, or methodologies
is crucial for deriving meaningful biological insights. This vignette
introduces several functions designed to facilitate such comparisons and
analyses by providing robust tools for evaluating and visualizing
similarities and differences in high-dimensional data.</p>
<div class="section level3">
<h3 id="functions-for-evaluation-of-dataset-alignment">Functions for Evaluation of Dataset Alignment<a class="anchor" aria-label="anchor" href="#functions-for-evaluation-of-dataset-alignment"></a>
</h3>
<ul>
<li><p><code><a href="../reference/compareCCA.html">compareCCA()</a></code>: This function enables the comparison
of datasets by applying Canonical Correlation Analysis (CCA). It helps
assess how well two datasets align with each other, providing insights
into the relationship between different single-cell experiments or
conditions.</p></li>
<li><p><code><a href="../reference/comparePCA.html">comparePCA()</a></code>: This function allows you to compare
datasets using Principal Component Analysis (PCA). It evaluates how
similar or different the principal components are between two datasets,
offering a way to understand the underlying structure and variance in
your data.</p></li>
<li><p><code><a href="../reference/comparePCASubspace.html">comparePCASubspace()</a></code>: Extending the comparison to
specific subspaces, this function focuses on subsets of principal
components. It provides a detailed analysis of how subspace structures
differ or align between datasets, which is valuable for fine-grained
comparative studies.</p></li>
<li><p><code><a href="../reference/plotPairwiseDistancesDensity.html">plotPairwiseDistancesDensity()</a></code>: To visualize the
distribution of distances between pairs of samples, this function
generates density plots. It helps in understanding the variation and
relationships between samples in high-dimensional spaces.</p></li>
<li><p><code><a href="../reference/plotWassersteinDistance.html">plotWassersteinDistance()</a></code>: This function visualizes
the Wasserstein distance, a metric for comparing distributions, across
datasets. It provides an intuitive view of how distributions differ
between datasets, aiding in the evaluation of alignment and
discrepancies.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="marker-gene-alignment">Marker Gene Alignment<a class="anchor" aria-label="anchor" href="#marker-gene-alignment"></a>
</h3>
<ul>
<li><p><code><a href="../reference/calculateHVGOverlap.html">calculateHVGOverlap()</a></code>: To assess the similarity
between datasets based on highly variable genes, this function computes
the overlap coefficient. It measures how well the sets of highly
variable genes from different datasets correspond to each
other.</p></li>
<li><p><code><a href="../reference/calculateVarImpOverlap.html">calculateVarImpOverlap()</a></code>: Using Random Forest, this
function identifies and compares the importance of genes for
differentiating cell types between datasets. It highlights which genes
are most critical in each dataset and compares their importance,
providing insights into shared and unique markers.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="purpose-and-applications">Purpose and Applications<a class="anchor" aria-label="anchor" href="#purpose-and-applications"></a>
</h3>
<p>These functions collectively offer a comprehensive toolkit for
comparing and analyzing single-cell data. Whether you are assessing
alignment between datasets, visualizing distance distributions, or
identifying key genes, these tools are designed to enhance your ability
to derive meaningful insights from complex, high-dimensional data.</p>
<p>In this vignette, we will guide you through the practical use of each
function, demonstrate how to interpret their outputs, and show how they
can be integrated into your single-cell genomics workflow.</p>
</div>
</div>
<div class="section level2">
<h2 id="preliminaries">Preliminaries<a class="anchor" aria-label="anchor" href="#preliminaries"></a>
</h2>
<p>In the context of the <code>scDiagnostics</code> package, this
vignette demonstrates how to leverage various functions to evaluate and
compare single-cell data across two distinct datasets:</p>
<ul>
<li>
<code>reference_data</code>: This dataset features meticulously
curated cell type annotations assigned by experts. It serves as a robust
benchmark for evaluating the accuracy and consistency of cell type
annotations across different datasets, offering a reliable standard
against which other annotations can be assessed.</li>
<li>
<code>query_data</code>: This dataset contains cell type annotations
from both expert assessments and those generated using the <em><a href="https://bioconductor.org/packages/3.19/SingleR" class="external-link">SingleR</a></em>
package. By comparing these annotations with those from the reference
dataset, you can identify discrepancies between manual and automated
results, highlighting potential inconsistencies or areas requiring
further investigation.</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/scDiagnostics" class="external-link">scDiagnostics</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>Some functions in the vignette are designed to work with <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects that contain data from only one cell type. We will create
separate <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects that only CD4 cells, to ensure compatibility with these
functions.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset to CD4 cells</span></span>
<span><span class="va">ref_data_cd4</span> <span class="op">&lt;-</span> <span class="va">reference_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span></span>
<span>    <span class="va">reference_data</span><span class="op">$</span><span class="va">expert_annotation</span> <span class="op">==</span> <span class="st">"CD4"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="va">query_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span></span>
<span>    <span class="va">query_data</span><span class="op">$</span><span class="va">expert_annotation</span> <span class="op">==</span> <span class="st">"CD4"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Select highly variable genes</span></span>
<span><span class="va">ref_top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">ref_data_cd4</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">query_top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">query_data_cd4</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">common_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">ref_top_genes</span>, <span class="va">query_top_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset data by common genes</span></span>
<span><span class="va">ref_data_cd4</span> <span class="op">&lt;-</span> <span class="va">ref_data_cd4</span><span class="op">[</span><span class="va">common_genes</span>,<span class="op">]</span></span>
<span><span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="va">query_data_cd4</span><span class="op">[</span><span class="va">common_genes</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Run PCA on both datasets</span></span>
<span><span class="va">ref_data_cd4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">ref_data_cd4</span><span class="op">)</span></span>
<span><span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">query_data_cd4</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="evaluation-of-dataset-alignment">Evaluation of Dataset Alignment<a class="anchor" aria-label="anchor" href="#evaluation-of-dataset-alignment"></a>
</h2>
<div class="section level3">
<h3 id="comparecca">
<code>compareCCA()</code><a class="anchor" aria-label="anchor" href="#comparecca"></a>
</h3>
<p>In single-cell genomics, datasets from different sources or
experimental conditions often need to be compared to understand how well
they align. The <code>compareCCA</code> function facilitates this
comparison by performing Canonical Correlation Analysis (CCA) between a
query dataset and a reference dataset. This function is particularly
useful when datasets have different sample sizes or distributions, and
it helps in assessing the similarity between them after projecting them
onto a common principal component space.</p>
<p><code><a href="../reference/compareCCA.html">compareCCA()</a></code> performs the following steps:</p>
<ul>
<li>
<strong>Project Data</strong>: Project the query dataset onto the
principal component space of the reference dataset. This ensures that
both datasets are represented in the same reduced-dimensional space for
a meaningful comparison.</li>
<li>
<strong>Perform CCA</strong>: Conduct Canonical Correlation Analysis
to find the relationships between the projected datasets. CCA helps to
identify how the canonical variables derived from the query dataset
correlate with those from the reference dataset.</li>
<li>
<strong>Compare Alignment</strong>: Evaluate the alignment between
the datasets by computing cosine similarities between the canonical
correlation vectors derived from the CCA. This provides a measure of how
well the two datasets correspond to each other in the canonical
space.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform CCA </span></span>
<span><span class="va">cca_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compareCCA.html">compareCCA</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_cd4</span>, </span>
<span>                             reference_data <span class="op">=</span> <span class="va">ref_data_cd4</span>, </span>
<span>                             query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                             ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                             pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cca_comparison</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/DatasetMarkerGeneAlignment/compareCCA.png"></p>
<p>Canonical Correlation Analysis (CCA) produces several key
outputs:</p>
<ul>
<li>
<strong>Canonical Variables</strong>: For each dataset, CCA computes
a set of canonical variables that maximize the correlation between the
datasets. These variables are linear combinations of the original
variables.</li>
<li>
<strong>Canonical Correlation Coefficients</strong>: These
coefficients indicate the strength of the correlation between the
canonical variables of the two datasets. Higher values suggest stronger
correlations.</li>
<li>
<strong>Cosine Similarities</strong>: After obtaining the canonical
variables, the function computes the cosine similarity between these
variables from the query and reference datasets. This similarity measure
helps assess how closely the canonical variables align between the two
datasets.</li>
</ul>
<p>In summary, the <code>compareCCA</code> function allows you to
compare how well two datasets are aligned by projecting them into a
shared PCA space, performing CCA, and then evaluating the similarity of
the canonical variables. This approach is valuable for integrative
analyses and understanding the relationships between different datasets
in single-cell studies.</p>
</div>
<div class="section level3">
<h3 id="comparepca">
<code>comparePCA()</code><a class="anchor" aria-label="anchor" href="#comparepca"></a>
</h3>
<p>The <code><a href="../reference/comparePCA.html">comparePCA()</a></code> function compares the PCA subspaces
between the query and reference datasets. It calculates the principal
angles between the PCA subspaces to assess the alignment and similarity
between them. This is useful for understanding how well the
dimensionality reduction spaces of your datasets match.</p>
<p><code><a href="../reference/comparePCA.html">comparePCA()</a></code> operates as follows:</p>
<ul>
<li>
<strong>PCA Computation</strong>: It computes PCA for both query and
reference datasets, reducing them into lower-dimensional spaces.</li>
<li>
<strong>Subspace Comparison</strong>: The function calculates the
principal angles between the PCA subspaces of the query and reference
datasets. These angles help determine how closely the subspaces
align.</li>
<li>
<strong>Distance Metrics</strong>: It uses distance metrics based on
principal angles to quantify the similarity between the datasets.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform PCA </span></span>
<span><span class="va">pca_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/comparePCA.html">comparePCA</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_cd4</span>, </span>
<span>                             reference_data <span class="op">=</span> <span class="va">ref_data_cd4</span>, </span>
<span>                             query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                             ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                             pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, </span>
<span>                             metric <span class="op">=</span> <span class="st">"cosine"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pca_comparison</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="comparepcasubspace">
<code>comparePCASubspace()</code><a class="anchor" aria-label="anchor" href="#comparepcasubspace"></a>
</h3>
<p>In single-cell RNA-seq analysis, it is essential to assess the
similarity between the subspaces spanned by the top principal components
(PCs) of query and reference datasets. This is particularly important
when comparing the structure and variation captured by each dataset. The
<code><a href="../reference/comparePCASubspace.html">comparePCASubspace()</a></code> function is designed to provide
insights into how well the subspaces align by computing the cosine
similarity between the loadings of the top variables for each PC. This
analysis helps in determining the degree of correspondence between the
datasets, which is critical for accurate cell type annotation and data
integration.</p>
<p><code><a href="../reference/comparePCASubspace.html">comparePCASubspace()</a></code> performs the following
operations:</p>
<ul>
<li>
<strong>Cosine Similarity Calculation</strong>: The function
computes the cosine similarity between the loadings of the top variables
for each PC in both the query and reference datasets. This similarity
measures how closely the two datasets align in the space spanned by
their principal components.</li>
<li>
<strong>Selection of Top Similarities</strong>: The function selects
the top cosine similarity scores and stores the corresponding PC indices
from both datasets. This step identifies the most aligned principal
components between the two datasets.</li>
<li>
<strong>Variance Explained Calculation</strong>: It then calculates
the average percentage of variance explained by the selected top PCs.
This helps in understanding how much of the data’s variance is captured
by these components.</li>
<li>
<strong>Weighted Cosine Similarity Score</strong>: Finally, the
function computes a weighted cosine similarity score based on the top
cosine similarities and the average percentage of variance explained.
This score provides an overall measure of subspace alignment between the
datasets.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compare PCA subspaces between query and reference data</span></span>
<span><span class="va">subspace_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/comparePCASubspace.html">comparePCASubspace</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data_cd4</span>,</span>
<span>    reference_data <span class="op">=</span> <span class="va">ref_data_cd4</span>, </span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View weighted cosine similarity score</span></span>
<span><span class="va">subspace_comparison</span><span class="op">$</span><span class="va">weighted_cosine_similarity</span></span>
<span><span class="co">#&gt; [1] 0.2578375</span></span>
<span></span>
<span><span class="co"># Plot output for PCA subspace comparison (if a plot method is available)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">subspace_comparison</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/DatasetMarkerGeneAlignment/comparePCASubspace.png"></p>
<p>In the results:</p>
<ul>
<li>
<strong>Cosine Similarity</strong>: The values in
principal_angles_cosines indicate the degree of alignment between the
principal components of the query and reference datasets. Higher values
suggest stronger alignment.</li>
<li>
<strong>Variance Explained</strong>: The average_variance_explained
vector provides the average percentage of variance captured by the
selected top PCs in both datasets. This helps assess the importance of
these PCs in explaining data variation.</li>
<li>
<strong>Weighted Cosine Similarity</strong>: The
weighted_cosine_similarity score combines the cosine similarity with
variance explained to give a comprehensive measure of how well the
subspaces align. A higher score indicates that the datasets are
well-aligned in the PCA space.</li>
</ul>
<p>By using <code><a href="../reference/comparePCASubspace.html">comparePCASubspace()</a></code>, you can quantify the
alignment of PCA subspaces between query and reference datasets, aiding
in the evaluation of dataset integration and the reliability of cell
type annotations.</p>
</div>
<div class="section level3">
<h3 id="plotpairwisedistancesdensity">
<code>plotPairwiseDistancesDensity()</code><a class="anchor" aria-label="anchor" href="#plotpairwisedistancesdensity"></a>
</h3>
<div class="section level4">
<h4 id="purpose">Purpose<a class="anchor" aria-label="anchor" href="#purpose"></a>
</h4>
<p>The <code><a href="../reference/plotPairwiseDistancesDensity.html">plotPairwiseDistancesDensity()</a></code> function is designed
to calculate and visualize the pairwise distances or correlations
between cell types in query and reference datasets. This function is
particularly useful in single-cell RNA sequencing (scRNA-seq) analysis,
where it is essential to evaluate the consistency and reliability of
cell type annotations by comparing their expression profiles.</p>
</div>
<div class="section level4">
<h4 id="functionality">Functionality<a class="anchor" aria-label="anchor" href="#functionality"></a>
</h4>
<p>The function operates on <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects, which are commonly used to store single-cell data, including
expression matrices and associated metadata. Users specify the cell
types of interest in both the query and reference datasets, and the
function computes either the distances or correlation coefficients
between these cells.</p>
<p>When principal component analysis (PCA) is applied, the function
projects the expression data into a lower-dimensional PCA space, which
can be specified by the user. This allows for a more focused analysis of
the major sources of variation in the data. Alternatively, if no
dimensionality reduction is desired, the function can directly use the
expression data for computation.</p>
<p>Depending on the user’s choice, the function can calculate pairwise
Euclidean distances or correlation coefficients. The resulting values
are used to compare the relationships between cells within the same
dataset (either query or reference) and between cells across the two
datasets.</p>
</div>
<div class="section level4">
<h4 id="interpretation">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation"></a>
</h4>
<p>The output of the function is a density plot generated by
<code>ggplot2</code>, which displays the distribution of pairwise
distances or correlations. The plot provides three key comparisons:</p>
<ul>
<li>Query vs. Query,</li>
<li>Reference vs. Reference,</li>
<li>Query vs. Reference.</li>
</ul>
<p>By examining these density curves, users can assess the similarity of
cells within each dataset and across datasets. For example, a higher
density of lower distances in the “Query vs. Reference” comparison would
suggest that the query and reference cells are similar in their
expression profiles, indicating consistency in the annotation of the
cell type across the datasets.</p>
<p>This visual approach offers an intuitive way to diagnose potential
discrepancies in cell type annotations, identify outliers, or confirm
the reliability of the cell type assignments.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example usage of the function</span></span>
<span><span class="fu"><a href="../reference/plotPairwiseDistancesDensity.html">plotPairwiseDistancesDensity</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                             reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                             query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                             ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                             cell_type_query <span class="op">=</span> <span class="st">"CD4"</span>, </span>
<span>                             cell_type_ref <span class="op">=</span> <span class="st">"CD4"</span>, </span>
<span>                             pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>                             distance_metric <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/DatasetMarkerGeneAlignment/plotPairwiseDistancesDensity.png"></p>
<p>This example demonstrates how to compare CD4 cells between a query
and reference dataset, with PCA applied to the first five principal
components and pairwise Euclidean distances calculated. The output is a
density plot that helps visualize the distribution of these distances,
aiding in the interpretation of the similarity between the two
datasets.</p>
</div>
</div>
<div class="section level3">
<h3 id="plotwassersteindistance">
<code>plotWassersteinDistance()</code><a class="anchor" aria-label="anchor" href="#plotwassersteindistance"></a>
</h3>
<p>The <code><a href="../reference/plotWassersteinDistance.html">plotWassersteinDistance()</a></code> function creates a density
plot to compare the Wasserstein distances between a reference dataset
and a query dataset under the null hypothesis. The null hypothesis
assumes that both datasets are drawn from the same distribution. This
function is useful for evaluating how different the query data is from
the reference data based on their Wasserstein distances.</p>
<div class="section level4">
<h4 id="how-the-function-operates">How the Function Operates<a class="anchor" aria-label="anchor" href="#how-the-function-operates"></a>
</h4>
<ol style="list-style-type: decimal">
<li>
<strong>Projecting Data into PCA Space</strong>:</li>
</ol>
<ul>
<li>The query data is projected into the PCA space defined by the
reference data. This step aligns the datasets by reducing their
dimensions using principal components.</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>
<strong>Calculating the Null Distribution</strong>:</li>
</ol>
<ul>
<li>The function generates a null distribution of Wasserstein distances
by repeatedly sampling subsets from the reference data. This represents
the distribution of distances you would expect if both datasets were
from the same source.</li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li>
<strong>Computing Reference-Query Distance</strong>:</li>
</ol>
<ul>
<li>The Wasserstein distance between the projected reference and query
data is computed. This distance quantifies how different the query
dataset is from the reference dataset.</li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li>
<strong>Creating the Density Plot</strong>:</li>
</ol>
<ul>
<li>The function generates a density plot of the null distribution of
Wasserstein distances.</li>
<li>Vertical lines are added to indicate:
<ul>
<li>The <strong>significance threshold</strong>: Distances greater than
this threshold suggest a significant difference between the
datasets.</li>
<li>The <strong>reference-query distance</strong>: The actual distance
between the reference and query datasets.</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="interpretation-1">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation-1"></a>
</h4>
<ul>
<li>
<strong>Density Plot</strong>: This plot shows the distribution of
Wasserstein distances under the assumption that the datasets are
similar.</li>
<li>
<strong>Significance Threshold Line</strong>: If the reference-query
distance exceeds this line, it indicates a significant difference
between the datasets.</li>
<li>
<strong>Reference-Query Distance Line</strong>: This line shows the
observed distance between the reference and query datasets. If this line
is to the right of the significance threshold, the difference is
statistically significant.</li>
</ul>
</div>
<div class="section level4">
<h4 id="code-example">Code Example<a class="anchor" aria-label="anchor" href="#code-example"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate the Wasserstein distance density plot</span></span>
<span><span class="fu"><a href="../reference/plotWassersteinDistance.html">plotWassersteinDistance</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_cd4</span>,</span>
<span>                        reference_data <span class="op">=</span> <span class="va">ref_data_cd4</span>, </span>
<span>                        query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                        ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                        pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>                        alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/DatasetMarkerGeneAlignment/plotWassersteinDistance.png"></p>
<p>This example demonstrates how to use the
<code><a href="../reference/plotWassersteinDistance.html">plotWassersteinDistance()</a></code> function to compare Wasserstein
distances between CD4 cells in the reference and query datasets. The
resulting plot helps determine whether the difference between the
datasets is statistically significant.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="marker-gene-alignment-1">Marker Gene Alignment<a class="anchor" aria-label="anchor" href="#marker-gene-alignment-1"></a>
</h2>
<div class="section level3">
<h3 id="calculatehvgoverlap">
<code>calculateHVGOverlap()</code><a class="anchor" aria-label="anchor" href="#calculatehvgoverlap"></a>
</h3>
<p>The <code><a href="../reference/calculateHVGOverlap.html">calculateHVGOverlap()</a></code> function computes the overlap
coefficient between two sets of highly variable genes (HVGs) from a
reference dataset and a query dataset. The overlap coefficient is a
measure of similarity between the two sets, reflecting how much the HVGs
in one dataset overlap with those in the other.</p>
<div class="section level4">
<h4 id="how-the-function-operates-1">How the Function Operates<a class="anchor" aria-label="anchor" href="#how-the-function-operates-1"></a>
</h4>
<p>The function begins by ensuring that the input vectors
<code>reference_genes</code> and <code>query_genes</code> are character
vectors and that neither of them is empty. Once these checks are
complete, the function identifies the common genes between the two sets
using the intersect function, which finds the intersection of the two
gene sets.</p>
<p>Next, the function calculates the size of this intersection,
representing the number of genes common to both sets. The overlap
coefficient is then computed by dividing the size of the intersection by
the size of the smaller set of genes. This ensures that the coefficient
is a value between 0 and 1, where 0 indicates no overlap and 1 indicates
complete overlap.</p>
<p>Finally, the function rounds the overlap coefficient to two decimal
places before returning it as the output.</p>
</div>
<div class="section level4">
<h4 id="interpretation-2">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation-2"></a>
</h4>
<p>The overlap coefficient quantifies the extent to which the HVGs in
the reference dataset align with those in the query dataset. A higher
coefficient indicates a stronger similarity between the two datasets in
terms of their HVGs, which can suggest that the datasets are more
comparable or that they capture similar biological variability.
Conversely, a lower coefficient indicates less overlap, suggesting that
the datasets may be capturing different biological signals or that they
are less comparable.</p>
</div>
<div class="section level4">
<h4 id="code-example-1">Code Example<a class="anchor" aria-label="anchor" href="#code-example-1"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load library to get top HVGs</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select the top 500 highly variable genes from both datasets</span></span>
<span><span class="va">ref_var_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">reference_data</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">query_var_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">query_data</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate the overlap coefficient between the reference and query HVGs</span></span>
<span><span class="va">overlap_coefficient</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVGOverlap.html">calculateHVGOverlap</a></span><span class="op">(</span>reference_genes <span class="op">=</span> <span class="va">ref_var_genes</span>, </span>
<span>                                           query_genes <span class="op">=</span> <span class="va">query_var_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Display the overlap coefficient</span></span>
<span><span class="va">overlap_coefficient</span></span>
<span><span class="co">#&gt; [1] 0.93</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="calculatevarimpoverlap">
<code>calculateVarImpOverlap()</code><a class="anchor" aria-label="anchor" href="#calculatevarimpoverlap"></a>
</h3>
<div class="section level4">
<h4 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h4>
<p>The <code><a href="../reference/calculateVarImpOverlap.html">calculateVarImpOverlap()</a></code> function helps you identify
and compare the most important genes for distinguishing cell types in
both a reference dataset and a query dataset. It does this using the
Random Forest algorithm, which calculates how important each gene is in
differentiating between cell types.</p>
</div>
<div class="section level4">
<h4 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h4>
<p>To use the function, you need to provide a reference dataset
containing expression data and cell type annotations. Optionally, you
can also provide a query dataset if you want to compare gene importance
across both datasets. The function allows you to specify which cell
types to analyze and how many trees to use in the Random Forest model.
Additionally, you can decide how many top genes you want to compare
between the datasets.</p>
</div>
<div class="section level4">
<h4 id="code-example-2">Code Example<a class="anchor" aria-label="anchor" href="#code-example-2"></a>
</h4>
<p>Let’s say you have a reference dataset (<code>reference_data</code>)
and a query dataset (<code>query_data</code>). Both datasets contain
expression data and cell type annotations, stored in columns named
“expert_annotation” and <code>SingleR_annotation</code>, respectively.
You want to calculate the importance of genes using 500 trees and
compare the top 50 genes between the datasets.</p>
<p>Here’s how you would use the function:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># RF function to compare which genes are best at differentiating cell types</span></span>
<span><span class="va">rf_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateVarImpOverlap.html">calculateVarImpOverlap</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                    query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                    query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                    n_tree <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                    n_top <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Comparison table</span></span>
<span><span class="va">rf_output</span><span class="op">$</span><span class="va">var_imp_comparison</span></span>
<span><span class="co">#&gt;              CD4-CD8     CD4-B_and_plasma          CD4-Myeloid </span></span>
<span><span class="co">#&gt;                 0.84                 0.82                 0.86 </span></span>
<span><span class="co">#&gt;     CD8-B_and_plasma          CD8-Myeloid B_and_plasma-Myeloid </span></span>
<span><span class="co">#&gt;                 0.84                 0.80                 0.76</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="interpretation-3">Interpretation:<a class="anchor" aria-label="anchor" href="#interpretation-3"></a>
</h4>
<p>After running the function, you’ll receive the importance scores of
genes for each pair of cell types in your reference dataset. If you
provided a query dataset, you’ll also get the importance scores for
those cell types. The function will tell you how much the top genes in
the reference and query datasets overlap, which helps you understand if
the same genes are important for distinguishing cell types across
different datasets.</p>
<p>For example, if there’s a high overlap, it suggests that similar
genes are crucial in both datasets for differentiating the cell types,
which could be important for validating your findings or identifying
robust markers.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>In this vignette, we have demonstrated a comprehensive suite of
functions designed to enhance the analysis and comparison of single-cell
genomics datasets. <code><a href="../reference/compareCCA.html">compareCCA()</a></code> and
<code><a href="../reference/comparePCA.html">comparePCA()</a></code> facilitate the evaluation of dataset alignment
through canonical correlation analysis and principal component analysis,
respectively. These tools help in assessing the correspondence between
datasets and identifying potential batch effects or differences in data
structure. <code><a href="../reference/comparePCASubspace.html">comparePCASubspace()</a></code> further refines this
analysis by focusing on specific subspaces within the PCA space,
providing a more granular view of dataset similarities.</p>
<p><code><a href="../reference/plotPairwiseDistancesDensity.html">plotPairwiseDistancesDensity()</a></code> and
<code><a href="../reference/plotWassersteinDistance.html">plotWassersteinDistance()</a></code> offer advanced visualization
techniques for comparing distances and distributions across datasets.
These functions are crucial for understanding the variability and
overlap between datasets in a more intuitive and interpretable
manner.</p>
<p>On the other hand, <code><a href="../reference/calculateHVGOverlap.html">calculateHVGOverlap()</a></code> and
<code><a href="../reference/calculateVarImpOverlap.html">calculateVarImpOverlap()</a></code> provide insights into gene
variability and importance, respectively, by comparing highly variable
genes and variable importance scores across reference and query
datasets.</p>
<p>Together, these functions form a robust toolkit for single-cell
genomics analysis, enabling researchers to conduct detailed comparisons,
visualize data differences, and validate the relevance of their findings
across different datasets. Incorporating these tools into your research
workflow will help ensure more accurate and insightful interpretations
of single-cell data, ultimately advancing the understanding of cellular
processes and improving experimental outcomes.</p>
<hr>
</div>
<div class="section level2">
<h2 id="r-session-info">R Session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h2>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scater_1.32.1               ggplot2_3.5.1              
 [3] scran_1.32.0                scuttle_1.14.0             
 [5] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0
 [7] Biobase_2.64.0              GenomicRanges_1.56.1       
 [9] GenomeInfoDb_1.40.1         IRanges_2.38.1             
[11] S4Vectors_0.42.1            BiocGenerics_0.50.0        
[13] MatrixGenerics_1.16.0       matrixStats_1.4.1          
[15] scDiagnostics_1.0.0         BiocStyle_2.32.1           

loaded via a namespace (and not attached):
 [1] gridExtra_2.3             rlang_1.1.4              
 [3] magrittr_2.0.3            compiler_4.4.1           
 [5] DelayedMatrixStats_1.26.0 systemfonts_1.1.0        
 [7] vctrs_0.6.5               pkgconfig_2.0.3          
 [9] crayon_1.5.3              fastmap_1.2.0            
[11] XVector_0.44.0            labeling_0.4.3           
[13] utf8_1.2.4                rmarkdown_2.28           
[15] UCSC.utils_1.0.0          ggbeeswarm_0.7.2         
[17] ragg_1.3.3                xfun_0.47                
[19] bluster_1.14.0            zlibbioc_1.50.0          
[21] cachem_1.1.0              beachmat_2.20.0          
[23] jsonlite_1.8.9            DelayedArray_0.30.1      
[25] BiocParallel_1.38.0       irlba_2.3.5.1            
[27] parallel_4.4.1            cluster_2.1.6            
[29] R6_2.5.1                  bslib_0.8.0              
[31] ranger_0.16.0             limma_3.60.4             
[33] jquerylib_0.1.4           Rcpp_1.0.13              
[35] bookdown_0.40             knitr_1.48               
[37] Matrix_1.7-0              igraph_2.0.3             
[39] tidyselect_1.2.1          abind_1.4-8              
[41] yaml_2.3.10               viridis_0.6.5            
[43] codetools_0.2-20          lattice_0.22-6           
[45] tibble_3.2.1              withr_3.0.1              
[47] evaluate_1.0.0            desc_1.4.3               
[49] pillar_1.9.0              BiocManager_1.30.25      
[51] generics_0.1.3            sparseMatrixStats_1.16.0 
[53] munsell_0.5.1             scales_1.3.0             
[55] glue_1.7.0                metapod_1.12.0           
[57] tools_4.4.1               data.table_1.16.0        
[59] BiocNeighbors_1.22.0      ScaledMatrix_1.12.0      
[61] locfit_1.5-9.10           fs_1.6.4                 
[63] grid_4.4.1                edgeR_4.2.1              
[65] colorspace_2.1-1          GenomeInfoDbData_1.2.12  
[67] beeswarm_0.4.0            BiocSingular_1.20.0      
[69] vipor_0.4.7               cli_3.6.3                
[71] rsvd_1.0.5                textshaping_0.4.0        
[73] fansi_1.0.6               S4Arrays_1.4.1           
[75] viridisLite_0.4.2         dplyr_1.1.4              
[77] gtable_0.3.5              sass_0.4.9               
[79] digest_0.6.37             SparseArray_1.4.8        
[81] ggrepel_0.9.6             dqrng_0.4.1              
[83] farver_2.1.2              htmlwidgets_1.6.4        
[85] htmltools_0.5.8.1         pkgdown_2.1.1            
[87] lifecycle_1.0.4           httr_1.4.7               
[89] transport_0.15-4          statmod_1.5.0            </code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
