<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Evaluation of Dataset and Marker Gene Alignment • scDiagnostics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Evaluation of Dataset and Marker Gene Alignment">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection and Analysis of Annotation Anomalies</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Evaluation of Dataset and Marker Gene
Alignment</h1>
                        <h4 data-toc-skip class="author">Anthony
Christidis</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Andrew
Ghazi</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Smriti
Chawla</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Nitesh
Turaga</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Ludwig
Geistlinger</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Robert
Gentleman</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/vignettes/DatasetMarkerGeneAlignment.Rmd" class="external-link"><code>vignettes/DatasetMarkerGeneAlignment.Rmd</code></a></small>
      <div class="hidden name"><code>DatasetMarkerGeneAlignment.Rmd</code></div>

    </address>
</address>
</address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In the realm of single-cell genomics, the ability to compare and
integrate data across different conditions, datasets, or methodologies
is crucial for deriving meaningful biological insights. This vignette
introduces several functions designed to facilitate such comparisons and
analyses by providing robust tools for evaluating and visualizing
similarities and differences in high-dimensional data.</p>
<div class="section level3">
<h3 id="functions-for-evaluation-of-dataset-alignment">Functions for Evaluation of Dataset Alignment<a class="anchor" aria-label="anchor" href="#functions-for-evaluation-of-dataset-alignment"></a>
</h3>
<ul>
<li><p><code>compareCCA()</code>: This function enables the comparison
of datasets by applying Canonical Correlation Analysis (CCA). It helps
assess how well two datasets align with each other, providing insights
into the relationship between different single-cell experiments or
conditions.</p></li>
<li><p><code><a href="../reference/comparePCA.html">comparePCA()</a></code>: This function allows you to compare
datasets using Principal Component Analysis (PCA). It evaluates how
similar or different the principal components are between two datasets,
offering a way to understand the underlying structure and variance in
your data.</p></li>
<li><p><code><a href="../reference/comparePCASubspace.html">comparePCASubspace()</a></code>: Extending the comparison to
specific subspaces, this function focuses on subsets of principal
components. It provides a detailed analysis of how subspace structures
differ or align between datasets, which is valuable for fine-grained
comparative studies.</p></li>
<li><p><code><a href="../reference/plotPairwiseDistancesDensity.html">plotPairwiseDistancesDensity()</a></code>: To visualize the
distribution of distances between pairs of samples, this function
generates density plots. It helps in understanding the variation and
relationships between samples in high-dimensional spaces.</p></li>
<li><p><code>plotWassersteinDistance()</code>: This function visualizes
the Wasserstein distance, a metric for comparing distributions, across
datasets. It provides an intuitive view of how distributions differ
between datasets, aiding in the evaluation of alignment and
discrepancies.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="statistical-measures-to-assess-dataset-alignment">Statistical Measures to Assess Dataset Alignment<a class="anchor" aria-label="anchor" href="#statistical-measures-to-assess-dataset-alignment"></a>
</h3>
<ul>
<li><p><code><a href="../reference/calculateCramerPValue.html">calculateCramerPValue()</a></code>: This function computes the
Cramér’s V statistic, which measures the strength of association between
categorical variables. Cramér’s V is particularly useful for quantifying
the association strength in contingency tables, providing insight into
how strongly different cell types or conditions are related.</p></li>
<li><p><code><a href="../reference/calculateHotellingPValue.html">calculateHotellingPValue()</a></code>: This function performs
Hotelling’s T-squared test, a multivariate statistical test used to
assess the differences between two groups of observations. It is useful
for comparing the mean vectors of multiple variables and is commonly
employed in single-cell data to evaluate group differences in principal
component space.</p></li>
<li><p><code>calculateNearestNeighborProbabilities()</code>: This
function calculates the probability of nearest neighbor assignments in a
high-dimensional space. It is used to evaluate the likelihood of a cell
belonging to a specific cluster based on its proximity to neighboring
cells. This function helps in assessing clustering quality and the
robustness of cell type assignments.</p></li>
<li><p><code><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation()</a></code>: This function
computes the average pairwise correlation between variables or features
across cells. It provides an overview of the relationships between
different markers or genes, helping to identify correlated features and
understand their interactions within the dataset.</p></li>
<li><p><code><a href="../reference/regressPC.html">regressPC()</a></code>: This function performs linear
regression of a covariate of interest onto one or more principal
components derived from single-cell data. This analysis helps quantify
the variance explained by the covariate and is useful for tasks such as
assessing batch effects, clustering homogeneity, and alignment of query
and reference datasets.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="marker-gene-alignment">Marker Gene Alignment<a class="anchor" aria-label="anchor" href="#marker-gene-alignment"></a>
</h3>
<ul>
<li><p><code><a href="../reference/calculateHVGOverlap.html">calculateHVGOverlap()</a></code>: To assess the similarity
between datasets based on highly variable genes, this function computes
the overlap coefficient. It measures how well the sets of highly
variable genes from different datasets correspond to each
other.</p></li>
<li><p><code><a href="../reference/calculateVarImpOverlap.html">calculateVarImpOverlap()</a></code>: Using Random Forest, this
function identifies and compares the importance of genes for
differentiating cell types between datasets. It highlights which genes
are most critical in each dataset and compares their importance,
providing insights into shared and unique markers.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="purpose-and-applications">Purpose and Applications<a class="anchor" aria-label="anchor" href="#purpose-and-applications"></a>
</h3>
<p>These functions collectively offer a comprehensive toolkit for
comparing and analyzing single-cell data. Whether you are assessing
alignment between datasets, visualizing distance distributions, or
identifying key genes, these tools are designed to enhance your ability
to derive meaningful insights from complex, high-dimensional data.</p>
<p>In this vignette, we will guide you through the practical use of each
function, demonstrate how to interpret their outputs, and show how they
can be integrated into your single-cell genomics workflow.</p>
</div>
</div>
<div class="section level2">
<h2 id="preliminaries">Preliminaries<a class="anchor" aria-label="anchor" href="#preliminaries"></a>
</h2>
<p>In the context of the <code>scDiagnostics</code> package, this
vignette demonstrates how to leverage various functions to evaluate and
compare single-cell data across two distinct datasets:</p>
<ul>
<li>
<code>reference_data</code>: This dataset features meticulously
curated cell type annotations assigned by experts. It serves as a robust
benchmark for evaluating the accuracy and consistency of cell type
annotations across different datasets, offering a reliable standard
against which other annotations can be assessed.</li>
<li>
<code>query_data</code>: This dataset contains cell type annotations
from both expert assessments and those generated using the <em><a href="https://bioconductor.org/packages/3.21/SingleR" class="external-link">SingleR</a></em>
package. By comparing these annotations with those from the reference
dataset, you can identify discrepancies between manual and automated
results, highlighting potential inconsistencies or areas requiring
further investigation.</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/scDiagnostics" class="external-link">scDiagnostics</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>Some functions in the vignette are designed to work with <em><a href="https://bioconductor.org/packages/3.21/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects that contain data from only one cell type. We will create
separate <em><a href="https://bioconductor.org/packages/3.21/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects that only CD4 cells, to ensure compatibility with these
functions.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset to CD4 cells</span></span>
<span><span class="va">ref_data_cd4</span> <span class="op">&lt;-</span> <span class="va">reference_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span></span>
<span>    <span class="va">reference_data</span><span class="op">$</span><span class="va">expert_annotation</span> <span class="op">==</span> <span class="st">"CD4"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="va">query_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span></span>
<span>    <span class="va">query_data</span><span class="op">$</span><span class="va">expert_annotation</span> <span class="op">==</span> <span class="st">"CD4"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Select highly variable genes</span></span>
<span><span class="va">ref_top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">ref_data_cd4</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">query_top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">query_data_cd4</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">common_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">ref_top_genes</span>, <span class="va">query_top_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset data by common genes</span></span>
<span><span class="va">ref_data_cd4</span> <span class="op">&lt;-</span> <span class="va">ref_data_cd4</span><span class="op">[</span><span class="va">common_genes</span>,<span class="op">]</span></span>
<span><span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="va">query_data_cd4</span><span class="op">[</span><span class="va">common_genes</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Run PCA on both datasets</span></span>
<span><span class="va">ref_data_cd4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">ref_data_cd4</span><span class="op">)</span></span>
<span><span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">query_data_cd4</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="evaluation-of-dataset-alignment">Evaluation of Dataset Alignment<a class="anchor" aria-label="anchor" href="#evaluation-of-dataset-alignment"></a>
</h2>
<div class="section level3">
<h3 id="comparepca">
<code>comparePCA()</code><a class="anchor" aria-label="anchor" href="#comparepca"></a>
</h3>
<p>The <code><a href="../reference/comparePCA.html">comparePCA()</a></code> function compares the PCA subspaces
between the query and reference datasets. It calculates the principal
angles between the PCA subspaces to assess the alignment and similarity
between them. This is useful for understanding how well the
dimensionality reduction spaces of your datasets match.</p>
<p><code><a href="../reference/comparePCA.html">comparePCA()</a></code> operates as follows:</p>
<ul>
<li>
<strong>PCA Computation</strong>: It computes PCA for both query and
reference datasets, reducing them into lower-dimensional spaces.</li>
<li>
<strong>Subspace Comparison</strong>: The function calculates the
principal angles between the PCA subspaces of the query and reference
datasets. These angles help determine how closely the subspaces
align.</li>
<li>
<strong>Distance Metrics</strong>: It uses distance metrics based on
principal angles to quantify the similarity between the datasets.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform PCA </span></span>
<span><span class="va">pca_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/comparePCA.html">comparePCA</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_cd4</span>, </span>
<span>                             reference_data <span class="op">=</span> <span class="va">ref_data_cd4</span>, </span>
<span>                             query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                             ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                             pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, </span>
<span>                             metric <span class="op">=</span> <span class="st">"cosine"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pca_comparison</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="comparepcasubspace">
<code>comparePCASubspace()</code><a class="anchor" aria-label="anchor" href="#comparepcasubspace"></a>
</h3>
<p>In single-cell RNA-seq analysis, it is essential to assess the
similarity between the subspaces spanned by the top principal components
(PCs) of query and reference datasets. This is particularly important
when comparing the structure and variation captured by each dataset. The
<code><a href="../reference/comparePCASubspace.html">comparePCASubspace()</a></code> function is designed to provide
insights into how well the subspaces align by computing the cosine
similarity between the loadings of the top variables for each PC. This
analysis helps in determining the degree of correspondence between the
datasets, which is critical for accurate cell type annotation and data
integration.</p>
<p><code><a href="../reference/comparePCASubspace.html">comparePCASubspace()</a></code> performs the following
operations:</p>
<ul>
<li>
<strong>Cosine Similarity Calculation</strong>: The function
computes the cosine similarity between the loadings of the top variables
for each PC in both the query and reference datasets. This similarity
measures how closely the two datasets align in the space spanned by
their principal components.</li>
<li>
<strong>Selection of Top Similarities</strong>: The function selects
the top cosine similarity scores and stores the corresponding PC indices
from both datasets. This step identifies the most aligned principal
components between the two datasets.</li>
<li>
<strong>Variance Explained Calculation</strong>: It then calculates
the average percentage of variance explained by the selected top PCs.
This helps in understanding how much of the data’s variance is captured
by these components.</li>
<li>
<strong>Weighted Cosine Similarity Score</strong>: Finally, the
function computes a weighted cosine similarity score based on the top
cosine similarities and the average percentage of variance explained.
This score provides an overall measure of subspace alignment between the
datasets.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compare PCA subspaces between query and reference data</span></span>
<span><span class="va">subspace_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/comparePCASubspace.html">comparePCASubspace</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data_cd4</span>,</span>
<span>    reference_data <span class="op">=</span> <span class="va">ref_data_cd4</span>, </span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View weighted cosine similarity score</span></span>
<span><span class="va">subspace_comparison</span><span class="op">$</span><span class="va">weighted_cosine_similarity</span></span>
<span><span class="co">#&gt; [1] 0.2620419</span></span>
<span></span>
<span><span class="co"># Plot output for PCA subspace comparison (if a plot method is available)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">subspace_comparison</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/DatasetMarkerGeneAlignment/comparePCASubspace.png"></p>
<p>In the results:</p>
<ul>
<li>
<strong>Cosine Similarity</strong>: The values in
principal_angles_cosines indicate the degree of alignment between the
principal components of the query and reference datasets. Higher values
suggest stronger alignment.</li>
<li>
<strong>Variance Explained</strong>: The average_variance_explained
vector provides the average percentage of variance captured by the
selected top PCs in both datasets. This helps assess the importance of
these PCs in explaining data variation.</li>
<li>
<strong>Weighted Cosine Similarity</strong>: The
weighted_cosine_similarity score combines the cosine similarity with
variance explained to give a comprehensive measure of how well the
subspaces align. A higher score indicates that the datasets are
well-aligned in the PCA space.</li>
</ul>
<p>By using <code><a href="../reference/comparePCASubspace.html">comparePCASubspace()</a></code>, you can quantify the
alignment of PCA subspaces between query and reference datasets, aiding
in the evaluation of dataset integration and the reliability of cell
type annotations.</p>
</div>
<div class="section level3">
<h3 id="plotpairwisedistancesdensity">
<code>plotPairwiseDistancesDensity()</code><a class="anchor" aria-label="anchor" href="#plotpairwisedistancesdensity"></a>
</h3>
<div class="section level4">
<h4 id="purpose">Purpose<a class="anchor" aria-label="anchor" href="#purpose"></a>
</h4>
<p>The <code><a href="../reference/plotPairwiseDistancesDensity.html">plotPairwiseDistancesDensity()</a></code> function is designed
to calculate and visualize the pairwise distances or correlations
between cell types in query and reference datasets. This function is
particularly useful in single-cell RNA sequencing (scRNA-seq) analysis,
where it is essential to evaluate the consistency and reliability of
cell type annotations by comparing their expression profiles.</p>
</div>
<div class="section level4">
<h4 id="functionality">Functionality<a class="anchor" aria-label="anchor" href="#functionality"></a>
</h4>
<p>The function operates on <em><a href="https://bioconductor.org/packages/3.21/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects, which are commonly used to store single-cell data, including
expression matrices and associated metadata. Users specify the cell
types of interest in both the query and reference datasets, and the
function computes either the distances or correlation coefficients
between these cells.</p>
<p>When principal component analysis (PCA) is applied, the function
projects the expression data into a lower-dimensional PCA space, which
can be specified by the user. This allows for a more focused analysis of
the major sources of variation in the data. Alternatively, if no
dimensionality reduction is desired, the function can directly use the
expression data for computation.</p>
<p>Depending on the user’s choice, the function can calculate pairwise
Euclidean distances or correlation coefficients. The resulting values
are used to compare the relationships between cells within the same
dataset (either query or reference) and between cells across the two
datasets.</p>
</div>
<div class="section level4">
<h4 id="interpretation">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation"></a>
</h4>
<p>The output of the function is a density plot generated by
<code>ggplot2</code>, which displays the distribution of pairwise
distances or correlations. The plot provides three key comparisons:</p>
<ul>
<li>Query vs. Query,</li>
<li>Reference vs. Reference,</li>
<li>Query vs. Reference.</li>
</ul>
<p>By examining these density curves, users can assess the similarity of
cells within each dataset and across datasets. For example, a higher
density of lower distances in the “Query vs. Reference” comparison would
suggest that the query and reference cells are similar in their
expression profiles, indicating consistency in the annotation of the
cell type across the datasets.</p>
<p>This visual approach offers an intuitive way to diagnose potential
discrepancies in cell type annotations, identify outliers, or confirm
the reliability of the cell type assignments.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example usage of the function</span></span>
<span><span class="fu"><a href="../reference/plotPairwiseDistancesDensity.html">plotPairwiseDistancesDensity</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                             reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                             query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                             ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                             cell_type_query <span class="op">=</span> <span class="st">"CD4"</span>, </span>
<span>                             cell_type_ref <span class="op">=</span> <span class="st">"CD4"</span>, </span>
<span>                             pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>                             distance_metric <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/DatasetMarkerGeneAlignment/plotPairwiseDistancesDensity.png"></p>
<p>This example demonstrates how to compare CD4 cells between a query
and reference dataset, with PCA applied to the first five principal
components and pairwise Euclidean distances calculated. The output is a
density plot that helps visualize the distribution of these distances,
aiding in the interpretation of the similarity between the two
datasets.</p>
</div>
</div>
<div class="section level3">
<h3 id="calculatewassersteindistance">
<code>calculateWassersteinDistance()</code><a class="anchor" aria-label="anchor" href="#calculatewassersteindistance"></a>
</h3>
<p>the <code><a href="../reference/calculateWassersteinDistance.html">calculateWassersteinDistance()</a></code> function uses the
concept of Wasserstein distances, a measure rooted in optimal transport
theory, to quantitatively compare these datasets. By projecting both
datasets into a common PCA space, the function enables a meaningful
comparison even when they originate from different experimental
conditions.</p>
<p>Before diving into a code example, it’s important to understand the
output of this function. The result is a list containing three key
components: the null distribution of Wasserstein distances computed from
the reference dataset alone, the mean Wasserstein distance between the
query and reference datasets, and a vector of the unique cell types
present in the reference dataset. This comparison provides insights into
how similar or different the query dataset is relative to the reference
dataset.</p>
<p>What the function reveals is particularly valuable: if the mean
Wasserstein distance between the query and reference is significantly
different from the null distribution, it indicates a notable variation
between the datasets, reflecting differences in cell-type compositions,
technical artifacts, or biological questions of interest.</p>
<div class="section level4">
<h4 id="code-example">Code Example<a class="anchor" aria-label="anchor" href="#code-example"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate the Wasserstein distance density plot</span></span>
<span><span class="va">wasserstein_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateWassersteinDistance.html">calculateWassersteinDistance</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data_cd4</span>,</span>
<span>    reference_data <span class="op">=</span> <span class="va">ref_data_cd4</span>, </span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">wasserstein_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0142</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/DatasetMarkerGeneAlignment/plotWassersteinDistance.png"></p>
<p>The plotting function generates a density plot that provides a clear
graphical representation of the Wasserstein distances. On the plot,
you’ll see the distribution of Wasserstein distances calculated within
the reference dataset, referred to as the null distribution.</p>
<p>Within this density plot, two significant vertical lines are
overlayed: one representing the significance threshold and another
indicating the mean Wasserstein distance between the reference and query
datasets. The significance threshold is determined based on your
specified alpha level, which is the probability threshold for
considering differences between the datasets as statistically
significant. This threshold line helps to establish a boundary; if the
observed reference-query distance exceeds this threshold, it suggests
that the differences between your query and reference datasets are
unlikely to occur by random chance.</p>
<p>The line for the reference-query distance helps identify where this
calculated distance lies in comparison to the null distribution. If this
dashed line is noticeably separate from the bulk of the null
distribution and beyond the significance threshold, it implies a
significant deviation between the query dataset and the reference
dataset. In practical terms, this could indicate distinct biological
differences, batch effects, or other analytical artefacts that
differentiate the datasets.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="statistical-measures-to-assess-dataset-alignment-1">Statistical Measures to Assess Dataset Alignment<a class="anchor" aria-label="anchor" href="#statistical-measures-to-assess-dataset-alignment-1"></a>
</h2>
<div class="section level3">
<h3 id="calculatecramerpvalue">
<code>calculateCramerPValue()</code><a class="anchor" aria-label="anchor" href="#calculatecramerpvalue"></a>
</h3>
<p>The calculateCramerPValue function is designed to perform the Cramer
test for comparing multivariate empirical cumulative distribution
functions (ECDFs) between two samples in single-cell data. This test is
particularly useful for assessing whether the distributions of principal
components (PCs) differ significantly between the reference and query
datasets for specific cell types.</p>
<p>To use this function, you first need to provide two key inputs:
<code>reference_data</code> and <code>query_data</code>, both of which
should be <em><a href="https://bioconductor.org/packages/3.21/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects containing numeric expression matrices. If
<code>query_data</code> is not supplied, the function will only use the
<code>reference_data.</code> You should also specify the column names
for cell type annotations in both datasets via
<code>ref_cell_type_col</code> and <code>query_cell_type_col.</code> If
<code>cell_types</code> is not provided, the function will automatically
include all unique cell types found in the datasets. The
<code>pc_subset</code> parameter allows you to define which principal
components to include in the analysis, with the default being the first
five PCs.</p>
<p>The function performs the following steps: it first projects the data
into PCA space, subsets the data according to the specified cell types
and principal components, and then applies the Cramer test to compare
the ECDFs between the reference and query datasets. The result is a
named vector of p-values from the Cramer test for each cell type, which
indicates whether there is a significant difference in the distributions
of PCs between the two datasets.</p>
<p>Here’s an example of how to use the
<code><a href="../reference/calculateCramerPValue.html">calculateCramerPValue()</a></code> function:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform Cramer test for the specified cell types and principal components</span></span>
<span><span class="va">cramer_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateCramerPValue.html">calculateCramerPValue</a></span><span class="op">(</span></span>
<span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>    cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Display the Cramer test p-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">cramer_test</span><span class="op">)</span></span>
<span><span class="co">#&gt;          CD4          CD8 B_and_plasma      Myeloid </span></span>
<span><span class="co">#&gt;   0.08491508   0.00000000   0.13186813   0.92207792</span></span></code></pre></div>
<p>In this example, the function compares the distributions of the first
five principal components between the reference and query datasets for
cell types such as “CD4”, “CD8”, “B_and_plasma”, and “Myeloid”. The
output is a vector of p-values indicating whether the differences in PC
distributions are statistically significant.</p>
</div>
<div class="section level3">
<h3 id="calculatehotellingpvalue">
<code>calculateHotellingPValue()</code><a class="anchor" aria-label="anchor" href="#calculatehotellingpvalue"></a>
</h3>
<p>The <code><a href="../reference/calculateHotellingPValue.html">calculateHotellingPValue()</a></code> function is designed to
compute Hotelling’s T-squared test statistic and corresponding p-values
for comparing multivariate means between reference and query datasets in
the context of single-cell RNA-seq data. This statistical test is
particularly useful for assessing whether the mean vectors of principal
components (PCs) differ significantly between the two datasets, which
can be indicative of differences in the cell type distributions.</p>
<p>To use this function, you need to provide two <em><a href="https://bioconductor.org/packages/3.21/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects: <code>query_data</code> and <code>reference_data</code>, each
containing numeric expression matrices. You also need to specify the
column names for cell type annotations in both datasets with
<code>query_cell_type_col</code> and <code>ref_cell_type_col.</code> The
<code>cell_types</code> parameter allows you to choose which cell types
to include in the analysis, and if not specified, the function will
automatically include all cell types present in the datasets. The
<code>pc_subset</code> parameter determines which principal components
to consider, with the default being the first five PCs. Additionally,
<code>n_permutation</code> specifies the number of permutations for
calculating p-values, with a default of 500.</p>
<p>The function works by first projecting the data into PCA space and
then performing Hotelling’s T-squared test for each specified cell type
to compare the means between the reference and query datasets. It uses
permutation testing to determine the p-values, indicating whether the
observed differences in means are statistically significant. The result
is a named numeric vector of p-values for each cell type.</p>
<p>Here is an example of how to use the
<code><a href="../reference/calculateHotellingPValue.html">calculateHotellingPValue()</a></code> function:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate Hotelling's T-squared test p-values</span></span>
<span><span class="va">p_values</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHotellingPValue.html">calculateHotellingPValue</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Display the p-values rounded to five decimal places</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">p_values</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;          CD4          CD8 B_and_plasma      Myeloid </span></span>
<span><span class="co">#&gt;        0.000        0.000        0.304        0.330</span></span></code></pre></div>
<p>In this example, the function calculates p-values for Hotelling’s
T-squared test, focusing on the first ten principal components to assess
whether the multivariate means differ significantly between the
reference and query datasets for each cell type. The p-values indicate
the likelihood of observing the differences by chance and help identify
significant disparities in cell type distributions.</p>
</div>
<div class="section level3">
<h3 id="calculateaveragepairwisecorrelation">
<code>calculateAveragePairwiseCorrelation()</code><a class="anchor" aria-label="anchor" href="#calculateaveragepairwisecorrelation"></a>
</h3>
<p>The <code><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation()</a></code> function is
designed to compute the average pairwise correlations between specified
cell types in single-cell gene expression data. This function operates
on <em><a href="https://bioconductor.org/packages/3.21/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects and is ideal for single-cell analysis workflows. It calculates
pairwise correlations between query and reference cells using a
specified correlation method, and then averages these correlations for
each cell type pair. This helps in assessing the similarity between
cells in the reference and query datasets and provides insights into the
reliability of cell type annotations.</p>
<p>To use the <code><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation()</a></code>
function, you need to supply it with two <em><a href="https://bioconductor.org/packages/3.21/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects: one for the query cells and one for the reference cells. The
function also requires column names specifying cell type annotations in
both datasets, and optionally a vector of cell types to include in the
analysis. Additionally, you can specify a subset of principal components
to use, or use the raw data directly if <code>pc_subset</code> is set to
<code>NULL</code>. The correlation method can be either “spearman” or
“pearson”.</p>
<p>Here’s an example of how to use this function:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute pairwise correlations between specified cell types</span></span>
<span><span class="va">cor_matrix_avg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation</a></span><span class="op">(</span></span>
<span>  query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>  reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>  query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>  ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>  cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span><span class="op">)</span>, </span>
<span>  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>  correlation_method <span class="op">=</span> <span class="st">"spearman"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize the average pairwise correlation matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cor_matrix_avg</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/DatasetMarkerGeneAlignment/calculateAveragePairwiseCorrelation.png"></p>
<p>In this example, <code><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation()</a></code>
computes the average pairwise correlations for the cell types “CD4”,
“CD8”, and “B_and_plasma”. This function is particularly useful for
understanding the relationships between different cell types in your
single-cell datasets and evaluating how well cell types in the query
data correspond to those in the reference data.</p>
</div>
<div class="section level3">
<h3 id="regresspc">
<code>regressPC()</code><a class="anchor" aria-label="anchor" href="#regresspc"></a>
</h3>
<p>The <code><a href="../reference/regressPC.html">regressPC()</a></code> function performs linear regression of a
covariate of interest onto one or more principal components using data
from a <em><a href="https://bioconductor.org/packages/3.21/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
object. This method helps quantify the variance explained by a
covariate, which can be useful in applications such as quantifying batch
effects, assessing clustering homogeneity, and evaluating alignment
between query and reference datasets in cell type annotation
settings.</p>
<p>The function supports multiple regression scenarios depending on the
data provided:</p>
<ul>
<li>
<strong>Query only</strong>: Regresses principal components against
cell types (<code>PC ~ cell_type</code>)</li>
<li>
<strong>Query + Reference</strong>: Uses an interaction model to
assess annotation quality (<code>PC ~ cell_type * dataset</code>)</li>
</ul>
<p>When reference data is provided, the function projects query data
onto the reference PCA space and performs unified interaction modeling
to identify cell types with annotation inconsistencies between
datasets.</p>
<p>The function calculates the R-squared value from the linear
regression for each principal component. For query-only analyses,
variance contributions are computed as the product of the variance
explained by each principal component and its corresponding R-squared
value. The total variance explained by the covariate is obtained by
summing these contributions across all principal components.</p>
<p>Here is how you can use the <code><a href="../reference/regressPC.html">regressPC()</a></code> function:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Query-only analysis: How well do cell types separate in PCA space?</span></span>
<span><span class="va">regress_res_query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regressPC.html">regressPC</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>    cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot R-squared values showing cell type separation quality</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res_query</span>, plot_type <span class="op">=</span> <span class="st">"r_squared"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/DatasetMarkerGeneAlignment/regressPC1.png"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot results showing p-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res_query</span>, plot_type <span class="op">=</span> <span class="st">"heatmap"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/DatasetMarkerGeneAlignment/regressPC2.png"></p>
<p>In this example, <code><a href="../reference/regressPC.html">regressPC()</a></code> is used to perform
regression analysis on principal components 1 to 10 from the query
dataset. The results are then visualized using plots showing the
R-squared values and cell type coefficients, with PC labels displaying
the percentage of variance explained by each component.</p>
<p>If you also have a reference dataset and want to assess annotation
quality by comparing it with the query dataset, you can include it in
the analysis:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform regression analysis using both reference and query data</span></span>
<span><span class="va">regress_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regressPC.html">regressPC</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>    cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot overall model fit with variance percentages on PC labels</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"r_squared"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/DatasetMarkerGeneAlignment/regressPC3.png"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot results showing p-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"heatmap"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/DatasetMarkerGeneAlignment/regressPC4.png"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot results showing p-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"heatmap"</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/regressPC.html">regressPC()</a></code> function also supports batch effect
analysis when sample or batch information is available. This capability
is particularly useful for assessing whether cell type annotations are
consistent across different batches or samples, and for identifying
batch-specific annotation problems.</p>
<p>When batch information is provided via the
<code>query_batch_col</code> parameter, the function performs different
regression analyses depending on the data configuration:</p>
<div class="section level4">
<h4 id="query-only-with-batch-information">Query-only with Batch Information<a class="anchor" aria-label="anchor" href="#query-only-with-batch-information"></a>
</h4>
<p>For query data with multiple batches, the function fits an
interaction model:</p>
<p><strong>PC ~ cell_type × batch</strong></p>
<p>This model captures: - <strong>Main effects of cell types</strong>:
How each cell type contributes to PC variation overall - <strong>Main
effects of batches</strong>: How each batch contributes to PC variation
overall<br>
- <strong>Interaction effects</strong>: How cell type effects vary
across different batches</p>
<p>Large interaction effects indicate that certain cell types behave
differently across batches, which could suggest batch-specific technical
artifacts or genuine biological differences between samples.</p>
</div>
<div class="section level4">
<h4 id="query-reference-with-batch-information">Query + Reference with Batch Information<a class="anchor" aria-label="anchor" href="#query-reference-with-batch-information"></a>
</h4>
<p>When both reference and query data are provided along with batch
information, the analysis becomes more sophisticated. The function
creates a unified batch labeling system where: - Reference data is
labeled as “Reference” - Query batches are labeled as “Query_BatchName”
(e.g., “Query_Sample1”, “Query_Sample2”)</p>
<p>The regression model then becomes:</p>
<p><strong>PC ~ cell_type × batch</strong></p>
<p>where the batch variable now includes both the reference and all
query batches. This allows assessment of: - How consistently each cell
type is annotated between reference and query datasets - Whether
annotation quality varies across different query batches - Which
specific query batches show the most deviation from reference
annotations</p>
</div>
<div class="section level4">
<h4 id="diagnostic-value">Diagnostic Value<a class="anchor" aria-label="anchor" href="#diagnostic-value"></a>
</h4>
<p>The interaction effects in batch analysis are particularly
diagnostic: - <strong>Small interactions</strong>: Consistent cell type
annotations across batches - <strong>Large interactions for specific
cell types</strong>: Potential annotation problems in particular batches
- <strong>Large interactions for specific batches</strong>: Systematic
annotation issues in those batches</p>
<p>This analysis framework enables identification of batch-specific
annotation problems and helps determine whether observed differences
between datasets are due to technical batch effects or genuine
biological variation.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="marker-gene-alignment-1">Marker Gene Alignment<a class="anchor" aria-label="anchor" href="#marker-gene-alignment-1"></a>
</h2>
<div class="section level3">
<h3 id="calculatehvgoverlap">
<code>calculateHVGOverlap()</code><a class="anchor" aria-label="anchor" href="#calculatehvgoverlap"></a>
</h3>
<p>The <code><a href="../reference/calculateHVGOverlap.html">calculateHVGOverlap()</a></code> function computes the overlap
coefficient between two sets of highly variable genes (HVGs) from a
reference dataset and a query dataset. The overlap coefficient is a
measure of similarity between the two sets, reflecting how much the HVGs
in one dataset overlap with those in the other.</p>
<div class="section level4">
<h4 id="how-the-function-operates">How the Function Operates<a class="anchor" aria-label="anchor" href="#how-the-function-operates"></a>
</h4>
<p>The function begins by ensuring that the input vectors
<code>reference_genes</code> and <code>query_genes</code> are character
vectors and that neither of them is empty. Once these checks are
complete, the function identifies the common genes between the two sets
using the intersect function, which finds the intersection of the two
gene sets.</p>
<p>Next, the function calculates the size of this intersection,
representing the number of genes common to both sets. The overlap
coefficient is then computed by dividing the size of the intersection by
the size of the smaller set of genes. This ensures that the coefficient
is a value between 0 and 1, where 0 indicates no overlap and 1 indicates
complete overlap.</p>
<p>Finally, the function rounds the overlap coefficient to two decimal
places before returning it as the output.</p>
</div>
<div class="section level4">
<h4 id="interpretation-1">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation-1"></a>
</h4>
<p>The overlap coefficient quantifies the extent to which the HVGs in
the reference dataset align with those in the query dataset. A higher
coefficient indicates a stronger similarity between the two datasets in
terms of their HVGs, which can suggest that the datasets are more
comparable or that they capture similar biological variability.
Conversely, a lower coefficient indicates less overlap, suggesting that
the datasets may be capturing different biological signals or that they
are less comparable.</p>
</div>
<div class="section level4">
<h4 id="code-example-1">Code Example<a class="anchor" aria-label="anchor" href="#code-example-1"></a>
</h4>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load library to get top HVGs</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select the top 500 highly variable genes from both datasets</span></span>
<span><span class="va">ref_var_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">reference_data</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">query_var_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">query_data</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate the overlap coefficient between the reference and query HVGs</span></span>
<span><span class="va">overlap_coefficient</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVGOverlap.html">calculateHVGOverlap</a></span><span class="op">(</span>reference_genes <span class="op">=</span> <span class="va">ref_var_genes</span>, </span>
<span>                                           query_genes <span class="op">=</span> <span class="va">query_var_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Display the overlap coefficient</span></span>
<span><span class="va">overlap_coefficient</span></span>
<span><span class="co">#&gt; [1] 0.93</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="calculatevarimpoverlap">
<code>calculateVarImpOverlap()</code><a class="anchor" aria-label="anchor" href="#calculatevarimpoverlap"></a>
</h3>
<div class="section level4">
<h4 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h4>
<p>The <code><a href="../reference/calculateVarImpOverlap.html">calculateVarImpOverlap()</a></code> function helps you identify
and compare the most important genes for distinguishing cell types in
both a reference dataset and a query dataset. It does this using the
Random Forest algorithm, which calculates how important each gene is in
differentiating between cell types.</p>
</div>
<div class="section level4">
<h4 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h4>
<p>To use the function, you need to provide a reference dataset
containing expression data and cell type annotations. Optionally, you
can also provide a query dataset if you want to compare gene importance
across both datasets. The function allows you to specify which cell
types to analyze and how many trees to use in the Random Forest model.
Additionally, you can decide how many top genes you want to compare
between the datasets.</p>
</div>
<div class="section level4">
<h4 id="code-example-2">Code Example<a class="anchor" aria-label="anchor" href="#code-example-2"></a>
</h4>
<p>Let’s say you have a reference dataset (<code>reference_data</code>)
and a query dataset (<code>query_data</code>). Both datasets contain
expression data and cell type annotations, stored in columns named
“expert_annotation” and <code>SingleR_annotation</code>, respectively.
You want to calculate the importance of genes using 500 trees and
compare the top 50 genes between the datasets.</p>
<p>Here’s how you would use the function:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># RF function to compare which genes are best at differentiating cell types</span></span>
<span><span class="va">rf_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateVarImpOverlap.html">calculateVarImpOverlap</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                    query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                    query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                    n_tree <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                    n_top <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Comparison table</span></span>
<span><span class="va">rf_output</span><span class="op">$</span><span class="va">var_imp_comparison</span></span>
<span><span class="co">#&gt;              CD4-CD8     CD4-B_and_plasma          CD4-Myeloid </span></span>
<span><span class="co">#&gt;                 0.84                 0.84                 0.86 </span></span>
<span><span class="co">#&gt;     CD8-B_and_plasma          CD8-Myeloid B_and_plasma-Myeloid </span></span>
<span><span class="co">#&gt;                 0.86                 0.84                 0.78</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="interpretation-2">Interpretation:<a class="anchor" aria-label="anchor" href="#interpretation-2"></a>
</h4>
<p>After running the function, you’ll receive the importance scores of
genes for each pair of cell types in your reference dataset. If you
provided a query dataset, you’ll also get the importance scores for
those cell types. The function will tell you how much the top genes in
the reference and query datasets overlap, which helps you understand if
the same genes are important for distinguishing cell types across
different datasets.</p>
<p>For example, if there’s a high overlap, it suggests that similar
genes are crucial in both datasets for differentiating the cell types,
which could be important for validating your findings or identifying
robust markers.</p>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="r-session-info">R Session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h2>
<pre><code>R version 4.5.1 (2025-06-13)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scater_1.36.0               ggplot2_3.5.2              
 [3] scran_1.36.0                scuttle_1.18.0             
 [5] SingleCellExperiment_1.30.1 SummarizedExperiment_1.38.1
 [7] Biobase_2.68.0              GenomicRanges_1.60.0       
 [9] GenomeInfoDb_1.44.0         IRanges_2.42.0             
[11] S4Vectors_0.46.0            BiocGenerics_0.54.0        
[13] generics_0.1.4              MatrixGenerics_1.20.0      
[15] matrixStats_1.5.0           scDiagnostics_1.1.0        
[17] BiocStyle_2.36.0           

loaded via a namespace (and not attached):
 [1] DBI_1.2.3               gridExtra_2.3           cramer_0.9-4           
 [4] rlang_1.1.6             magrittr_2.0.3          ggridges_0.5.6         
 [7] compiler_4.5.1          systemfonts_1.2.3       vctrs_0.6.5            
[10] stringr_1.5.1           pkgconfig_2.0.3         crayon_1.5.3           
[13] fastmap_1.2.0           XVector_0.48.0          labeling_0.4.3         
[16] rmarkdown_2.29          UCSC.utils_1.4.0        ggbeeswarm_0.7.2       
[19] ragg_1.4.0              xfun_0.52               bluster_1.18.0         
[22] cachem_1.1.0            beachmat_2.24.0         jsonlite_2.0.0         
[25] DelayedArray_0.34.1     BiocParallel_1.42.1     irlba_2.3.5.1          
[28] parallel_4.5.1          cluster_2.1.8.1         biglm_0.9-3            
[31] R6_2.6.1                bslib_0.9.0             stringi_1.8.7          
[34] RColorBrewer_1.1-3      ranger_0.17.0           limma_3.64.1           
[37] boot_1.3-31             jquerylib_0.1.4         Rcpp_1.1.0             
[40] bookdown_0.43           knitr_1.50              Matrix_1.7-3           
[43] igraph_2.1.4            tidyselect_1.2.1        abind_1.4-8            
[46] yaml_2.3.10             viridis_0.6.5           codetools_0.2-20       
[49] lattice_0.22-7          tibble_3.3.0            withr_3.0.2            
[52] evaluate_1.0.4          desc_1.4.3              pillar_1.11.0          
[55] BiocManager_1.30.26     scales_1.4.0            glue_1.8.0             
[58] metapod_1.16.0          tools_4.5.1             speedglm_0.3-5         
[61] BiocNeighbors_2.2.0     data.table_1.17.8       ScaledMatrix_1.16.0    
[64] locfit_1.5-9.12         fs_1.6.6                grid_4.5.1             
[67] edgeR_4.6.3             GenomeInfoDbData_1.2.14 beeswarm_0.4.0         
[70] BiocSingular_1.24.0     vipor_0.4.7             cli_3.6.5              
[73] rsvd_1.0.5              textshaping_1.0.1       S4Arrays_1.8.1         
[76] viridisLite_0.4.2       dplyr_1.1.4             gtable_0.3.6           
[79] sass_0.4.10             digest_0.6.37           SparseArray_1.8.0      
[82] ggrepel_0.9.6           dqrng_0.4.1             htmlwidgets_1.6.4      
[85] farver_2.1.2            htmltools_0.5.8.1       pkgdown_2.1.3          
[88] lifecycle_1.0.4         httr_1.4.7              statmod_1.5.0          
[91] transport_0.15-4        MASS_7.3-65            </code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
