<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Visualization of QC and Annotation Scores • scDiagnostics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Visualization of QC and Annotation Scores">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/QCandAnnotationScores.html">Visualization of QC and Annotation Scores</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/StatisticalMeasures.html">Statistical Measures to Assess Dataset Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection of Annotation Anomalies</a>
    </li>
    <li>
      <a href="../articles/CellDistancesDiagnostics.html">Calculation of Distances Between Specific Cells and Cell Populations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Visualization of QC and Annotation Scores</h1>
                        <h4 data-toc-skip class="author">Anthony
Christidis</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Andrew
Ghazi</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Smriti
Chawla</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Nitesh
Turaga</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Ludwig
Geistlinger</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Robert
Gentleman</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/vignettes/QCandAnnotationScores.Rmd" class="external-link"><code>vignettes/QCandAnnotationScores.Rmd</code></a></small>
      <div class="hidden name"><code>QCandAnnotationScores.Rmd</code></div>

    </address>
</address>
</address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this vignette, we will demonstrate how to use functions for
visualizing various aspects of single-cell RNA sequencing data. We will
cover:</p>
<ul>
<li><strong>Scatter Plot of QC Stats vs Cell Type Annotation
Scores</strong></li>
<li><strong>Histograms of QC Stats and Annotation Scores</strong></li>
<li><strong>Gene Set Scores on Dimensional Reduction Plots</strong></li>
</ul>
</div>
<div class="section level2">
<h2 id="datasets">Datasets<a class="anchor" aria-label="anchor" href="#datasets"></a>
</h2>
<p>We will use two datasets in this vignette: <code>query_data</code>,
and <code>qc_data.</code> Each dataset has been preprocessed to include
log-normalized counts, specific metadata columns, PCA, t-SNE, and UMAP
results.</p>
<div class="section level3">
<h3 id="query-data">Query Data<a class="anchor" aria-label="anchor" href="#query-data"></a>
</h3>
<p>The <code>query_data</code> is another dataset from HeOrganAtlas, but
it serves as a query set for comparison. It contains:</p>
<ul>
<li>
<strong>Log-Normalized Counts and Metadata</strong>: Similar to the
reference data.</li>
<li>
<strong>SingleR Annotations</strong>: Predictions of cell types
using <em><a href="https://bioconductor.org/packages/3.19/SingleR" class="external-link">SingleR</a></em>,
including <code>SingleR_annotation</code> and
<code>annotation_scores.</code>
</li>
<li>
<strong>Gene Set Scores</strong>: Scores calculated for gene sets or
pathways, providing insights into gene activities across cells.</li>
</ul>
</div>
<div class="section level3">
<h3 id="qc-data">QC Data<a class="anchor" aria-label="anchor" href="#qc-data"></a>
</h3>
<p>The <code>qc_data</code> originates from the Bunis et al. study
focusing on haematopoietic stem and progenitor cells. This dataset
includes:</p>
<ul>
<li>
<strong>QC Metrics</strong>: Metrics such as total library size and
mitochondrial gene content.</li>
<li>
<strong><em><a href="https://bioconductor.org/packages/3.19/SingleR" class="external-link">SingleR</a></em>
Predictions</strong>: Predicted cell types based on <em><a href="https://bioconductor.org/packages/3.19/SingleR" class="external-link">SingleR</a></em>
annotations.</li>
<li>
<strong>Annotation Scores</strong>: Scores reflecting the confidence
in cell type predictions.</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/scDiagnostics" class="external-link">scDiagnostics</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"qc_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="qc-scores">QC Scores<a class="anchor" aria-label="anchor" href="#qc-scores"></a>
</h2>
<div class="section level3">
<h3 id="scatter-plot-qc-stats-vs-cell-type-annotation-scores">Scatter Plot: QC Stats vs Cell Type Annotation Scores<a class="anchor" aria-label="anchor" href="#scatter-plot-qc-stats-vs-cell-type-annotation-scores"></a>
</h3>
<p>The scatter plot visualizes the relationship between QC statistics
(e.g., total library size or percentage of mitochondrial genes) and cell
type annotation scores. This plot helps in understanding how QC metrics
influence or correlate with the predicted cell types.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate scatter plot</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotQCvsAnnotation.html">plotQCvsAnnotation</a></span><span class="op">(</span>se_object <span class="op">=</span> <span class="va">qc_data</span>,</span>
<span>                         cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>                         qc_col <span class="op">=</span> <span class="st">"total"</span>,</span>
<span>                         score_col <span class="op">=</span> <span class="st">"annotation_scores"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Library Size"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/QCandAnnotationScores/plotQCvsAnnotation.png"></p>
<p>A scatter plot can reveal patterns such as whether cells with higher
library sizes or mitochondrial content tend to be associated with
specific annotations. For instance, cells with unusually high
mitochondrial content might be identified as low-quality or stressed,
potentially affecting their annotations.</p>
</div>
<div class="section level3">
<h3 id="histograms-qc-stats-and-annotation-scores-visualization">Histograms: QC Stats and Annotation Scores Visualization<a class="anchor" aria-label="anchor" href="#histograms-qc-stats-and-annotation-scores-visualization"></a>
</h3>
<p>Histograms provide a distribution view of QC metrics and annotation
scores. They help in evaluating the range, central tendency, and spread
of these variables across cells.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate histograms</span></span>
<span><span class="fu"><a href="../reference/histQCvsAnnotation.html">histQCvsAnnotation</a></span><span class="op">(</span>se_object <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                   cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                   qc_col <span class="op">=</span> <span class="st">"percent_mito"</span>, </span>
<span>                   score_col <span class="op">=</span> <span class="st">"annotation_scores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/QCandAnnotationScores/histQCvsAnnotation.png"></p>
<p>Histograms are useful for assessing the overall distribution of QC
metrics and annotation scores. For example, if the majority of cells
have high percent_mito, it might indicate that many cells are stressed
or dying, which could impact the quality of the data.</p>
</div>
</div>
<div class="section level2">
<h2 id="visualization-of-gene-sets-or-pathway-scores-on-dimensional-reduction-plots">Visualization of Gene Sets or Pathway Scores on Dimensional
Reduction Plots<a class="anchor" aria-label="anchor" href="#visualization-of-gene-sets-or-pathway-scores-on-dimensional-reduction-plots"></a>
</h2>
<p>Dimensional reduction plots (PCA, t-SNE, UMAP) are used to visualize
the relationships between cells in reduced dimensions. Overlaying gene
set scores on these plots provides insights into how specific gene
activities are distributed across cell clusters.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot gene set scores on PCA</span></span>
<span><span class="fu"><a href="../reference/plotGeneSetScores.html">plotGeneSetScores</a></span><span class="op">(</span>se_object <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                  method <span class="op">=</span> <span class="st">"PCA"</span>, </span>
<span>                  score_col <span class="op">=</span> <span class="st">"gene_set_scores"</span>, </span>
<span>                  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/QCandAnnotationScores/plotGeneSetScores.png"></p>
<p>By visualizing gene set scores on PCA or UMAP plots, one can identify
clusters of cells with high or low gene set activities. This can help in
understanding the biological relevance of different gene sets or
pathways in various cell states or types.</p>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>This vignette illustrates how to visualize and interpret QC
statistics, cell type annotation scores, and gene set scores using
single-cell RNA sequencing data. These visualizations are crucial for
assessing data quality, understanding cell type annotations, and
exploring gene activities, ultimately aiding in the comprehensive
analysis of single-cell datasets.</p>
<hr>
</div>
<div class="section level2">
<h2 id="r-session-info">R Session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h2>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] scDiagnostics_0.99.10 BiocStyle_2.32.1     

loaded via a namespace (and not attached):
 [1] SummarizedExperiment_1.34.0 gtable_0.3.5               
 [3] xfun_0.47                   bslib_0.8.0                
 [5] ggplot2_3.5.1               htmlwidgets_1.6.4          
 [7] Biobase_2.64.0              lattice_0.22-6             
 [9] vctrs_0.6.5                 tools_4.4.1                
[11] generics_0.1.3              stats4_4.4.1               
[13] tibble_3.2.1                fansi_1.0.6                
[15] pkgconfig_2.0.3             Matrix_1.7-0               
[17] desc_1.4.3                  S4Vectors_0.42.1           
[19] lifecycle_1.0.4             GenomeInfoDbData_1.2.12    
[21] farver_2.1.2                compiler_4.4.1             
[23] textshaping_0.4.0           munsell_0.5.1              
[25] GenomeInfoDb_1.40.1         htmltools_0.5.8.1          
[27] sass_0.4.9                  yaml_2.3.10                
[29] pkgdown_2.1.1               pillar_1.9.0               
[31] crayon_1.5.3                jquerylib_0.1.4            
[33] SingleCellExperiment_1.26.0 DelayedArray_0.30.1        
[35] cachem_1.1.0                abind_1.4-8                
[37] tidyselect_1.2.1            digest_0.6.37              
[39] dplyr_1.1.4                 bookdown_0.40              
[41] labeling_0.4.3              fastmap_1.2.0              
[43] grid_4.4.1                  colorspace_2.1-1           
[45] cli_3.6.3                   SparseArray_1.4.8          
[47] magrittr_2.0.3              S4Arrays_1.4.1             
[49] utf8_1.2.4                  withr_3.0.1                
[51] UCSC.utils_1.0.0            scales_1.3.0               
[53] rmarkdown_2.28              XVector_0.44.0             
[55] httr_1.4.7                  matrixStats_1.4.1          
[57] ragg_1.3.3                  evaluate_1.0.0             
[59] knitr_1.48                  GenomicRanges_1.56.1       
[61] IRanges_2.38.1              rlang_1.1.4                
[63] glue_1.7.0                  BiocManager_1.30.25        
[65] BiocGenerics_0.50.0         jsonlite_1.8.8             
[67] R6_2.5.1                    MatrixGenerics_1.16.0      
[69] systemfonts_1.1.0           fs_1.6.4                   
[71] zlibbioc_1.50.0            </code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
