<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting Started with `scDiagnostics` • scDiagnostics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Getting Started with `scDiagnostics`">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/QCandAnnotationScores.html">Visualization of QC and Annotation Scores</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/StatisticalMeasures.html">Statistical Measures to Assess Dataset Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection of Annotation Anomalies</a>
    </li>
    <li>
      <a href="../articles/CellDistancesDiagnostics.html">Calculation of Distances Between Specific Cells and Cell Populations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting Started with
<code>scDiagnostics</code>
</h1>
                        <h4 data-toc-skip class="author">Anthony
Christidis</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Andrew
Ghazi</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Smriti
Chawla</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Nitesh
Turaga</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Ludwig
Geistlinger</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Robert
Gentleman</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/vignettes/scDiagnostics.Rmd" class="external-link"><code>vignettes/scDiagnostics.Rmd</code></a></small>
      <div class="hidden name"><code>scDiagnostics.Rmd</code></div>

    </address>
</address>
</address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="purpose">Purpose<a class="anchor" aria-label="anchor" href="#purpose"></a>
</h2>
<p>Annotation transfer from a reference dataset is a key process for
annotating cell types in new single-cell RNA-sequencing (scRNA-seq)
experiments. This approach provides a quick, automated, and reproducible
alternative to manual annotation based on marker gene expression.
Despite its advantages, challenges such as dataset imbalance and
unrecognized discrepancies between query and reference datasets can lead
to inaccurate annotations and affect subsequent analyses.</p>
<p>The <code>scDiagnostics</code> package is designed to address these
issues by offering a suite of diagnostic tools for the systematic
evaluation of cell type assignments in scRNA-seq data. It provides
functionality to assess whether query and reference datasets are
well-aligned, which is crucial for ensuring accurate annotation
transfer. In addition, <code>scDiagnostics</code> helps evaluate
annotation ambiguity, cluster heterogeneity, and marker gene alignment.
By providing insights into these aspects, <code>scDiagnostics</code>
enables researchers to determine the precision with which cells from a
new scRNA-seq experiment can be assigned to known cell types, thereby
supporting more accurate and reliable downstream analysis.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="release-version">Release Version<a class="anchor" aria-label="anchor" href="#release-version"></a>
</h3>
<p>To install the release version of the package from Bioconductor use
the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scDiagnostics"</span><span class="op">)</span></span></code></pre></div>
<p>NOTE: you will need the <em><a href="https://CRAN.R-project.org/package=BiocManager" class="external-link">BiocManager</a></em>
package to install from GitHub.</p>
<p>To build the package vignettes upon installation use:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scDiagnostics"</span>,</span>
<span>                     build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="development-version">Development Version<a class="anchor" aria-label="anchor" href="#development-version"></a>
</h3>
<p>To install the development version of the package from Github, use
the following command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ccb-hms/scDiagnostics"</span><span class="op">)</span></span></code></pre></div>
<p>NOTE: you will need the <em><a href="https://bioconductor.org/packages/3.19/remotes" class="external-link">remotes</a></em>
package to install from GitHub.</p>
<p>To build the package vignettes upon installation use:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ccb-hms/scDiagnostics"</span>,</span>
<span>                        build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                        dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Once you have installed the package, you can load it with the
following code:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/scDiagnostics" class="external-link">scDiagnostics</a></span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="preliminaries">Preliminaries<a class="anchor" aria-label="anchor" href="#preliminaries"></a>
</h2>
<p>To explore the full capabilities of the <code>scDiagnostics</code>
package, you have the option to use your own data or leverage the
datasets included within the <code>scDiagnostics</code> package itself.
In this guide, we will focus on utilizing these built-in datasets, which
provide a practical and convenient resource for demonstrating the
features of <code>scDiagnostics.</code> These datasets are specifically
designed to facilitate the exploration of the package’s functionalities
and to help evaluate the accuracy of cell type assignments. You can
learn more about the datasets by looking at the documentation of the
datasets available in the reference manual.</p>
<div class="section level3">
<h3 id="loading-datasets">Loading Datasets<a class="anchor" aria-label="anchor" href="#loading-datasets"></a>
</h3>
<p>In these datasets available in the <code>scDiagnostics</code>
package, <code>reference_data</code>, <code>query_data</code>, and
<code>qc_data</code> are all <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects that include a <code>logcounts</code> assay, which stores the
log-transformed expression values for the genes.</p>
<p>The <code>reference_data</code> and <code>query_data</code> objects
both originate from scRNA-seq experiments on hematopoietic tissues,
specifically bone marrow samples, as provided by the <em><a href="https://bioconductor.org/packages/3.19/scRNAseq" class="external-link">scRNAseq</a></em>
package. These datasets have undergone comprehensive processing and
cleaning, ensuring high-quality data for downstream analysis.
Log-normalized counts were added to both datasets using the <em><a href="https://bioconductor.org/packages/3.19/scuttle" class="external-link">scuttle</a></em>
package. The <code>query_data</code> object has been further annotated
with cell type assignments using the <em><a href="https://bioconductor.org/packages/3.19/SingleR" class="external-link">SingleR</a></em>
package, and it includes <code>annotation_scores</code> that reflect the
confidence in these annotations. Additionally, gene set scores were
computed and incorporated into the <code>query_data</code> using the
<em><a href="https://bioconductor.org/packages/3.19/AUCell" class="external-link">AUCell</a></em>
package. For feature selection, the top 500 highly variable genes (HVGs)
common to both datasets were identified and retained using the <em><a href="https://bioconductor.org/packages/3.19/scran" class="external-link">scran</a></em>
package. Finally, dimensionality reduction techniques including PCA,
t-SNE, and UMAP were applied to both datasets, with the results stored
within each object using the <em><a href="https://bioconductor.org/packages/3.19/scater" class="external-link">scater</a></em>
package.</p>
<p>The <code>qc_data</code> dataset in this package is derived from the
<code>hpca</code> dataset available in the <em><a href="https://bioconductor.org/packages/3.19/celldex" class="external-link">celldex</a></em>
package. Like the other datasets, <code>qc_data</code> has undergone
significant cleaning and processing to ensure high data quality. Quality
control (QC) metrics were added using the <em><a href="https://bioconductor.org/packages/3.19/scuttle" class="external-link">scuttle</a></em>
package. Cell type annotations and associated annotation_scores were
generated using the <em><a href="https://bioconductor.org/packages/3.19/SingleR" class="external-link">SingleR</a></em>
package. Additionally, the top highly variable genes were selected using
the <em><a href="https://bioconductor.org/packages/3.19/scran" class="external-link">scran</a></em>
package to enhance the dataset’s utility for downstream analyses.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"qc_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>The <code>reference_data</code> contains a column data labeled
<code>expert_annotation</code>, which provides cell type annotations
assigned by experts. On the other hand, <code>query_data</code> also
includes <code>expert_annotation</code>, but it additionally features
<code>SingleR_annotation</code>, which is the cell type annotation
generated by the <em><a href="https://bioconductor.org/packages/3.19/SingleR" class="external-link">SingleR</a></em>
package, a popular package for cell type assignment based on reference
datasets. The <code>qc_data</code> object contains a special column
called <code>annotation_scores</code>, which holds the scores from the
<code>SingleR</code> annotations, providing a measure of confidence or
relevance for the assigned cell types.</p>
<p>By working with these datasets, you can gain hands-on experience with
the various diagnostic tools and functions offered by
<code>scDiagnostics</code>, allowing you to better understand how well
it aligns query and reference datasets, assesses annotation ambiguity,
and evaluates cluster heterogeneity and marker gene alignment.</p>
</div>
<div class="section level3">
<h3 id="subsetting-the-datasets">Subsetting the Datasets<a class="anchor" aria-label="anchor" href="#subsetting-the-datasets"></a>
</h3>
<p>Some functions in the vignette are designed to work with <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects that contain data from only one cell type. We will create
separate <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects that only CD4 cells, to ensure compatibility with these
functions.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset to CD4 cells</span></span>
<span><span class="va">ref_data_cd4</span> <span class="op">&lt;-</span> <span class="va">reference_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span></span>
<span>    <span class="va">reference_data</span><span class="op">$</span><span class="va">expert_annotation</span> <span class="op">==</span> <span class="st">"CD4"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="va">query_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span></span>
<span>    <span class="va">query_data</span><span class="op">$</span><span class="va">expert_annotation</span> <span class="op">==</span> <span class="st">"CD4"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Select highly variable genes</span></span>
<span><span class="va">ref_top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">ref_data_cd4</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">query_top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">query_data_cd4</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">common_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">ref_top_genes</span>, <span class="va">query_top_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset data by common genes</span></span>
<span><span class="va">ref_data_cd4</span> <span class="op">&lt;-</span> <span class="va">ref_data_cd4</span><span class="op">[</span><span class="va">common_genes</span>,<span class="op">]</span></span>
<span><span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="va">query_data_cd4</span><span class="op">[</span><span class="va">common_genes</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Run PCA on both datasets</span></span>
<span><span class="va">ref_data_cd4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">ref_data_cd4</span><span class="op">)</span></span>
<span><span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">query_data_cd4</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="getting-started-with-scdiagnostics">Getting Started with <code>scDiagnostics</code><a class="anchor" aria-label="anchor" href="#getting-started-with-scdiagnostics"></a>
</h2>
<p><strong>The functions introduced in this section represent just a
subset of the functions available in the <code>scDiagnostics</code>
package.</strong></p>
<p>For a complete overview and detailed demonstrations of all the
functions included in the package, please refer to the designated
vignettes which you may browse from the <a href="https://ccb-hms.github.io/scDiagnostics/">pkgdown site for
<code>scDiagnostics</code></a>. Each vignette is designed to address
specific aspects of <code>scDiagnostics</code>, and this vignette
highlights key functionalities to illustrate their applications. These
vignettes provide in-depth guidance and examples for each function,
helping users fully leverage the capabilities of
<code>scDiagnostics</code> in their single-cell analyses.</p>
</div>
<div class="section level2">
<h2 id="visualization-of-cell-type-annotations">Visualization of Cell Type Annotations<a class="anchor" aria-label="anchor" href="#visualization-of-cell-type-annotations"></a>
</h2>
<p>For a detailed example of all possible functions to visualize
reference and query datasets, please refer to the <a href="https://ccb-hms.github.io/scDiagnostics/articles/VisualizationTools.html">Visualization
of Cell Type Annotations</a> vignette.</p>
<div class="section level3">
<h3 id="plotcelltypepca">
<code>plotCellTypePCA()</code><a class="anchor" aria-label="anchor" href="#plotcelltypepca"></a>
</h3>
<p>The <code><a href="../reference/plotCellTypePCA.html">plotCellTypePCA()</a></code> function provides a visual
comparison of principal components (PCs) for different cell types across
query and reference datasets. By projecting the query data onto the PCA
space of the reference dataset, it creates informative plots to help you
understand how various cell types are distributed in the principal
component space.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot PCA data</span></span>
<span><span class="va">pc_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCellTypePCA.html">plotCellTypePCA</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>    cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Display plot</span></span>
<span><span class="va">pc_plot</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/scDiagnostics/plotCellTypePCA.png"></p>
<p>The function returns a <code>ggplot</code> object featuring pairwise
scatter plots of the selected principal components. Each plot compares
how different cell types from the query and reference datasets project
onto the PCA space. This visualization aids in identifying how cell
types distribute across PCs and facilitates comparisons between
datasets.</p>
<p>The <code>reference_data</code> argument contains the reference cell
data, which serves as the foundation for defining the PC space. The
<code>query_data</code> parameter includes the query cell data that will
be projected. The function uses the <code>ref_cell_type_col</code> and
<code>query_cell_type_col</code> to identify the relevant cell type
annotations in the reference and query datasets.</p>
</div>
<div class="section level3">
<h3 id="calculatediscriminantspace">
<code>calculateDiscriminantSpace()</code><a class="anchor" aria-label="anchor" href="#calculatediscriminantspace"></a>
</h3>
<p>Alternatively, you can also use the
<code><a href="../reference/calculateDiscriminantSpace.html">calculateDiscriminantSpace()</a></code> function, which projects query
single-cell RNA-seq data onto a discriminant space defined by a
reference dataset. This approach helps evaluate the similarity between
the query and reference data, offering insights into the classification
of query cells.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">disc_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateDiscriminantSpace.html">calculateDiscriminantSpace</a></span><span class="op">(</span></span>
<span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The function returns a comprehensive output that includes
discriminant eigenvalues and eigenvectors, which represent the variance
explained by each discriminant axis and are used to project the data. It
also provides the projections of the reference and query data onto the
discriminant space. The Mahalanobis distances between the query and
reference cell types are calculated, offering insights into how close
the query projections are to the reference. The cosine similarity scores
provide another metric to assess the similarity between the
datasets.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">disc_output</span>, plot_type <span class="op">=</span> <span class="st">"scatterplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/scDiagnostics/calculateDiscriminantSpace.png"></p>
<p>Alternatively, you can create a boxplot comparing query and reference
projections for a specific cell type. See the reference manual for an
example.</p>
</div>
<div class="section level3">
<h3 id="visualization-of-marker-expressions">Visualization of Marker Expressions<a class="anchor" aria-label="anchor" href="#visualization-of-marker-expressions"></a>
</h3>
<p>Visualizing gene expression distributions is crucial for
understanding dataset similarity and cell type-specific expression
patterns. The <code><a href="../reference/plotMarkerExpression.html">plotMarkerExpression()</a></code> function allows you to
compare the expression levels of a specific gene between a reference
dataset and a query dataset, both overall and within a specified cell
type. This comparison is done using density plots, which help in
assessing the alignment and potential discrepancies between
datasets.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMarkerExpression.html">plotMarkerExpression</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                     query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                     ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                     query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                     gene_name <span class="op">=</span> <span class="st">"VPREB3"</span>, </span>
<span>                     cell_type <span class="op">=</span> <span class="st">"B_and_plasma"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/scDiagnostics/plotMarkerExpression.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="visualization-of-qc-and-annotation-scores">Visualization of QC and Annotation Scores<a class="anchor" aria-label="anchor" href="#visualization-of-qc-and-annotation-scores"></a>
</h2>
<p>For a detailed example of all possible functions to visualize quality
control and annotation score metrics, please refer to the <a href="https://ccb-hms.github.io/scDiagnostics/articles/VisualizationTools.html">Visualization
of QC and Annotation Scores</a> vignette.</p>
<p>The scatter plot illustrates the relationship between key QC metrics
(e.g., total library size, percentage of mitochondrial genes) and cell
type annotation scores. This visualization aids in assessing how QC
factors might influence or correlate with the assigned cell type
annotations.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate scatter plot</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotQCvsAnnotation.html">plotQCvsAnnotation</a></span><span class="op">(</span>se_object <span class="op">=</span> <span class="va">qc_data</span>,</span>
<span>                         cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>                         qc_col <span class="op">=</span> <span class="st">"total"</span>,</span>
<span>                         score_col <span class="op">=</span> <span class="st">"annotation_scores"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Library Size"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/scDiagnostics/plotQCvsAnnotation.png"></p>
<p>This scatter plot can uncover patterns, such as whether cells with
larger library sizes or higher mitochondrial content are linked to
specific annotations. For example, cells exhibiting unusually high
mitochondrial content may be flagged as low-quality or stressed, which
could impact their annotation accuracy.</p>
</div>
<div class="section level2">
<h2 id="evaluation-of-dataset-and-marker-gene-alignment">Evaluation of Dataset and Marker Gene Alignment<a class="anchor" aria-label="anchor" href="#evaluation-of-dataset-and-marker-gene-alignment"></a>
</h2>
<p>For a detailed example of all possible functions to assess the
alignment of datasets and marker genes, please refer to the <a href="https://ccb-hms.github.io/scDiagnostics/articles/https://ccb-hms.github.io/scDiagnostics/articles/DatasetAlignment.html">Evaluation
of Dataset and Marker Gene Alignment</a> vignette.</p>
<div class="section level3">
<h3 id="comparepcasubspace">
<code>comparePCASubspace()</code><a class="anchor" aria-label="anchor" href="#comparepcasubspace"></a>
</h3>
<p>In single-cell RNA-seq analysis, evaluating how closely the subspaces
defined by the leading principal components (PCs) of query and reference
datasets match is crucial. This evaluation is key to understanding how
each dataset captures and represents structure and variation. The
<code><a href="../reference/comparePCASubspace.html">comparePCASubspace()</a></code> function is specifically designed to
assess this alignment by calculating the cosine similarity between the
loadings of the most significant variables for each principal component.
This analysis is essential for measuring the extent of similarity
between datasets, which is vital for precise cell type annotation and
effective data integration.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compare PCA subspaces between query and reference data</span></span>
<span><span class="va">subspace_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/comparePCASubspace.html">comparePCASubspace</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data_cd4</span>,</span>
<span>    reference_data <span class="op">=</span> <span class="va">ref_data_cd4</span>, </span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View weighted cosine similarity score</span></span>
<span><span class="va">subspace_comparison</span><span class="op">$</span><span class="va">weighted_cosine_similarity</span></span>
<span><span class="co">#&gt; [1] 0.2578375</span></span>
<span></span>
<span><span class="co"># Plot output for PCA subspace comparison (if a plot method is available)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">subspace_comparison</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/scDiagnostics/comparePCASubspace.png"></p>
</div>
<div class="section level3">
<h3 id="plotpairwisedistancesdensity">
<code>plotPairwiseDistancesDensity()</code><a class="anchor" aria-label="anchor" href="#plotpairwisedistancesdensity"></a>
</h3>
<p>The <code><a href="../reference/plotPairwiseDistancesDensity.html">plotPairwiseDistancesDensity()</a></code> function calculates
and visualizes pairwise distances or correlations between cell types in
query and reference datasets, aiding in the evaluation of cell type
annotation consistency in single-cell RNA sequencing (scRNA-seq)
analysis. Operating on <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects, it allows users to specify cell types of interest and compute
either distances or correlation coefficients, with the option to project
data into PCA space for focused analysis. The function generates a
density plot using <code>ggplot2</code>, comparing cell relationships
within and between datasets.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example usage of the function</span></span>
<span><span class="fu"><a href="../reference/plotPairwiseDistancesDensity.html">plotPairwiseDistancesDensity</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                             reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                             query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                             ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                             cell_type_query <span class="op">=</span> <span class="st">"CD8"</span>, </span>
<span>                             cell_type_ref <span class="op">=</span> <span class="st">"CD8"</span>, </span>
<span>                             pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                             distance_metric <span class="op">=</span> <span class="st">"correlation"</span>,</span>
<span>                             correlation_method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/scDiagnostics/plotPairwiseDistancesDensity.png"></p>
</div>
<div class="section level3">
<h3 id="plotwassersteindistance">
<code>plotWassersteinDistance()</code><a class="anchor" aria-label="anchor" href="#plotwassersteindistance"></a>
</h3>
<p>The code below illustrates how to use the
<code><a href="../reference/plotWassersteinDistance.html">plotWassersteinDistance()</a></code> function to compare the
Wasserstein distances between CD4 cells in the reference and query
datasets. The resulting plot provides insight into whether the
differences between the datasets are statistically significant.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate the Wasserstein distance density plot</span></span>
<span><span class="fu"><a href="../reference/plotWassersteinDistance.html">plotWassersteinDistance</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_cd4</span>,</span>
<span>                        reference_data <span class="op">=</span> <span class="va">ref_data_cd4</span>, </span>
<span>                        query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                        ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                        pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>                        alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/scDiagnostics/plotWassersteinDistance.png"></p>
</div>
<div class="section level3">
<h3 id="calculatevarimpoverlap">
<code>calculateVarImpOverlap()</code><a class="anchor" aria-label="anchor" href="#calculatevarimpoverlap"></a>
</h3>
<p>Imagine you have two sets of data: one called reference_data and
another called query_data. Both sets include information about gene
expression and cell types, with columns named
<code>expert_annotation</code> for the reference data and
<code>SingleR_annotation</code> for the query data. You want to find out
which genes are most important for each dataset and then compare
them.</p>
<p>Here’s how you can do it with the function:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># RF function to compare (between datasets) which genes are best at differentiating cell types</span></span>
<span><span class="va">rf_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateVarImpOverlap.html">calculateVarImpOverlap</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                    query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                                    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                    n_tree <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                    n_top <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Comparison table</span></span>
<span><span class="va">rf_output</span><span class="op">$</span><span class="va">var_imp_comparison</span></span>
<span><span class="co">#&gt;              CD4-CD8     CD4-B_and_plasma          CD4-Myeloid </span></span>
<span><span class="co">#&gt;                 0.74                 0.76                 0.68 </span></span>
<span><span class="co">#&gt;     CD8-B_and_plasma          CD8-Myeloid B_and_plasma-Myeloid </span></span>
<span><span class="co">#&gt;                 0.70                 0.56                 0.52</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="statistical-measures-to-assess-dataset-alignment">Statistical Measures to Assess Dataset Alignment<a class="anchor" aria-label="anchor" href="#statistical-measures-to-assess-dataset-alignment"></a>
</h2>
<p>For a detailed example of all possible functions that implement
statistical methods to assess cell type annotation, please refer to the
<a href="https://ccb-hms.github.io/scDiagnostics/articles/StatisticalMeasures.html">Statistical
Measures to Assess Dataset Alignment</a> vignette.</p>
<div class="section level3">
<h3 id="calculateaveragepairwisecorrelation">
<code>calculateAveragePairwiseCorrelation()</code><a class="anchor" aria-label="anchor" href="#calculateaveragepairwisecorrelation"></a>
</h3>
<p>The <code><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation()</a></code> function
computes the average pairwise correlations between specified cell types
in single-cell gene expression data. It calculates pairwise correlations
between query and reference cells using a specified correlation method,
and then averages these correlations for each cell type pair. This
approach helps assess the similarity between cells in reference and
query datasets and provides insights into the reliability of cell type
annotations.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute pairwise correlations between specified cell types</span></span>
<span><span class="va">cor_matrix_avg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation</a></span><span class="op">(</span></span>
<span>  query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>  reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>  query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>  ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>  cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span><span class="op">)</span>, </span>
<span>  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>  correlation_method <span class="op">=</span> <span class="st">"spearman"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize the average pairwise correlation matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cor_matrix_avg</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/scDiagnostics/calculateAveragePairwiseCorrelation.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="detection-of-annotation-anomalies">Detection of Annotation Anomalies<a class="anchor" aria-label="anchor" href="#detection-of-annotation-anomalies"></a>
</h2>
<p>For a detailed example of all possible functions that implement
statistical methods to assess cell type annotation, please refer to the
<a href="https://ccb-hms.github.io/scDiagnostics/articles/AnnotationAnomalies.html">Detection
of Annotation Anomalies</a> vignette.</p>
<p>The <code><a href="../reference/detectAnomaly.html">detectAnomaly()</a></code> function is designed to identify
anomalies in single-cell data by leveraging PCA projections and the
Isolation Forest algorithm. This method is useful for detecting
anomalies or unusual patterns in single-cell datasets, whether you’re
analyzing a reference dataset or comparing a query dataset against
it.</p>
<p>The function projects single-cell data onto a PCA space and builds an
Isolation Forest model on this PCA space to detect anomalies. If a query
dataset is provided, the function computes anomaly scores for the query
data based on its PCA projections relative to the reference data. If no
query data is provided, it computes anomaly scores for the reference
data itself.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform anomaly detection</span></span>
<span><span class="va">anomaly_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectAnomaly.html">detectAnomaly</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>                                pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co"># Plot the results for a specific cell type</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">anomaly_output</span>, </span>
<span>     cell_type <span class="op">=</span> <span class="st">"CD4"</span>, </span>
<span>     data_type <span class="op">=</span> <span class="st">"query"</span>,</span>
<span>     pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/scDiagnostics/detectAnomaly.png"></p>
</div>
<div class="section level2">
<h2 id="calculation-of-distances-between-specific-cells-and-cell-types">Calculation of Distances Between Specific Cells and Cell Types<a class="anchor" aria-label="anchor" href="#calculation-of-distances-between-specific-cells-and-cell-types"></a>
</h2>
<p>For a detailed example of all possible functions to analyze distances
between specific cells and cell distributions, please refer to the <a href="https://ccb-hms.github.io/scDiagnostics/articles/CellDistancesDiagnostics.html">Calculation
of Distances Between Specific Cells and Cell Types</a> vignette.</p>
<p>The <code><a href="../reference/calculateCellDistances.html">calculateCellDistances()</a></code> function computes distances
both within a reference dataset and between query cells and reference
cells for each specified cell type. By first projecting data onto a PCA
space, the function calculates Euclidean distances to quantify
similarities and dissimilarities. For each cell type, it generates
pairwise distances within the reference dataset and measures how far
each query cell is from all reference cells. This approach enables
detailed analysis of cell type-specific distances, aiding in the
identification of outliers and other patterns of interest.</p>
<p>To identify anomalous cells within the query data, we first use the
<code><a href="../reference/detectAnomaly.html">detectAnomaly()</a></code> function, focusing specifically on the CD4
cell type. This function will compute anomaly scores for each CD4 cell
in the query dataset based on their projection in the PCA space of the
reference data. Next, we will plot the distance distributions for the
top 6 CD4 cells with the highest anomaly scores. These distances,
computed using the <code><a href="../reference/calculateCellDistances.html">calculateCellDistances()</a></code> function, will
illustrate how these anomalous cells differ from the reference cells,
providing insight into their potential outlier status and helping to
visualize patterns within the data.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify outliers for CD4</span></span>
<span><span class="va">cd4_anomalies</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectAnomaly.html">detectAnomaly</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                               query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                               query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                               ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span><span class="op">)</span></span>
<span><span class="va">cd4_top6_anomalies</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">cd4_anomalies</span><span class="op">$</span><span class="va">CD4</span><span class="op">$</span><span class="va">query_anomaly_scores</span>, </span>
<span>                                 decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the PC data</span></span>
<span><span class="va">distance_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateCellDistances.html">calculateCellDistances</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span></span>
<span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Plot the densities of the distances</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">distance_data</span>, ref_cell_type <span class="op">=</span> <span class="st">"CD4"</span>, cell_names <span class="op">=</span> <span class="va">cd4_top6_anomalies</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/scDiagnostics/calculateCellDistances.png"></p>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>In this vignette, we have introduced the core functionalities of the
<code>scDiagnostics</code> package, providing a detailed guide on its
use for scRNA-seq data analysis. We started by exploring the basic
capabilities, including the visualization of query versus reference
datasets through dimensionality reduction techniques. This approach
allows for both the evaluation of multiple cell types simultaneously and
the comparison of distributions for individual cell types.</p>
<p>We then delved into the visualization of annotation scores, which is
essential for assessing the accuracy of cell type assignments. Following
this, we discussed statistical methods for evaluating annotation
precision. This includes assessing correlation between cell types and
performing statistical tests to compare two datasets, offering a
rigorous approach to validate cell type classifications.</p>
<p>The vignette also covered techniques for detecting annotation
anomalies, crucial for identifying and addressing potential issues that
could impact data validity. Additionally, we explored cell distance
diagnostics, focusing on analyzing distances between specific cells and
various cell populations or distributions to better understand internal
relationships and variations.</p>
<p>Overall, this guide equips you with practical tools and methods to
enhance the reliability and accuracy of your scRNA-seq data analysis
using <code>scDiagnostics</code>, paving the way for more informed and
meaningful biological interpretations.</p>
<hr>
</div>
<div class="section level2">
<h2 id="r-session-info">R Session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h2>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scater_1.32.1               ggplot2_3.5.1              
 [3] scran_1.32.0                scuttle_1.14.0             
 [5] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0
 [7] Biobase_2.64.0              GenomicRanges_1.56.1       
 [9] GenomeInfoDb_1.40.1         IRanges_2.38.1             
[11] S4Vectors_0.42.1            BiocGenerics_0.50.0        
[13] MatrixGenerics_1.16.0       matrixStats_1.4.1          
[15] scDiagnostics_0.99.8        BiocStyle_2.32.1           

loaded via a namespace (and not attached):
 [1] gridExtra_2.3             rlang_1.1.4              
 [3] magrittr_2.0.3            compiler_4.4.1           
 [5] DelayedMatrixStats_1.26.0 systemfonts_1.1.0        
 [7] vctrs_0.6.5               pkgconfig_2.0.3          
 [9] crayon_1.5.3              fastmap_1.2.0            
[11] XVector_0.44.0            labeling_0.4.3           
[13] utf8_1.2.4                rmarkdown_2.28           
[15] UCSC.utils_1.0.0          ggbeeswarm_0.7.2         
[17] ragg_1.3.3                xfun_0.47                
[19] bluster_1.14.0            zlibbioc_1.50.0          
[21] cachem_1.1.0              beachmat_2.20.0          
[23] jsonlite_1.8.8            DelayedArray_0.30.1      
[25] BiocParallel_1.38.0       irlba_2.3.5.1            
[27] parallel_4.4.1            cluster_2.1.6            
[29] R6_2.5.1                  bslib_0.8.0              
[31] ranger_0.16.0             limma_3.60.4             
[33] jquerylib_0.1.4           Rcpp_1.0.13              
[35] bookdown_0.40             knitr_1.48               
[37] Matrix_1.7-0              igraph_2.0.3             
[39] tidyselect_1.2.1          abind_1.4-8              
[41] yaml_2.3.10               viridis_0.6.5            
[43] codetools_0.2-20          lattice_0.22-6           
[45] tibble_3.2.1              withr_3.0.1              
[47] evaluate_1.0.0            desc_1.4.3               
[49] pillar_1.9.0              BiocManager_1.30.25      
[51] generics_0.1.3            sparseMatrixStats_1.16.0 
[53] munsell_0.5.1             scales_1.3.0             
[55] glue_1.7.0                metapod_1.12.0           
[57] tools_4.4.1               data.table_1.16.0        
[59] BiocNeighbors_1.22.0      ScaledMatrix_1.12.0      
[61] locfit_1.5-9.10           fs_1.6.4                 
[63] grid_4.4.1                edgeR_4.2.1              
[65] colorspace_2.1-1          GenomeInfoDbData_1.2.12  
[67] beeswarm_0.4.0            BiocSingular_1.20.0      
[69] vipor_0.4.7               cli_3.6.3                
[71] rsvd_1.0.5                textshaping_0.4.0        
[73] fansi_1.0.6               S4Arrays_1.4.1           
[75] viridisLite_0.4.2         dplyr_1.1.4              
[77] gtable_0.3.5              isotree_0.6.1-1          
[79] sass_0.4.9                digest_0.6.37            
[81] SparseArray_1.4.8         ggrepel_0.9.6            
[83] dqrng_0.4.1               farver_2.1.2             
[85] htmlwidgets_1.6.4         htmltools_0.5.8.1        
[87] pkgdown_2.1.1             lifecycle_1.0.4          
[89] httr_1.4.7                transport_0.15-4         
[91] statmod_1.5.0            </code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
