<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Statistical Measures to Assess Dataset Alignment • scDiagnostics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Statistical Measures to Assess Dataset Alignment">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/QCandAnnotationScores.html">Visualization of QC and Annotation Scores</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/StatisticalMeasures.html">Statistical Measures to Assess Dataset Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection of Annotation Anomalies</a>
    </li>
    <li>
      <a href="../articles/CellDistancesDiagnostics.html">Calculation of Distances Between Specific Cells and Cell Populations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Statistical Measures to Assess Dataset
Alignment</h1>
                        <h4 data-toc-skip class="author">Anthony
Christidis</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Andrew
Ghazi</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Smriti
Chawla</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Nitesh
Turaga</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Ludwig
Geistlinger</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Robert
Gentleman</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/vignettes/StatisticalMeasures.Rmd" class="external-link"><code>vignettes/StatisticalMeasures.Rmd</code></a></small>
      <div class="hidden name"><code>StatisticalMeasures.Rmd</code></div>

    </address>
</address>
</address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In single-cell transcriptomics, it is essential to rigorously analyze
and interpret the complex data generated from high-throughput
experiments. This vignette introduces several key functions that
facilitate advanced statistical analysis and data integration within the
context of single-cell experiments. Each function serves a specific
purpose, aiding in the comprehensive assessment of single-cell data and
its associated covariates.</p>
<ol style="list-style-type: decimal">
<li><p><code><a href="../reference/calculateCramerPValue.html">calculateCramerPValue()</a></code>: This function computes the
Cramér’s V statistic, which measures the strength of association between
categorical variables. Cramér’s V is particularly useful for quantifying
the association strength in contingency tables, providing insight into
how strongly different cell types or conditions are related.</p></li>
<li><p><code><a href="../reference/calculateHotellingPValue.html">calculateHotellingPValue()</a></code>: This function performs
Hotelling’s T-squared test, a multivariate statistical test used to
assess the differences between two groups of observations. It is useful
for comparing the mean vectors of multiple variables and is commonly
employed in single-cell data to evaluate group differences in principal
component space.</p></li>
<li><p><code><a href="../reference/calculateNearestNeighborProbabilities.html">calculateNearestNeighborProbabilities()</a></code>: This
function calculates the probability of nearest neighbor assignments in a
high-dimensional space. It is used to evaluate the likelihood of a cell
belonging to a specific cluster based on its proximity to neighboring
cells. This function helps in assessing clustering quality and the
robustness of cell type assignments.</p></li>
<li><p><code><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation()</a></code>: This function
computes the average pairwise correlation between variables or features
across cells. It provides an overview of the relationships between
different markers or genes, helping to identify correlated features and
understand their interactions within the dataset.</p></li>
<li><p><code><a href="../reference/regressPC.html">regressPC()</a></code>: This function performs linear
regression of a covariate of interest onto one or more principal
components derived from single-cell data. This analysis helps quantify
the variance explained by the covariate and is useful for tasks such as
assessing batch effects, clustering homogeneity, and alignment of query
and reference datasets.</p></li>
</ol>
<p>Together, these functions provide a reliable toolkit for analyzing
and interpreting single-cell data, allowing researchers to gain deeper
insights into the underlying biological processes and the relationships
between different variables and cell types. The following sections of
this vignette will guide you through the usage and application of each
function, illustrating their role in comprehensive data analysis.</p>
</div>
<div class="section level2">
<h2 id="preliminaries">Preliminaries<a class="anchor" aria-label="anchor" href="#preliminaries"></a>
</h2>
<p>In the context of the <code>scDiagnostics</code> package, this
vignette demonstrates how to leverage five key functions to evaluate
cell type annotations across two distinct datasets:</p>
<ul>
<li><p><code>reference_data</code>: This dataset provides expert-curated
cell type annotations, serving as a gold standard for evaluating the
quality of cell type annotations. It acts as a benchmark to assess and
detect anomalies or inconsistencies in cell type
classification.</p></li>
<li><p><code>query_data</code>: This dataset contains cell type
annotations from both expert assessments and those generated by the
<em><a href="https://bioconductor.org/packages/3.19/SingleR" class="external-link">SingleR</a></em>
package. By comparing these annotations with the reference dataset, you
can identify discrepancies between manual and automated annotations.
This comparison helps in pinpointing inconsistencies and areas that may
require further investigation or refinement.</p></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/scDiagnostics" class="external-link">scDiagnostics</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculatecramerpvalue">
<code>calculateCramerPValue()</code><a class="anchor" aria-label="anchor" href="#calculatecramerpvalue"></a>
</h2>
<p>The calculateCramerPValue function is designed to perform the Cramer
test for comparing multivariate empirical cumulative distribution
functions (ECDFs) between two samples in single-cell data. This test is
particularly useful for assessing whether the distributions of principal
components (PCs) differ significantly between the reference and query
datasets for specific cell types.</p>
<p>To use this function, you first need to provide two key inputs:
<code>reference_data</code> and <code>query_data</code>, both of which
should be <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects containing numeric expression matrices. If
<code>query_data</code> is not supplied, the function will only use the
<code>reference_data.</code> You should also specify the column names
for cell type annotations in both datasets via
<code>ref_cell_type_col</code> and <code>query_cell_type_col.</code> If
<code>cell_types</code> is not provided, the function will automatically
include all unique cell types found in the datasets. The
<code>pc_subset</code> parameter allows you to define which principal
components to include in the analysis, with the default being the first
five PCs.</p>
<p>The function performs the following steps: it first projects the data
into PCA space, subsets the data according to the specified cell types
and principal components, and then applies the Cramer test to compare
the ECDFs between the reference and query datasets. The result is a
named vector of p-values from the Cramer test for each cell type, which
indicates whether there is a significant difference in the distributions
of PCs between the two datasets.</p>
<p>Here’s an example of how to use the
<code><a href="../reference/calculateCramerPValue.html">calculateCramerPValue()</a></code> function:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform Cramer test for the specified cell types and principal components</span></span>
<span><span class="va">cramer_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateCramerPValue.html">calculateCramerPValue</a></span><span class="op">(</span></span>
<span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>    cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Display the Cramer test p-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">cramer_test</span><span class="op">)</span></span>
<span><span class="co">#&gt;          CD4          CD8 B_and_plasma      Myeloid </span></span>
<span><span class="co">#&gt;    0.0000000    0.0000000    0.1368631    0.4105894</span></span></code></pre></div>
<p>In this example, the function compares the distributions of the first
five principal components between the reference and query datasets for
cell types such as “CD4”, “CD8”, “B_and_plasma”, and “Myeloid”. The
output is a vector of p-values indicating whether the differences in PC
distributions are statistically significant.</p>
</div>
<div class="section level2">
<h2 id="calculatehotellingpvalue">
<code>calculateHotellingPValue()</code><a class="anchor" aria-label="anchor" href="#calculatehotellingpvalue"></a>
</h2>
<p>The <code><a href="../reference/calculateHotellingPValue.html">calculateHotellingPValue()</a></code> function is designed to
compute Hotelling’s T-squared test statistic and corresponding p-values
for comparing multivariate means between reference and query datasets in
the context of single-cell RNA-seq data. This statistical test is
particularly useful for assessing whether the mean vectors of principal
components (PCs) differ significantly between the two datasets, which
can be indicative of differences in the cell type distributions.</p>
<p>To use this function, you need to provide two <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects: <code>query_data</code> and <code>reference_data</code>, each
containing numeric expression matrices. You also need to specify the
column names for cell type annotations in both datasets with
<code>query_cell_type_col</code> and <code>ref_cell_type_col.</code> The
<code>cell_types</code> parameter allows you to choose which cell types
to include in the analysis, and if not specified, the function will
automatically include all cell types present in the datasets. The
<code>pc_subset</code> parameter determines which principal components
to consider, with the default being the first five PCs. Additionally,
<code>n_permutation</code> specifies the number of permutations for
calculating p-values, with a default of 500.</p>
<p>The function works by first projecting the data into PCA space and
then performing Hotelling’s T-squared test for each specified cell type
to compare the means between the reference and query datasets. It uses
permutation testing to determine the p-values, indicating whether the
observed differences in means are statistically significant. The result
is a named numeric vector of p-values for each cell type.</p>
<p>Here is an example of how to use the
<code><a href="../reference/calculateHotellingPValue.html">calculateHotellingPValue()</a></code> function:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate Hotelling's T-squared test p-values</span></span>
<span><span class="va">p_values</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHotellingPValue.html">calculateHotellingPValue</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Display the p-values rounded to five decimal places</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">p_values</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;          CD4          CD8 B_and_plasma      Myeloid </span></span>
<span><span class="co">#&gt;         0.00         0.00         0.12         0.00</span></span></code></pre></div>
<p>In this example, the function calculates p-values for Hotelling’s
T-squared test, focusing on the first ten principal components to assess
whether the multivariate means differ significantly between the
reference and query datasets for each cell type. The p-values indicate
the likelihood of observing the differences by chance and help identify
significant disparities in cell type distributions.</p>
</div>
<div class="section level2">
<h2 id="calculatenearestneighborprobabilities">
<code>calculateNearestNeighborProbabilities()</code><a class="anchor" aria-label="anchor" href="#calculatenearestneighborprobabilities"></a>
</h2>
<p>In this section, we will explore the
<code><a href="../reference/calculateNearestNeighborProbabilities.html">calculateNearestNeighborProbabilities()</a></code> function, which is
designed to estimate the probability of each query cell belonging to
either the reference or query dataset for each specified cell type. This
function uses nearest neighbor analysis on PCA-reduced data to achieve
this.</p>
<p>The function begins by reducing the dimensionality of both the query
and reference datasets using Principal Component Analysis (PCA). It then
compares each query cell to its nearest neighbors within the reference
dataset. To ensure that sample sizes between datasets are comparable,
the function uses data augmentation if necessary.</p>
<p>Here is a detailed explanation of how to use the
<code><a href="../reference/calculateNearestNeighborProbabilities.html">calculateNearestNeighborProbabilities()</a></code> function:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Loading the Data</strong>: First, ensure you have the
data available in the form of <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects for both query and reference datasets.</p></li>
<li><p><strong>Function Call</strong>: The function
<code><a href="../reference/calculateNearestNeighborProbabilities.html">calculateNearestNeighborProbabilities()</a></code> is called with
several parameters including the query and reference datasets, the names
of the columns containing cell type annotations, the subset of principal
components to use, and the number of nearest neighbors to
consider.</p></li>
</ol>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute nearest-neighbor probabilities</span></span>
<span><span class="va">nn_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateNearestNeighborProbabilities.html">calculateNearestNeighborProbabilities</a></span><span class="op">(</span></span>
<span>  query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>  reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>  query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>  ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>  n_neighbor <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>
<strong>Understanding the Output</strong>: The output of the
function is a list where each element corresponds to a specific cell
type. Each element of this list contains:</li>
</ol>
<ul>
<li>
<code>n_neighbor</code>: The number of nearest neighbors
considered.</li>
<li>
<code>n_query</code>: The number of cells in the query dataset for
that cell type.</li>
<li>
<code>query_prob</code>: The average probability of each query cell
belonging to the reference dataset.</li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li>
<strong>Plotting the Results</strong>: To visualize the results, you
can use the <code>plot</code> function provided for objects of class
<code>calculateNearestNeighborProbabilities</code>.</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">nn_output</span>, cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/StatisticalMeasures/calculateNearestNeighborProbabilities.png"></p>
<p>In summary, the <code><a href="../reference/calculateNearestNeighborProbabilities.html">calculateNearestNeighborProbabilities()</a></code>
function provides a robust method for classifying query cells based on
nearest neighbor analysis in PCA space. This can be particularly useful
for understanding how query cells relate to known reference datasets and
for visualizing these relationships across different cell types.</p>
</div>
<div class="section level2">
<h2 id="calculateaveragepairwisecorrelation">
<code>calculateAveragePairwiseCorrelation()</code><a class="anchor" aria-label="anchor" href="#calculateaveragepairwisecorrelation"></a>
</h2>
<p>The <code><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation()</a></code> function is
designed to compute the average pairwise correlations between specified
cell types in single-cell gene expression data. This function operates
on <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects and is ideal for single-cell analysis workflows. It calculates
pairwise correlations between query and reference cells using a
specified correlation method, and then averages these correlations for
each cell type pair. This helps in assessing the similarity between
cells in the reference and query datasets and provides insights into the
reliability of cell type annotations.</p>
<p>To use the <code><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation()</a></code>
function, you need to supply it with two <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects: one for the query cells and one for the reference cells. The
function also requires column names specifying cell type annotations in
both datasets, and optionally a vector of cell types to include in the
analysis. Additionally, you can specify a subset of principal components
to use, or use the raw data directly if <code>pc_subset</code> is set to
<code>NULL</code>. The correlation method can be either “spearman” or
“pearson”.</p>
<p>Here’s an example of how to use this function:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute pairwise correlations between specified cell types</span></span>
<span><span class="va">cor_matrix_avg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation</a></span><span class="op">(</span></span>
<span>  query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>  reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>  query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>  ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>  cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span><span class="op">)</span>, </span>
<span>  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>  correlation_method <span class="op">=</span> <span class="st">"spearman"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize the average pairwise correlation matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cor_matrix_avg</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/StatisticalMeasures/calculateAveragePairwiseCorrelation.png"></p>
<p>In this example, <code><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation()</a></code>
computes the average pairwise correlations for the cell types “CD4”,
“CD8”, and “B_and_plasma”. This function is particularly useful for
understanding the relationships between different cell types in your
single-cell datasets and evaluating how well cell types in the query
data correspond to those in the reference data.</p>
</div>
<div class="section level2">
<h2 id="regresspc">
<code>regressPC()</code><a class="anchor" aria-label="anchor" href="#regresspc"></a>
</h2>
<p>The <code><a href="../reference/regressPC.html">regressPC()</a></code> function performs linear regression of a
covariate of interest onto one or more principal components using data
from a <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
object. This method helps quantify the variance explained by a
covariate, which can be useful in applications such as quantifying batch
effects, assessing clustering homogeneity, and evaluating alignment
between query and reference datasets in cell type annotation
settings.</p>
<p>The function calculates the R-squared value from the linear
regression of the covariate onto each principal component. The variance
contribution of the covariate effect per principal component is computed
as the product of the variance explained by the principal component and
the R-squared value. The total variance explained by the covariate is
obtained by summing these contributions across all principal
components.</p>
<p>Here is how you can use the <code><a href="../reference/regressPC.html">regressPC()</a></code> function:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform regression analysis using only reference data</span></span>
<span><span class="va">regress_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regressPC.html">regressPC</a></span><span class="op">(</span></span>
<span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>    cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot results showing R-squared values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"r_squared"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/StatisticalMeasures/regressPC1.png"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot results showing p-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"p-value"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/StatisticalMeasures/regressPC2.png"></p>
<p>In this example, <code><a href="../reference/regressPC.html">regressPC()</a></code> is used to perform
regression analysis on principal components 1 to 15 from the reference
dataset. The results are then visualized using plots showing the
R-squared values and p-values.</p>
<p>If you also have a query dataset and want to compare it with the
reference dataset, you can include it in the analysis:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform regression analysis using both reference and query data</span></span>
<span><span class="va">regress_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regressPC.html">regressPC</a></span><span class="op">(</span></span>
<span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>    cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>    pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot results showing R-squared values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"r_squared"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/StatisticalMeasures/regressPC3.png"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot results showing p-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"p-value"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/StatisticalMeasures/regressPC4.png"></p>
<p>In this case, the function projects the query data onto the principal
components of the reference data and performs regression analysis. The
results are visualized in the same way, providing insights into how well
the principal components from the query data align with those from the
reference data.</p>
<p>This function is useful for understanding the impact of different
covariates on principal components and for comparing how cell types in
different datasets are represented in the principal component space.</p>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>The functions introduced in this vignette offer powerful tools for
in-depth analysis and interpretation of single-cell transcriptomics
data. By applying <code><a href="../reference/calculateCramerPValue.html">calculateCramerPValue()</a></code> and
<code><a href="../reference/calculateHotellingPValue.html">calculateHotellingPValue()</a></code>, researchers can quantify
associations and differences between categorical variables and
multivariate observations, respectively.
<code><a href="../reference/calculateNearestNeighborProbabilities.html">calculateNearestNeighborProbabilities()</a></code> enhances clustering
analysis by evaluating the likelihood of cell assignments, while
<code><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation()</a></code> reveals relationships
between markers and features. Lastly, <code><a href="../reference/regressPC.html">regressPC()</a></code> provides
insights into how principal components relate to covariates, aiding in
the understanding of variance explained by different factors.</p>
<p>Together, these functions enable a comprehensive approach to
single-cell data analysis, facilitating the identification of key
relationships and enhancing the robustness of conclusions drawn from
complex datasets. By leveraging these tools, researchers can advance
their understanding of cellular processes and improve the accuracy of
data interpretations in single-cell genomics.</p>
<hr>
</div>
<div class="section level2">
<h2 id="r-session-info">R Session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h2>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] scDiagnostics_0.99.9 BiocStyle_2.32.1    

loaded via a namespace (and not attached):
 [1] SummarizedExperiment_1.34.0 biglm_0.9-3                
 [3] gtable_0.3.5                xfun_0.47                  
 [5] bslib_0.8.0                 ggplot2_3.5.1              
 [7] htmlwidgets_1.6.4           Biobase_2.64.0             
 [9] lattice_0.22-6              generics_0.1.3             
[11] vctrs_0.6.5                 tools_4.4.1                
[13] stats4_4.4.1                tibble_3.2.1               
[15] fansi_1.0.6                 pkgconfig_2.0.3            
[17] Matrix_1.7-0                desc_1.4.3                 
[19] S4Vectors_0.42.1            lifecycle_1.0.4            
[21] GenomeInfoDbData_1.2.12     farver_2.1.2               
[23] compiler_4.4.1              textshaping_0.4.0          
[25] munsell_0.5.1               GenomeInfoDb_1.40.1        
[27] htmltools_0.5.8.1           sass_0.4.9                 
[29] yaml_2.3.10                 pkgdown_2.1.1              
[31] pillar_1.9.0                crayon_1.5.3               
[33] jquerylib_0.1.4             MASS_7.3-60.2              
[35] SingleCellExperiment_1.26.0 DelayedArray_0.30.1        
[37] cachem_1.1.0                boot_1.3-30                
[39] abind_1.4-8                 tidyselect_1.2.1           
[41] digest_0.6.37               dplyr_1.1.4                
[43] bookdown_0.40               labeling_0.4.3             
[45] fastmap_1.2.0               grid_4.4.1                 
[47] colorspace_2.1-1            cli_3.6.3                  
[49] SparseArray_1.4.8           magrittr_2.0.3             
[51] S4Arrays_1.4.1              utf8_1.2.4                 
[53] withr_3.0.1                 UCSC.utils_1.0.0           
[55] scales_1.3.0                rmarkdown_2.28             
[57] XVector_0.44.0              httr_1.4.7                 
[59] matrixStats_1.4.1           ragg_1.3.3                 
[61] evaluate_1.0.0              knitr_1.48                 
[63] GenomicRanges_1.56.1        IRanges_2.38.1             
[65] rlang_1.1.4                 Rcpp_1.0.13                
[67] DBI_1.2.3                   glue_1.7.0                 
[69] BiocManager_1.30.25         cramer_0.9-4               
[71] BiocGenerics_0.50.0         speedglm_0.3-5             
[73] jsonlite_1.8.8              R6_2.5.1                   
[75] MatrixGenerics_1.16.0       systemfonts_1.1.0          
[77] fs_1.6.4                    zlibbioc_1.50.0            </code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
