<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Visualization of Cell Type Annotations • scDiagnostics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Visualization of Cell Type Annotations">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scDiagnostics.html">Getting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/DatasetMarkerGeneAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection and Analysis of Annotation Anomalies</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Visualization of Cell Type Annotations</h1>
                        <h4 data-toc-skip class="author">Anthony
Christidis</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Andrew
Ghazi</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Smriti
Chawla</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Nitesh
Turaga</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Ludwig
Geistlinger</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Robert
Gentleman</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/main/vignettes/VisualizationTools.Rmd" class="external-link"><code>vignettes/VisualizationTools.Rmd</code></a></small>
      <div class="hidden name"><code>VisualizationTools.Rmd</code></div>

    </address>
</address>
</address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette demonstrates the use <code><a href="../reference/plotCellTypeMDS.html">plotCellTypeMDS()</a></code>,
<code><a href="../reference/plotCellTypePCA.html">plotCellTypePCA()</a></code>, <code><a href="../reference/boxplotPCA.html">boxplotPCA()</a></code> and
<code><a href="../reference/calculateDiscriminantSpace.html">calculateDiscriminantSpace()</a></code>, which are designed to help
assess and compare cell types between query and reference datasets using
Multidimensional Scaling (MDS) and Principal Component Analysis (PCA),
and Fisher Discriminant Analysis (FDA) respectively.</p>
<p>This vignette also demonstrates how to visualize gene expression data
from single-cell RNA sequencing (scRNA-seq) experiments using two
functions: <code><a href="../reference/plotGeneExpressionDimred.html">plotGeneExpressionDimred()</a></code> and
<code><a href="../reference/plotMarkerExpression.html">plotMarkerExpression()</a></code>. These functions allow researchers
to explore gene expression patterns across various dimensions and
compare expression distributions between different datasets or cell
types.</p>
<p>Lastly this vignette will demonstrate how to visualize quality
control (QC) scores and annotation scores, potentially identifying
relationship between QC statistics (e.g., total library size or
percentage of mitochondrial genes) and cell type annotation scores.</p>
</div>
<div class="section level2">
<h2 id="preliminaries">Preliminaries<a class="anchor" aria-label="anchor" href="#preliminaries"></a>
</h2>
<p>In the context of the <code>scDiagnostics</code> package, this
vignette illustrates how to evaluate cell type annotations using two
distinct datasets:</p>
<ul>
<li><p><code>reference_data</code>: This dataset consists of
meticulously curated cell type annotations assigned by experts,
providing a robust benchmark for assessing the quality of annotations.
It serves as a standard reference to evaluate and detect anomalies or
inconsistencies within the cell type annotations.</p></li>
<li><p><code>query_data</code>: This dataset includes annotations both
from expert assessments and those generated by the <em><a href="https://bioconductor.org/packages/3.21/SingleR" class="external-link">SingleR</a></em>
package. By comparing these annotations, you can identify discrepancies
between manual and automated results, thereby pinpointing potential
inconsistencies or areas that may need further scrutiny.</p></li>
<li><p><code>qc_data</code>: This dataset includes data from
haematopoietic stem and progenitor cells, including QC metrics like
library size and mitochondrial content, <em><a href="https://bioconductor.org/packages/3.21/SingleR" class="external-link">SingleR</a></em>-predicted
cell types and annotation scores indicating prediction
confidence.</p></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/scDiagnostics" class="external-link">scDiagnostics</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"qc_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization-of-query-vs--reference-dataset">Visualization of Query vs. Reference Dataset<a class="anchor" aria-label="anchor" href="#visualization-of-query-vs--reference-dataset"></a>
</h2>
<div class="section level3">
<h3 id="plot-reference-and-query-cell-types-using-mds">Plot Reference and Query Cell Types Using MDS<a class="anchor" aria-label="anchor" href="#plot-reference-and-query-cell-types-using-mds"></a>
</h3>
<p>The <code><a href="../reference/plotCellTypeMDS.html">plotCellTypeMDS()</a></code> function creates a scatter plot
using Multidimensional Scaling (MDS) to visualize the similarity between
cell types in query and reference datasets. This function generates a
MDS plot that colors cells according to their types based on a
dissimilarity matrix calculated from a specified gene set.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate the MDS scatter plot with cell type coloring</span></span>
<span><span class="fu"><a href="../reference/plotCellTypeMDS.html">plotCellTypeMDS</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>                query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing MDS from expression data.</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/VisualizationTools/plotCellTypeMDS.png"></p>
</div>
<div class="section level3">
<h3 id="plot-principal-components-for-different-cell-types">Plot Principal Components for Different Cell Types<a class="anchor" aria-label="anchor" href="#plot-principal-components-for-different-cell-types"></a>
</h3>
<p>The <code><a href="../reference/plotCellTypePCA.html">plotCellTypePCA()</a></code> function projects the query
dataset onto the PCA space of the reference dataset. It then plots
specified principal components for different cell types, allowing
comparison of PCA results between query and reference datasets.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate the MDS scatter plot with cell type coloring</span></span>
<span><span class="fu"><a href="../reference/plotCellTypePCA.html">plotCellTypePCA</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>                cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>                query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.503</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.502</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.462</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/VisualizationTools/plotCellTypePCA.png"></p>
<p>Here, <code><a href="../reference/plotCellTypePCA.html">plotCellTypePCA()</a></code> projects the query data into the
PCA space of the reference data and plots the specified principal
components. The <code>pc_subset</code> argument allows you to select
which principal components to display.</p>
</div>
<div class="section level3">
<h3 id="plot-principal-components-as-boxplots">Plot Principal Components as Boxplots<a class="anchor" aria-label="anchor" href="#plot-principal-components-as-boxplots"></a>
</h3>
<p>The <code><a href="../reference/boxplotPCA.html">boxplotPCA()</a></code> function provides a boxplot
visualization of principal components for different cell types across
query and reference datasets. This function helps in comparing the
distributions of PCA scores for specified cell types and datasets.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the PCA data as boxplots</span></span>
<span><span class="fu"><a href="../reference/boxplotPCA.html">boxplotPCA</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>           reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>           cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>           query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>           ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>           pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/VisualizationTools/boxplotPCA.png"></p>
<p>In this example, <code><a href="../reference/boxplotPCA.html">boxplotPCA()</a></code> generates boxplots for the
specified principal components, showing the distribution of PCA scores
for different cell types and datasets.</p>
</div>
<div class="section level3">
<h3 id="project-query-data-onto-discriminant-space-of-reference-data">Project Query Data onto Discriminant Space of Reference Data<a class="anchor" aria-label="anchor" href="#project-query-data-onto-discriminant-space-of-reference-data"></a>
</h3>
<p>The <code><a href="../reference/calculateDiscriminantSpace.html">calculateDiscriminantSpace()</a></code> function projects query
single-cell RNA-seq data onto a discriminant space defined by reference
data. This process involves using the reference data to identify key
variables, compute discriminant vectors, and project both the reference
and query data onto this space. The function then assesses the
similarity between the projections of the query data and the reference
data using cosine similarity and Mahalanobis distance.</p>
</div>
<div class="section level3">
<h3 id="function-details">Function Details<a class="anchor" aria-label="anchor" href="#function-details"></a>
</h3>
<p>The function performs the following steps for each pairwise
combination of cell types:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Identify Important Variables</strong>: Determine the most
significant variables that distinguish the two cell types from the
reference data.</li>
<li>
<strong>Compute Covariance Matrices</strong>: Estimate the
covariance matrices for each cell type using the Ledoit-Wolf shrinkage
method.</li>
<li>
<strong>Construct Scatter Matrices</strong>: Create within-class and
between-class scatter matrices.</li>
<li>
<strong>Solve Eigenvalue Problem</strong>: Solve the generalized
eigenvalue problem to obtain discriminant vectors.</li>
<li>
<strong>Project Data</strong>: Project both reference and query data
onto the discriminant space.</li>
<li>
<strong>Assess Similarity</strong>: Measure the similarity of the
query data projections to the reference projections using cosine
similarity and Mahalanobis distance.</li>
</ol>
<div class="section level4">
<h4 id="example-application">Example Application<a class="anchor" aria-label="anchor" href="#example-application"></a>
</h4>
<p>First, applying the function to the <code>SingleR</code>
annotation.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute important variables for all pairwise cell comparisons</span></span>
<span><span class="va">disc_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateDiscriminantSpace.html">calculateDiscriminantSpace</a></span><span class="op">(</span></span>
<span>  reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>  query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>  query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>  ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate scatter and boxplot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">disc_output</span>, plot_type <span class="op">=</span> <span class="st">"scatterplot"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.107</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.167</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.164</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/VisualizationTools/calculateDiscriminantSpace1.png"></p>
</div>
<div class="section level4">
<h4 id="using-mahalanobis-distance-for-anomaly-detection-in-single-cell-rna-seq-data">Using Mahalanobis Distance for Anomaly Detection in Single-Cell
RNA-Seq Data<a class="anchor" aria-label="anchor" href="#using-mahalanobis-distance-for-anomaly-detection-in-single-cell-rna-seq-data"></a>
</h4>
<p>We aim to evaluate whether a new batch of query cells fits well
within established reference cell types or if there are cells exhibiting
unusual characteristics. We will use the Mahalanobis distance calculated
by the <code><a href="../reference/calculateDiscriminantSpace.html">calculateDiscriminantSpace()</a></code> function to achieve
this.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform discriminant analysis and projection</span></span>
<span><span class="va">discriminant_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateDiscriminantSpace.html">calculateDiscriminantSpace</a></span><span class="op">(</span></span>
<span>  reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>  query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>  ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>  query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>  calculate_metrics <span class="op">=</span> <span class="cn">TRUE</span>,  <span class="co"># Compute Mahalanobis distance and cosine similarity</span></span>
<span>  alpha <span class="op">=</span> <span class="fl">0.01</span>  <span class="co"># Significance level for Mahalanobis distance cutoff</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>You can extract and analyze the Mahalanobis distances of query cells
relative to reference cell types. High Mahalanobis distances may
indicate potential anomalies or novel cell states.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract Mahalanobis distances</span></span>
<span><span class="va">mahalanobis_distances</span> <span class="op">&lt;-</span> <span class="va">discriminant_results</span><span class="op">$</span><span class="va">query_mahalanobis_dist</span></span>
<span></span>
<span><span class="co"># Identify anomalies based on a threshold</span></span>
<span><span class="va">threshold</span> <span class="op">&lt;-</span> <span class="va">discriminant_results</span><span class="op">$</span><span class="va">mahalanobis_crit</span>  <span class="co"># Use the computed cutoff value</span></span>
<span><span class="va">anomalies</span> <span class="op">&lt;-</span> <span class="va">mahalanobis_distances</span><span class="op">[</span><span class="va">mahalanobis_distances</span> <span class="op">&gt;</span> <span class="va">threshold</span><span class="op">]</span></span></code></pre></div>
<p>You can now create visualizations to illustrate the distribution of
Mahalanobis distances and highlight potential anomalies.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load ggplot2 for visualization</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert distances to a data frame</span></span>
<span><span class="va">distances_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Distance <span class="op">=</span> <span class="va">mahalanobis_distances</span>,</span>
<span>                           Anomaly <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">mahalanobis_distances</span> <span class="op">&gt;</span> <span class="va">threshold</span>, <span class="st">"Anomaly"</span>, <span class="st">"Normal"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot histogram of Mahalanobis distances</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">distances_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Distance</span>, fill <span class="op">=</span> <span class="va">Anomaly</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.5</span>, position <span class="op">=</span> <span class="st">"identity"</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Histogram of Mahalanobis Distances"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Mahalanobis Distance"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Frequency"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/VisualizationTools/calculateDiscriminantSpace2.png"></p>
<ul>
<li><p><strong>Low Mahalanobis Distances</strong>: These query cells
align well with the reference cell types, suggesting they are similar to
the established cell types.</p></li>
<li><p><strong>High Mahalanobis Distances</strong>: Cells with high
distances may represent outliers or novel, previously uncharacterized
cell types. Further investigation may be needed to understand these
anomalies.</p></li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="project-data-onto-sliced-inverse-regression-sir-space-of-reference-data">Project Data onto Sliced Inverse Regression (SIR) Space of Reference
Data<a class="anchor" aria-label="anchor" href="#project-data-onto-sliced-inverse-regression-sir-space-of-reference-data"></a>
</h3>
<p>The <code><a href="../reference/calculateSIRSpace.html">calculateSIRSpace()</a></code> function leverages Sliced
Inverse Regression (SIR) to analyze high-dimensional cell type data by
focusing on the directions that best separate different cell types and
their subtypes. The function projects data onto a lower-dimensional SIR
space, aiming to highlight the most informative features for
distinguishing between cell types.</p>
<p>The process starts by handling each cell type as a distinct slice of
the data, allowing for a more detailed analysis. When
<code>multiple_cond_means</code> is set to <code>TRUE</code>, the
function refines this by computing conditional means for each cell type,
using clustering to capture finer distinctions within each type. It then
performs Singular Value Decomposition (SVD) on these refined means to
uncover the principal directions that most effectively differentiate the
cell types.</p>
<p>Ultimately, the function provides projections that reveal how cell
types and their subtypes relate to each other in a reduced space,
facilitating a deeper understanding of cellular variability and
distinctions in complex datasets.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate the SIR space projections</span></span>
<span><span class="va">sir_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateSIRSpace.html">calculateSIRSpace</a></span><span class="op">(</span></span>
<span>  reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>  query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>  query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>  ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>  multiple_cond_means <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the SIR space projections</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sir_output</span>,</span>
<span>     sir_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>     cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>     lower_facet <span class="op">=</span> <span class="st">"scatter"</span>,</span>
<span>     diagonal_facet <span class="op">=</span> <span class="st">"boxplot"</span>,</span>
<span>     upper_facet <span class="op">=</span> <span class="st">"blank"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/VisualizationTools/calculateSIRSpace1.png"></p>
<p>In this example, we compute the SIR space projections using the
<code><a href="../reference/calculateSIRSpace.html">calculateSIRSpace()</a></code> function, and then we visualize the
projections with a boxplot to analyze the distribution of the data in
the SIR space. This example helps you understand how different cell
types and their subtypes are projected into a common space, facilitating
their comparison and interpretation.</p>
<p>The plot below visualizes the contribution of different markers to
the variance along the SIR dimensions, helping identify which markers
are most influential in differentiating e.g. between B cells (including
plasma cells) from other cell types.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the top contributing markers</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sir_output</span>, </span>
<span>     plot_type <span class="op">=</span> <span class="st">"loadings"</span>, </span>
<span>     sir_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>     n_top <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/VisualizationTools/calculateSIRSpace2.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="visualization-of-marker-expressions">Visualization of Marker Expressions<a class="anchor" aria-label="anchor" href="#visualization-of-marker-expressions"></a>
</h2>
<div class="section level3">
<h3 id="visualizing-gene-expression-in-reduced-dimensions">Visualizing Gene Expression in Reduced Dimensions<a class="anchor" aria-label="anchor" href="#visualizing-gene-expression-in-reduced-dimensions"></a>
</h3>
<p>The <code><a href="../reference/plotGeneExpressionDimred.html">plotGeneExpressionDimred()</a></code> function plots the
expression of a specific gene on a dimensional reduction plot. This
function is useful for exploring how gene expression varies across cells
in the context of reduced dimensions, such as those derived from PCA,
t-SNE, or UMAP. Each point on the plot represents a single cell,
color-coded according to the expression level of the specified gene.</p>
<p>Let’s visualize the expression of the gene <strong>NKG7</strong> for
CD4 on a PCA plot:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize VPREB3 gene on a PCA plot</span></span>
<span><span class="fu"><a href="../reference/plotGeneExpressionDimred.html">plotGeneExpressionDimred</a></span><span class="op">(</span>sce_object <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>                         method <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>                         pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>                         feature <span class="op">=</span> <span class="st">"NKG7"</span>,</span>
<span>                         cell_types <span class="op">=</span> <span class="st">"CD4"</span>,</span>
<span>                         cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/VisualizationTools/plotGeneExpressionDimred.png"></p>
<p>In this example, we visualize the expression of the
<strong>NKG7</strong> gene on a PCA plot derived from the
<code>query_data</code> dataset. The <code>method</code> argument
specifies the dimensional reduction technique, which in this case is
“PCA.” The <code>pc_subset</code> argument allows us to focus on
specific principal components (PCs), here selecting the first five. The
<code>feature</code> argument is the name of the gene we want to
visualize.</p>
<p>When executed, this function generates a scatter plot with cells
colored by their expression levels of <strong>NKG7</strong>. The
resulting plot can help identify clusters of cells with similar
expression patterns or highlight how expression varies across the
dataset.</p>
<p>You can also visualize gene expression using other dimensional
reduction methods, such as t-SNE or UMAP.</p>
</div>
<div class="section level3">
<h3 id="plotting-gene-expression-distribution">Plotting Gene Expression Distribution<a class="anchor" aria-label="anchor" href="#plotting-gene-expression-distribution"></a>
</h3>
<p>The <code><a href="../reference/plotMarkerExpression.html">plotMarkerExpression()</a></code> function compares the
distribution of a specific gene’s expression across an overall dataset
and within specific cell types. It generates density plots to visualize
these distributions, providing a graphical means of assessing the
similarity between reference and query datasets. The function also
applies a z-rank normalization to make expression values comparable
between datasets.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMarkerExpression.html">plotMarkerExpression</a></span><span class="op">(</span></span>
<span>    reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>    gene_name <span class="op">=</span> <span class="st">"NKG7"</span>,</span>
<span>    cell_type <span class="op">=</span> <span class="st">"CD4"</span>,</span>
<span>    normalization <span class="op">=</span> <span class="st">"z_score"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/VisualizationTools/plotMarkerExpression.png"></p>
<p>In this example, the function compares the expression distribution of
the <strong>VPREB3</strong> gene between the reference and query
datasets. The <code>ref_cell_type_col</code> and
<code>query_cell_type_col</code> arguments specify the columns in
<code>colData</code> that identify the cell types in the reference and
query datasets, respectively. The <code>cell_type</code> argument
specifies the cell type of interest, and <code>gene_name</code> is the
gene whose expression distribution we wish to compare.</p>
<p>When executed, this function generates two density plots: one showing
the overall gene expression distribution across all cells and another
focusing on the specified cell type. These plots help identify whether
the gene’s expression levels are similar between the reference and query
datasets, providing insight into dataset alignment.</p>
</div>
</div>
<div class="section level2">
<h2 id="visualization-of-qc-and-annotation-scores">Visualization of QC and Annotation Scores<a class="anchor" aria-label="anchor" href="#visualization-of-qc-and-annotation-scores"></a>
</h2>
<div class="section level3">
<h3 id="scatter-plot-qc-stats-vs-cell-type-annotation-scores">Scatter Plot: QC Stats vs Cell Type Annotation Scores<a class="anchor" aria-label="anchor" href="#scatter-plot-qc-stats-vs-cell-type-annotation-scores"></a>
</h3>
<p>The scatter plot visualizes the relationship between QC statistics
(e.g., total library size or percentage of mitochondrial genes) and cell
type annotation scores. This plot helps in understanding how QC metrics
influence or correlate with the predicted cell types.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remove cell types with very few cells</span></span>
<span><span class="va">qc_data_subset</span> <span class="op">&lt;-</span> <span class="va">qc_data</span><span class="op">[</span>, <span class="op">!</span><span class="op">(</span><span class="va">qc_data</span><span class="op">$</span><span class="va">SingleR_annotation</span> </span>
<span>                              <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Chondrocytes"</span>, <span class="st">"DC"</span>, </span>
<span>                                     <span class="st">"Neurons"</span>,<span class="st">"Platelets"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Generate scatter plot</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotQCvsAnnotation.html">plotQCvsAnnotation</a></span><span class="op">(</span>sce_object <span class="op">=</span> <span class="va">qc_data_subset</span>,</span>
<span>                         cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>                         qc_col <span class="op">=</span> <span class="st">"total"</span>,</span>
<span>                         score_col <span class="op">=</span> <span class="st">"annotation_scores"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Library Size"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/VisualizationTools/plotQCvsAnnotation.png"></p>
<p>A scatter plot can reveal patterns such as whether cells with higher
library sizes or mitochondrial content tend to be associated with
specific annotations. For instance, cells with unusually high
mitochondrial content might be identified as low-quality or stressed,
potentially affecting their annotations.</p>
</div>
<div class="section level3">
<h3 id="histograms-qc-stats-and-annotation-scores-visualization">Histograms: QC Stats and Annotation Scores Visualization<a class="anchor" aria-label="anchor" href="#histograms-qc-stats-and-annotation-scores-visualization"></a>
</h3>
<p>Histograms provide a distribution view of QC metrics and annotation
scores. They help in evaluating the range, central tendency, and spread
of these variables across cells.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate histograms</span></span>
<span><span class="fu"><a href="../reference/histQCvsAnnotation.html">histQCvsAnnotation</a></span><span class="op">(</span>sce_object <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                   cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                   qc_col <span class="op">=</span> <span class="st">"percent_mito"</span>, </span>
<span>                   score_col <span class="op">=</span> <span class="st">"annotation_scores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/VisualizationTools/histQCvsAnnotation.png"></p>
<p>Histograms are useful for assessing the overall distribution of QC
metrics and annotation scores. For example, if the majority of cells
have high percent_mito, it might indicate that many cells are stressed
or dying, which could impact the quality of the data.</p>
</div>
</div>
<div class="section level2">
<h2 id="visualization-of-gene-sets-or-pathway-scores-on-dimensional-reduction-plots">Visualization of Gene Sets or Pathway Scores on Dimensional
Reduction Plots<a class="anchor" aria-label="anchor" href="#visualization-of-gene-sets-or-pathway-scores-on-dimensional-reduction-plots"></a>
</h2>
<p>Dimensional reduction plots (PCA, t-SNE, UMAP) are used to visualize
the relationships between cells in reduced dimensions. Overlaying gene
set scores on these plots provides insights into how specific gene
activities are distributed across cell clusters.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot gene set scores on PCA</span></span>
<span><span class="fu"><a href="../reference/plotGeneSetScores.html">plotGeneSetScores</a></span><span class="op">(</span>sce_object <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                  cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>                  cell_types <span class="op">=</span> <span class="st">"CD8"</span>,</span>
<span>                  method <span class="op">=</span> <span class="st">"PCA"</span>, </span>
<span>                  score_col <span class="op">=</span> <span class="st">"gene_set_scores"</span>, </span>
<span>                  pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/ccb-hms/scDiagnostics/main/inst/extdata/compressed/VisualizationTools/plotGeneSetScores.png"></p>
<p>By visualizing gene set scores on PCA or UMAP plots, one can identify
clusters of cells with high or low gene set activities. This can help in
understanding the biological relevance of different gene sets or
pathways in various cell states or types.</p>
<hr>
</div>
<div class="section level2">
<h2 id="r-session-info">R Session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h2>
<pre><code>R version 4.5.1 (2025-06-13)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_4.0.0       scDiagnostics_1.1.0 BiocStyle_2.36.0   

loaded via a namespace (and not attached):
 [1] SummarizedExperiment_1.38.1 gtable_0.3.6               
 [3] xfun_0.53                   bslib_0.9.0                
 [5] htmlwidgets_1.6.4           GGally_2.4.0               
 [7] Biobase_2.68.0              lattice_0.22-7             
 [9] vctrs_0.6.5                 tools_4.5.1                
[11] generics_0.1.4              parallel_4.5.1             
[13] stats4_4.5.1                tibble_3.3.0               
[15] cluster_2.1.8.1             BiocNeighbors_2.2.0        
[17] pkgconfig_2.0.3             Matrix_1.7-3               
[19] RColorBrewer_1.1-3          S7_0.2.0                   
[21] desc_1.4.3                  S4Vectors_0.46.0           
[23] ggridges_0.5.7              lifecycle_1.0.4            
[25] GenomeInfoDbData_1.2.14     compiler_4.5.1             
[27] farver_2.1.2                textshaping_1.0.3          
[29] bluster_1.18.0              codetools_0.2-20           
[31] GenomeInfoDb_1.44.3         htmltools_0.5.8.1          
[33] sass_0.4.10                 yaml_2.3.10                
[35] tidyr_1.3.1                 pkgdown_2.1.3              
[37] pillar_1.11.1               crayon_1.5.3               
[39] jquerylib_0.1.4             BiocParallel_1.42.2        
[41] SingleCellExperiment_1.30.1 DelayedArray_0.34.1        
[43] cachem_1.1.0                abind_1.4-8                
[45] ggstats_0.11.0              tidyselect_1.2.1           
[47] digest_0.6.37               purrr_1.1.0                
[49] dplyr_1.1.4                 bookdown_0.44              
[51] labeling_0.4.3              fastmap_1.2.0              
[53] grid_4.5.1                  cli_3.6.5                  
[55] SparseArray_1.8.1           magrittr_2.0.4             
[57] S4Arrays_1.8.1              withr_3.0.2                
[59] UCSC.utils_1.4.0            scales_1.4.0               
[61] rmarkdown_2.30              XVector_0.48.0             
[63] httr_1.4.7                  matrixStats_1.5.0          
[65] igraph_2.1.4                ranger_0.17.0              
[67] ragg_1.5.0                  evaluate_1.0.5             
[69] knitr_1.50                  GenomicRanges_1.60.0       
[71] IRanges_2.42.0              rlang_1.1.6                
[73] Rcpp_1.1.0                  glue_1.8.0                 
[75] BiocManager_1.30.26         BiocGenerics_0.54.0        
[77] jsonlite_2.0.0              R6_2.6.1                   
[79] MatrixGenerics_1.20.0       systemfonts_1.3.1          
[81] fs_1.6.6                   </code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
